// @base-ui/react/preview-card/arrow/PreviewCardArrowDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PreviewCardArrowDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let PreviewCardArrowDataAttributes = exports.PreviewCardArrowDataAttributes = function (PreviewCardArrowDataAttributes) {
  /**
   * Present when the preview card is open.
   */
  PreviewCardArrowDataAttributes[PreviewCardArrowDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the preview card is closed.
   */
  PreviewCardArrowDataAttributes[PreviewCardArrowDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  PreviewCardArrowDataAttributes[PreviewCardArrowDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  PreviewCardArrowDataAttributes[PreviewCardArrowDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  /**
   * Present when the preview card arrow is uncentered.
   */
  PreviewCardArrowDataAttributes["uncentered"] = "data-uncentered";
  return PreviewCardArrowDataAttributes;
}({});
