// @base-ui/react/esm/combobox/popup/ComboboxPopupDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let ComboboxPopupDataAttributes = function (ComboboxPopupDataAttributes) {
  /**
   * Present when the popup is open.
   */
  ComboboxPopupDataAttributes[ComboboxPopupDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  ComboboxPopupDataAttributes[ComboboxPopupDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the popup is animating in.
   */
  ComboboxPopupDataAttributes[ComboboxPopupDataAttributes["startingStyle"] = CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the popup is animating out.
   */
  ComboboxPopupDataAttributes[ComboboxPopupDataAttributes["endingStyle"] = CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  ComboboxPopupDataAttributes[ComboboxPopupDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  ComboboxPopupDataAttributes[ComboboxPopupDataAttributes["align"] = CommonPopupDataAttributes.align] = "align";
  /**
   * Present if animations should be instant.
   * @type {'click' | 'dismiss'}
   */
  ComboboxPopupDataAttributes["instant"] = "data-instant";
  /**
   * Present when the items list is empty.
   */
  ComboboxPopupDataAttributes["empty"] = "data-empty";
  return ComboboxPopupDataAttributes;
}({});
