// @base-ui/react/switch/thumb/SwitchThumbDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SwitchThumbDataAttributes = void 0;
let SwitchThumbDataAttributes = exports.SwitchThumbDataAttributes = /*#__PURE__*/function (SwitchThumbDataAttributes) {
  /**
   * Present when the switch is checked.
   */
  SwitchThumbDataAttributes["checked"] = "data-checked";
  /**
   * Present when the switch is not checked.
   */
  SwitchThumbDataAttributes["unchecked"] = "data-unchecked";
  /**
   * Present when the switch is disabled.
   */
  SwitchThumbDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the switch is readonly.
   */
  SwitchThumbDataAttributes["readonly"] = "data-readonly";
  /**
   * Present when the switch is required.
   */
  SwitchThumbDataAttributes["required"] = "data-required";
  /**
   * Present when the switch is in valid state (when wrapped in Field.Root).
   */
  SwitchThumbDataAttributes["valid"] = "data-valid";
  /**
   * Present when the switch is in invalid state (when wrapped in Field.Root).
   */
  SwitchThumbDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the switch has been touched (when wrapped in Field.Root).
   */
  SwitchThumbDataAttributes["touched"] = "data-touched";
  /**
   * Present when the switch's value has changed (when wrapped in Field.Root).
   */
  SwitchThumbDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the switch is active (when wrapped in Field.Root).
   */
  SwitchThumbDataAttributes["filled"] = "data-filled";
  /**
   * Present when the switch is focused (when wrapped in Field.Root).
   */
  SwitchThumbDataAttributes["focused"] = "data-focused";
  return SwitchThumbDataAttributes;
}({});
