// @base-ui/react/navigation-menu/content/NavigationMenuContentDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NavigationMenuContentDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let NavigationMenuContentDataAttributes = exports.NavigationMenuContentDataAttributes = function (NavigationMenuContentDataAttributes) {
  /**
   * Present when the popup is open.
   */
  NavigationMenuContentDataAttributes[NavigationMenuContentDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  NavigationMenuContentDataAttributes[NavigationMenuContentDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the content is animating in.
   */
  NavigationMenuContentDataAttributes[NavigationMenuContentDataAttributes["startingStyle"] = _popupStateMapping.CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the content is animating out.
   */
  NavigationMenuContentDataAttributes[NavigationMenuContentDataAttributes["endingStyle"] = _popupStateMapping.CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Which direction another trigger was activated from.
   */
  NavigationMenuContentDataAttributes["activationDirection"] = "data-activation-direction";
  return NavigationMenuContentDataAttributes;
}({});
