// @base-ui/react/esm/select/scroll-down-arrow/SelectScrollDownArrowDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let SelectScrollDownArrowDataAttributes = function (SelectScrollDownArrowDataAttributes) {
  /**
   * Present when the scroll arrow is animating in.
   */
  SelectScrollDownArrowDataAttributes[SelectScrollDownArrowDataAttributes["startingStyle"] = CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the scroll arrow is animating out.
   */
  SelectScrollDownArrowDataAttributes[SelectScrollDownArrowDataAttributes["endingStyle"] = CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates the direction of the scroll arrow.
   * @type {'down'}
   */
  SelectScrollDownArrowDataAttributes["direction"] = "data-direction";
  /**
   * Present when the scroll arrow is visible.
   */
  SelectScrollDownArrowDataAttributes["visible"] = "data-visible";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'none' | 'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  SelectScrollDownArrowDataAttributes[SelectScrollDownArrowDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  return SelectScrollDownArrowDataAttributes;
}({});
