// @base-ui/react/esm/drawer/viewport/DrawerViewportDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let DrawerViewportDataAttributes = function (DrawerViewportDataAttributes) {
  /**
   * Present when the drawer is open.
   */
  DrawerViewportDataAttributes[DrawerViewportDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the drawer is closed.
   */
  DrawerViewportDataAttributes[DrawerViewportDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the drawer is animating in.
   */
  DrawerViewportDataAttributes[DrawerViewportDataAttributes["startingStyle"] = CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the drawer is animating out.
   */
  DrawerViewportDataAttributes[DrawerViewportDataAttributes["endingStyle"] = CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Present when the drawer is nested within another drawer.
   */
  DrawerViewportDataAttributes["nested"] = "data-nested";
  return DrawerViewportDataAttributes;
}({});
