// @base-ui/react/select/arrow/SelectArrowDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectArrowDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let SelectArrowDataAttributes = exports.SelectArrowDataAttributes = function (SelectArrowDataAttributes) {
  /**
   * Present when the select popup is open.
   */
  SelectArrowDataAttributes[SelectArrowDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the select popup is closed.
   */
  SelectArrowDataAttributes[SelectArrowDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'none' | 'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  SelectArrowDataAttributes[SelectArrowDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  SelectArrowDataAttributes[SelectArrowDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  /**
   * Present when the select arrow is uncentered.
   */
  SelectArrowDataAttributes["uncentered"] = "data-uncentered";
  return SelectArrowDataAttributes;
}({});
