// @base-ui/react/drawer/index.parts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Backdrop", {
  enumerable: true,
  get: function () {
    return _DrawerBackdrop.DrawerBackdrop;
  }
});
Object.defineProperty(exports, "Close", {
  enumerable: true,
  get: function () {
    return _DrawerClose.DrawerClose;
  }
});
Object.defineProperty(exports, "Content", {
  enumerable: true,
  get: function () {
    return _DrawerContent.DrawerContent;
  }
});
Object.defineProperty(exports, "Description", {
  enumerable: true,
  get: function () {
    return _DrawerDescription.DrawerDescription;
  }
});
Object.defineProperty(exports, "Handle", {
  enumerable: true,
  get: function () {
    return _DialogHandle.DialogHandle;
  }
});
Object.defineProperty(exports, "Indent", {
  enumerable: true,
  get: function () {
    return _DrawerIndent.DrawerIndent;
  }
});
Object.defineProperty(exports, "IndentBackground", {
  enumerable: true,
  get: function () {
    return _DrawerIndentBackground.DrawerIndentBackground;
  }
});
Object.defineProperty(exports, "Popup", {
  enumerable: true,
  get: function () {
    return _DrawerPopup.DrawerPopup;
  }
});
Object.defineProperty(exports, "Portal", {
  enumerable: true,
  get: function () {
    return _DrawerPortal.DrawerPortal;
  }
});
Object.defineProperty(exports, "Provider", {
  enumerable: true,
  get: function () {
    return _DrawerProvider.DrawerProvider;
  }
});
Object.defineProperty(exports, "Root", {
  enumerable: true,
  get: function () {
    return _DrawerRoot.DrawerRoot;
  }
});
Object.defineProperty(exports, "SwipeArea", {
  enumerable: true,
  get: function () {
    return _DrawerSwipeArea.DrawerSwipeArea;
  }
});
Object.defineProperty(exports, "Title", {
  enumerable: true,
  get: function () {
    return _DrawerTitle.DrawerTitle;
  }
});
Object.defineProperty(exports, "Trigger", {
  enumerable: true,
  get: function () {
    return _DrawerTrigger.DrawerTrigger;
  }
});
Object.defineProperty(exports, "Viewport", {
  enumerable: true,
  get: function () {
    return _DrawerViewport.DrawerViewport;
  }
});
Object.defineProperty(exports, "createHandle", {
  enumerable: true,
  get: function () {
    return _DialogHandle.createDialogHandle;
  }
});
var _DrawerBackdrop = require("./backdrop/DrawerBackdrop");
var _DrawerClose = require("./close/DrawerClose");
var _DrawerContent = require("./content/DrawerContent");
var _DrawerDescription = require("./description/DrawerDescription");
var _DrawerIndent = require("./indent/DrawerIndent");
var _DrawerIndentBackground = require("./indent-background/DrawerIndentBackground");
var _DrawerPopup = require("./popup/DrawerPopup");
var _DrawerPortal = require("./portal/DrawerPortal");
var _DrawerProvider = require("./provider/DrawerProvider");
var _DrawerRoot = require("./root/DrawerRoot");
var _DrawerSwipeArea = require("./swipe-area/DrawerSwipeArea");
var _DrawerTitle = require("./title/DrawerTitle");
var _DrawerTrigger = require("./trigger/DrawerTrigger");
var _DrawerViewport = require("./viewport/DrawerViewport");
var _DialogHandle = require("../dialog/store/DialogHandle");
