// @base-ui/react/popover/positioner/PopoverPositionerDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PopoverPositionerDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let PopoverPositionerDataAttributes = exports.PopoverPositionerDataAttributes = function (PopoverPositionerDataAttributes) {
  /**
   * Present when the popup is open.
   */
  PopoverPositionerDataAttributes[PopoverPositionerDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  PopoverPositionerDataAttributes[PopoverPositionerDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  PopoverPositionerDataAttributes[PopoverPositionerDataAttributes["anchorHidden"] = _popupStateMapping.CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  PopoverPositionerDataAttributes[PopoverPositionerDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  PopoverPositionerDataAttributes[PopoverPositionerDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  return PopoverPositionerDataAttributes;
}({});
