// @base-ui/react/esm/navigation-menu/positioner/NavigationPositionerDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let NavigationMenuPositionerDataAttributes = function (NavigationMenuPositionerDataAttributes) {
  /**
   * Present when the popup is open.
   */
  NavigationMenuPositionerDataAttributes[NavigationMenuPositionerDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  NavigationMenuPositionerDataAttributes[NavigationMenuPositionerDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  NavigationMenuPositionerDataAttributes[NavigationMenuPositionerDataAttributes["anchorHidden"] = CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  NavigationMenuPositionerDataAttributes[NavigationMenuPositionerDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to the specified side.
   * @type {'start' | 'center' | 'end'}
   */
  NavigationMenuPositionerDataAttributes[NavigationMenuPositionerDataAttributes["align"] = CommonPopupDataAttributes.align] = "align";
  /**
   * Present if animations should be instant.
   */
  NavigationMenuPositionerDataAttributes["instant"] = "data-instant";
  return NavigationMenuPositionerDataAttributes;
}({});
