// @base-ui/react/esm/combobox/arrow/ComboboxArrowDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let ComboboxArrowDataAttributes = function (ComboboxArrowDataAttributes) {
  /**
   * Present when the popup is open.
   */
  ComboboxArrowDataAttributes[ComboboxArrowDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  ComboboxArrowDataAttributes[ComboboxArrowDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  ComboboxArrowDataAttributes[ComboboxArrowDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  ComboboxArrowDataAttributes[ComboboxArrowDataAttributes["align"] = CommonPopupDataAttributes.align] = "align";
  /**
   * Present when the arrow is uncentered.
   */
  ComboboxArrowDataAttributes["uncentered"] = "data-uncentered";
  return ComboboxArrowDataAttributes;
}({});
