// @base-ui/react/preview-card/viewport/PreviewCardViewportDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PreviewCardViewportDataAttributes = void 0;
let PreviewCardViewportDataAttributes = exports.PreviewCardViewportDataAttributes = /*#__PURE__*/function (PreviewCardViewportDataAttributes) {
  /**
   * Applied to the direct child of the viewport when no transitions are present or the new content when it's entering.
   */
  PreviewCardViewportDataAttributes["current"] = "data-current";
  /**
   * Applied to the direct child of the viewport that contains the exiting content when transitions are present.
   */
  PreviewCardViewportDataAttributes["previous"] = "data-previous";
  /**
   * Indicates the direction from which the popup was activated.
   * This can be used to create directional animations based on how the popup was triggered.
   * Contains space-separated values for both horizontal and vertical axes.
   * @type {`${'left' | 'right'} {'top' | 'bottom'}`}
   */
  PreviewCardViewportDataAttributes["activationDirection"] = "data-activation-direction";
  /**
   * Indicates that the viewport is currently transitioning between old and new content.
   */
  PreviewCardViewportDataAttributes["transitioning"] = "data-transitioning";
  /**
   * Present if animations should be instant.
   * @type {'delay' | 'dismiss' | 'focus'}
   */
  PreviewCardViewportDataAttributes["instant"] = "data-instant";
  return PreviewCardViewportDataAttributes;
}({});
