// @base-ui/react/esm/combobox/positioner/ComboboxPositionerDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let ComboboxPositionerDataAttributes = function (ComboboxPositionerDataAttributes) {
  /**
   * Present when the popup is open.
   */
  ComboboxPositionerDataAttributes[ComboboxPositionerDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  ComboboxPositionerDataAttributes[ComboboxPositionerDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  ComboboxPositionerDataAttributes[ComboboxPositionerDataAttributes["anchorHidden"] = CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  ComboboxPositionerDataAttributes[ComboboxPositionerDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  ComboboxPositionerDataAttributes[ComboboxPositionerDataAttributes["align"] = CommonPopupDataAttributes.align] = "align";
  /**
   * Present when the items list is empty.
   */
  ComboboxPositionerDataAttributes["empty"] = "data-empty";
  return ComboboxPositionerDataAttributes;
}({});
