// @base-ui/react/number-field/scrub-area-cursor/NumberFieldScrubAreaCursorDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NumberFieldScrubAreaCursorDataAttributes = void 0;
let NumberFieldScrubAreaCursorDataAttributes = exports.NumberFieldScrubAreaCursorDataAttributes = /*#__PURE__*/function (NumberFieldScrubAreaCursorDataAttributes) {
  /**
   * Present while scrubbing.
   */
  NumberFieldScrubAreaCursorDataAttributes["scrubbing"] = "data-scrubbing";
  /**
   * Present when the number field is disabled.
   */
  NumberFieldScrubAreaCursorDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the number field is readonly.
   */
  NumberFieldScrubAreaCursorDataAttributes["readonly"] = "data-readonly";
  /**
   * Present when the number field is required.
   */
  NumberFieldScrubAreaCursorDataAttributes["required"] = "data-required";
  /**
   * Present when the number field is in valid state (when wrapped in Field.Root).
   */
  NumberFieldScrubAreaCursorDataAttributes["valid"] = "data-valid";
  /**
   * Present when the number field is in invalid state (when wrapped in Field.Root).
   */
  NumberFieldScrubAreaCursorDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the number field has been touched (when wrapped in Field.Root).
   */
  NumberFieldScrubAreaCursorDataAttributes["touched"] = "data-touched";
  /**
   * Present when the number field's value has changed (when wrapped in Field.Root).
   */
  NumberFieldScrubAreaCursorDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the number field is filled (when wrapped in Field.Root).
   */
  NumberFieldScrubAreaCursorDataAttributes["filled"] = "data-filled";
  /**
   * Present when the number field is focused (when wrapped in Field.Root).
   */
  NumberFieldScrubAreaCursorDataAttributes["focused"] = "data-focused";
  return NumberFieldScrubAreaCursorDataAttributes;
}({});
