// @base-ui/react/toast/positioner/ToastPositionerCssVars.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ToastPositionerCssVars = void 0;
let ToastPositionerCssVars = exports.ToastPositionerCssVars = /*#__PURE__*/function (ToastPositionerCssVars) {
  /**
   * The available width between the anchor and the edge of the viewport.
   * @type {number}
   */
  ToastPositionerCssVars["availableWidth"] = "--available-width";
  /**
   * The available height between the anchor and the edge of the viewport.
   * @type {number}
   */
  ToastPositionerCssVars["availableHeight"] = "--available-height";
  /**
   * The anchor's width.
   * @type {number}
   */
  ToastPositionerCssVars["anchorWidth"] = "--anchor-width";
  /**
   * The anchor's height.
   * @type {number}
   */
  ToastPositionerCssVars["anchorHeight"] = "--anchor-height";
  /**
   * The coordinates that this element is anchored to. Used for animations and transitions.
   * @type {string}
   */
  ToastPositionerCssVars["transformOrigin"] = "--transform-origin";
  return ToastPositionerCssVars;
}({});
