// @base-ui/react/slider/control/SliderControlDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SliderControlDataAttributes = void 0;
let SliderControlDataAttributes = exports.SliderControlDataAttributes = /*#__PURE__*/function (SliderControlDataAttributes) {
  /**
   * Present while the user is dragging.
   */
  SliderControlDataAttributes["dragging"] = "data-dragging";
  /**
   * Indicates the orientation of the slider.
   * @type {'horizontal' | 'vertical'}
   */
  SliderControlDataAttributes["orientation"] = "data-orientation";
  /**
   * Present when the slider is disabled.
   */
  SliderControlDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the slider is in valid state (when wrapped in Field.Root).
   */
  SliderControlDataAttributes["valid"] = "data-valid";
  /**
   * Present when the slider is in invalid state (when wrapped in Field.Root).
   */
  SliderControlDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the slider has been touched (when wrapped in Field.Root).
   */
  SliderControlDataAttributes["touched"] = "data-touched";
  /**
   * Present when the slider's value has changed (when wrapped in Field.Root).
   */
  SliderControlDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the slider is focused (when wrapped in Field.Root).
   */
  SliderControlDataAttributes["focused"] = "data-focused";
  return SliderControlDataAttributes;
}({});
