// @base-ui/react/menu/radio-item-indicator/MenuRadioItemIndicatorDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MenuRadioItemIndicatorDataAttributes = void 0;
var _stateAttributesMapping = require("../../utils/stateAttributesMapping");
let MenuRadioItemIndicatorDataAttributes = exports.MenuRadioItemIndicatorDataAttributes = function (MenuRadioItemIndicatorDataAttributes) {
  /**
   * Present when the menu radio item is selected.
   */
  MenuRadioItemIndicatorDataAttributes["checked"] = "data-checked";
  /**
   * Present when the menu radio item is not selected.
   */
  MenuRadioItemIndicatorDataAttributes["unchecked"] = "data-unchecked";
  /**
   * Present when the menu radio item is disabled.
   */
  MenuRadioItemIndicatorDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the radio indicator is animating in.
   */
  MenuRadioItemIndicatorDataAttributes[MenuRadioItemIndicatorDataAttributes["startingStyle"] = _stateAttributesMapping.TransitionStatusDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the radio indicator is animating out.
   */
  MenuRadioItemIndicatorDataAttributes[MenuRadioItemIndicatorDataAttributes["endingStyle"] = _stateAttributesMapping.TransitionStatusDataAttributes.endingStyle] = "endingStyle";
  return MenuRadioItemIndicatorDataAttributes;
}({});
