// @base-ui/react/esm/navigation-menu/popup/NavigationMenuPopupDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let NavigationMenuPopupDataAttributes = function (NavigationMenuPopupDataAttributes) {
  /**
   * Present when the popup is open.
   */
  NavigationMenuPopupDataAttributes[NavigationMenuPopupDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  NavigationMenuPopupDataAttributes[NavigationMenuPopupDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the popup is animating in.
   */
  NavigationMenuPopupDataAttributes[NavigationMenuPopupDataAttributes["startingStyle"] = CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the popup is animating out.
   */
  NavigationMenuPopupDataAttributes[NavigationMenuPopupDataAttributes["endingStyle"] = CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  NavigationMenuPopupDataAttributes[NavigationMenuPopupDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to the specified side.
   * @type {'start' | 'center' | 'end'}
   */
  NavigationMenuPopupDataAttributes[NavigationMenuPopupDataAttributes["align"] = CommonPopupDataAttributes.align] = "align";
  return NavigationMenuPopupDataAttributes;
}({});
