// @base-ui/react/number-field/group/NumberFieldGroupDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NumberFieldGroupDataAttributes = void 0;
let NumberFieldGroupDataAttributes = exports.NumberFieldGroupDataAttributes = /*#__PURE__*/function (NumberFieldGroupDataAttributes) {
  /**
   * Present while scrubbing.
   */
  NumberFieldGroupDataAttributes["scrubbing"] = "data-scrubbing";
  /**
   * Present when the number field is disabled.
   */
  NumberFieldGroupDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the number field is readonly.
   */
  NumberFieldGroupDataAttributes["readonly"] = "data-readonly";
  /**
   * Present when the number field is required.
   */
  NumberFieldGroupDataAttributes["required"] = "data-required";
  /**
   * Present when the number field is in valid state (when wrapped in Field.Root).
   */
  NumberFieldGroupDataAttributes["valid"] = "data-valid";
  /**
   * Present when the number field is in invalid state (when wrapped in Field.Root).
   */
  NumberFieldGroupDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the number field has been touched (when wrapped in Field.Root).
   */
  NumberFieldGroupDataAttributes["touched"] = "data-touched";
  /**
   * Present when the number field's value has changed (when wrapped in Field.Root).
   */
  NumberFieldGroupDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the number field is filled (when wrapped in Field.Root).
   */
  NumberFieldGroupDataAttributes["filled"] = "data-filled";
  /**
   * Present when the number field is focused (when wrapped in Field.Root).
   */
  NumberFieldGroupDataAttributes["focused"] = "data-focused";
  return NumberFieldGroupDataAttributes;
}({});
