// @base-ui/react/esm/number-field/scrub-area/NumberFieldScrubAreaDataAttributes.js
export let NumberFieldScrubAreaDataAttributes = /*#__PURE__*/function (NumberFieldScrubAreaDataAttributes) {
  /**
   * Present while scrubbing.
   */
  NumberFieldScrubAreaDataAttributes["scrubbing"] = "data-scrubbing";
  /**
   * Present when the number field is disabled.
   */
  NumberFieldScrubAreaDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the number field is readonly.
   */
  NumberFieldScrubAreaDataAttributes["readonly"] = "data-readonly";
  /**
   * Present when the number field is required.
   */
  NumberFieldScrubAreaDataAttributes["required"] = "data-required";
  /**
   * Present when the number field is in valid state (when wrapped in Field.Root).
   */
  NumberFieldScrubAreaDataAttributes["valid"] = "data-valid";
  /**
   * Present when the number field is in invalid state (when wrapped in Field.Root).
   */
  NumberFieldScrubAreaDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the number field has been touched (when wrapped in Field.Root).
   */
  NumberFieldScrubAreaDataAttributes["touched"] = "data-touched";
  /**
   * Present when the number field's value has changed (when wrapped in Field.Root).
   */
  NumberFieldScrubAreaDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the number field is filled (when wrapped in Field.Root).
   */
  NumberFieldScrubAreaDataAttributes["filled"] = "data-filled";
  /**
   * Present when the number field is focused (when wrapped in Field.Root).
   */
  NumberFieldScrubAreaDataAttributes["focused"] = "data-focused";
  return NumberFieldScrubAreaDataAttributes;
}({});
