// @base-ui/react/drawer/swipe-area/DrawerSwipeAreaDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DrawerSwipeAreaDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let DrawerSwipeAreaDataAttributes = exports.DrawerSwipeAreaDataAttributes = function (DrawerSwipeAreaDataAttributes) {
  /**
   * Present when the drawer is open.
   */
  DrawerSwipeAreaDataAttributes[DrawerSwipeAreaDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the drawer is closed.
   */
  DrawerSwipeAreaDataAttributes[DrawerSwipeAreaDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the swipe area is disabled.
   */
  DrawerSwipeAreaDataAttributes["disabled"] = "data-disabled";
  /**
   * Indicates the swipe direction.
   * @type {'up' | 'down' | 'left' | 'right'}
   */
  DrawerSwipeAreaDataAttributes["swipeDirection"] = "data-swipe-direction";
  /**
   * Present when the drawer is being swiped.
   */
  DrawerSwipeAreaDataAttributes["swiping"] = "data-swiping";
  return DrawerSwipeAreaDataAttributes;
}({});
