// @base-ui/react/esm/select/popup/SelectPopupDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let SelectPopupDataAttributes = function (SelectPopupDataAttributes) {
  /**
   * Present when the select is open.
   */
  SelectPopupDataAttributes[SelectPopupDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the select is closed.
   */
  SelectPopupDataAttributes[SelectPopupDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the select is animating in.
   */
  SelectPopupDataAttributes[SelectPopupDataAttributes["startingStyle"] = CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the select is animating out.
   */
  SelectPopupDataAttributes[SelectPopupDataAttributes["endingStyle"] = CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'none' | 'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  SelectPopupDataAttributes[SelectPopupDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  SelectPopupDataAttributes[SelectPopupDataAttributes["align"] = CommonPopupDataAttributes.align] = "align";
  return SelectPopupDataAttributes;
}({});
