// @base-ui/react/esm/select/scroll-up-arrow/SelectScrollUpArrowDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let SelectScrollUpArrowDataAttributes = function (SelectScrollUpArrowDataAttributes) {
  /**
   * Present when the scroll arrow is animating in.
   */
  SelectScrollUpArrowDataAttributes[SelectScrollUpArrowDataAttributes["startingStyle"] = CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the scroll arrow is animating out.
   */
  SelectScrollUpArrowDataAttributes[SelectScrollUpArrowDataAttributes["endingStyle"] = CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates the direction of the scroll arrow.
   * @type {'up'}
   */
  SelectScrollUpArrowDataAttributes["direction"] = "data-direction";
  /**
   * Present when the scroll arrow is visible.
   */
  SelectScrollUpArrowDataAttributes["visible"] = "data-visible";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'none' | 'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  SelectScrollUpArrowDataAttributes[SelectScrollUpArrowDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  return SelectScrollUpArrowDataAttributes;
}({});
