// @base-ui/react/select/positioner/SelectPositionerDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectPositionerDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let SelectPositionerDataAttributes = exports.SelectPositionerDataAttributes = function (SelectPositionerDataAttributes) {
  /**
   * Present when the select popup is open.
   */
  SelectPositionerDataAttributes[SelectPositionerDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the select popup is closed.
   */
  SelectPositionerDataAttributes[SelectPositionerDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  SelectPositionerDataAttributes[SelectPositionerDataAttributes["anchorHidden"] = _popupStateMapping.CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'none' | 'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  SelectPositionerDataAttributes[SelectPositionerDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  SelectPositionerDataAttributes[SelectPositionerDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  return SelectPositionerDataAttributes;
}({});
