// @base-ui/react/esm/select/positioner/SelectPositionerDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let SelectPositionerDataAttributes = function (SelectPositionerDataAttributes) {
  /**
   * Present when the select popup is open.
   */
  SelectPositionerDataAttributes[SelectPositionerDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the select popup is closed.
   */
  SelectPositionerDataAttributes[SelectPositionerDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  SelectPositionerDataAttributes[SelectPositionerDataAttributes["anchorHidden"] = CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'none' | 'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  SelectPositionerDataAttributes[SelectPositionerDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  SelectPositionerDataAttributes[SelectPositionerDataAttributes["align"] = CommonPopupDataAttributes.align] = "align";
  return SelectPositionerDataAttributes;
}({});
