// @base-ui/react/esm/drawer/popup/DrawerPopupCssVars.js
export let DrawerPopupCssVars = /*#__PURE__*/function (DrawerPopupCssVars) {
  /**
   * The number of nested drawers that are currently open.
   * @type {number}
   */
  DrawerPopupCssVars["nestedDrawers"] = "--nested-drawers";
  /**
   * The height of the drawer popup.
   * @type {CSS length}
   */
  DrawerPopupCssVars["height"] = "--drawer-height";
  /**
   * The height of the frontmost open drawer in the current nested drawer stack.
   * @type {CSS length}
   */
  DrawerPopupCssVars["frontmostHeight"] = "--drawer-frontmost-height";
  /**
   * The swipe movement on the X axis.
   * @type {CSS length}
   */
  DrawerPopupCssVars["swipeMovementX"] = "--drawer-swipe-movement-x";
  /**
   * The swipe movement on the Y axis.
   * @type {CSS length}
   */
  DrawerPopupCssVars["swipeMovementY"] = "--drawer-swipe-movement-y";
  /**
   * The snap point offset used for translating the drawer.
   * @type {CSS length}
   */
  DrawerPopupCssVars["snapPointOffset"] = "--drawer-snap-point-offset";
  /**
   * A scalar (0.1-1) used to scale the swipe release transition duration in CSS.
   * @type {number}
   */
  DrawerPopupCssVars["swipeStrength"] = "--drawer-swipe-strength";
  return DrawerPopupCssVars;
}({});
