// @base-ui/react/alert-dialog/index.parts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Backdrop", {
  enumerable: true,
  get: function () {
    return _DialogBackdrop.DialogBackdrop;
  }
});
Object.defineProperty(exports, "Close", {
  enumerable: true,
  get: function () {
    return _DialogClose.DialogClose;
  }
});
Object.defineProperty(exports, "Description", {
  enumerable: true,
  get: function () {
    return _DialogDescription.DialogDescription;
  }
});
Object.defineProperty(exports, "Handle", {
  enumerable: true,
  get: function () {
    return _DialogHandle.DialogHandle;
  }
});
Object.defineProperty(exports, "Popup", {
  enumerable: true,
  get: function () {
    return _DialogPopup.DialogPopup;
  }
});
Object.defineProperty(exports, "Portal", {
  enumerable: true,
  get: function () {
    return _DialogPortal.DialogPortal;
  }
});
Object.defineProperty(exports, "Root", {
  enumerable: true,
  get: function () {
    return _AlertDialogRoot.AlertDialogRoot;
  }
});
Object.defineProperty(exports, "Title", {
  enumerable: true,
  get: function () {
    return _DialogTitle.DialogTitle;
  }
});
Object.defineProperty(exports, "Trigger", {
  enumerable: true,
  get: function () {
    return _DialogTrigger.DialogTrigger;
  }
});
Object.defineProperty(exports, "Viewport", {
  enumerable: true,
  get: function () {
    return _DialogViewport.DialogViewport;
  }
});
Object.defineProperty(exports, "createHandle", {
  enumerable: true,
  get: function () {
    return _handle.createAlertDialogHandle;
  }
});
var _AlertDialogRoot = require("./root/AlertDialogRoot");
var _DialogBackdrop = require("../dialog/backdrop/DialogBackdrop");
var _DialogClose = require("../dialog/close/DialogClose");
var _DialogDescription = require("../dialog/description/DialogDescription");
var _DialogPopup = require("../dialog/popup/DialogPopup");
var _DialogPortal = require("../dialog/portal/DialogPortal");
var _DialogTitle = require("../dialog/title/DialogTitle");
var _DialogTrigger = require("../dialog/trigger/DialogTrigger");
var _DialogViewport = require("../dialog/viewport/DialogViewport");
var _handle = require("./handle");
var _DialogHandle = require("../dialog/store/DialogHandle");
