// @base-ui/react/number-field/increment/NumberFieldIncrementDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NumberFieldIncrementDataAttributes = void 0;
let NumberFieldIncrementDataAttributes = exports.NumberFieldIncrementDataAttributes = /*#__PURE__*/function (NumberFieldIncrementDataAttributes) {
  /**
   * Present while scrubbing.
   */
  NumberFieldIncrementDataAttributes["scrubbing"] = "data-scrubbing";
  /**
   * Present when the number field is disabled.
   */
  NumberFieldIncrementDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the number field is readonly.
   */
  NumberFieldIncrementDataAttributes["readonly"] = "data-readonly";
  /**
   * Present when the number field is required.
   */
  NumberFieldIncrementDataAttributes["required"] = "data-required";
  /**
   * Present when the number field is in valid state (when wrapped in Field.Root).
   */
  NumberFieldIncrementDataAttributes["valid"] = "data-valid";
  /**
   * Present when the number field is in invalid state (when wrapped in Field.Root).
   */
  NumberFieldIncrementDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the number field has been touched (when wrapped in Field.Root).
   */
  NumberFieldIncrementDataAttributes["touched"] = "data-touched";
  /**
   * Present when the number field's value has changed (when wrapped in Field.Root).
   */
  NumberFieldIncrementDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the number field is filled (when wrapped in Field.Root).
   */
  NumberFieldIncrementDataAttributes["filled"] = "data-filled";
  /**
   * Present when the number field is focused (when wrapped in Field.Root).
   */
  NumberFieldIncrementDataAttributes["focused"] = "data-focused";
  return NumberFieldIncrementDataAttributes;
}({});
