// @base-ui/react/tooltip/positioner/TooltipPositionerCssVars.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TooltipPositionerCssVars = void 0;
let TooltipPositionerCssVars = exports.TooltipPositionerCssVars = /*#__PURE__*/function (TooltipPositionerCssVars) {
  /**
   * The available width between the trigger and the edge of the viewport.
   * @type {number}
   */
  TooltipPositionerCssVars["availableWidth"] = "--available-width";
  /**
   * The available height between the trigger and the edge of the viewport.
   * @type {number}
   */
  TooltipPositionerCssVars["availableHeight"] = "--available-height";
  /**
   * The anchor's width.
   * @type {number}
   */
  TooltipPositionerCssVars["anchorWidth"] = "--anchor-width";
  /**
   * The anchor's height.
   * @type {number}
   */
  TooltipPositionerCssVars["anchorHeight"] = "--anchor-height";
  /**
   * The coordinates that this element is anchored to. Used for animations and transitions.
   * @type {string}
   */
  TooltipPositionerCssVars["transformOrigin"] = "--transform-origin";
  return TooltipPositionerCssVars;
}({});
