// @base-ui/react/tooltip/positioner/TooltipPositionerDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TooltipPositionerDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let TooltipPositionerDataAttributes = exports.TooltipPositionerDataAttributes = function (TooltipPositionerDataAttributes) {
  /**
   * Present when the tooltip is open.
   */
  TooltipPositionerDataAttributes[TooltipPositionerDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the tooltip is closed.
   */
  TooltipPositionerDataAttributes[TooltipPositionerDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  TooltipPositionerDataAttributes[TooltipPositionerDataAttributes["anchorHidden"] = _popupStateMapping.CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  TooltipPositionerDataAttributes[TooltipPositionerDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  TooltipPositionerDataAttributes[TooltipPositionerDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  return TooltipPositionerDataAttributes;
}({});
