// @base-ui/react/number-field/decrement/NumberFieldDecrementDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NumberFieldDecrementDataAttributes = void 0;
let NumberFieldDecrementDataAttributes = exports.NumberFieldDecrementDataAttributes = /*#__PURE__*/function (NumberFieldDecrementDataAttributes) {
  /**
   * Present while scrubbing.
   */
  NumberFieldDecrementDataAttributes["scrubbing"] = "data-scrubbing";
  /**
   * Present when the number field is disabled.
   */
  NumberFieldDecrementDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the number field is readonly.
   */
  NumberFieldDecrementDataAttributes["readonly"] = "data-readonly";
  /**
   * Present when the number field is required.
   */
  NumberFieldDecrementDataAttributes["required"] = "data-required";
  /**
   * Present when the number field is in valid state (when wrapped in Field.Root).
   */
  NumberFieldDecrementDataAttributes["valid"] = "data-valid";
  /**
   * Present when the number field is in invalid state (when wrapped in Field.Root).
   */
  NumberFieldDecrementDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the number field has been touched (when wrapped in Field.Root).
   */
  NumberFieldDecrementDataAttributes["touched"] = "data-touched";
  /**
   * Present when the number field's value has changed (when wrapped in Field.Root).
   */
  NumberFieldDecrementDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the number field is filled (when wrapped in Field.Root).
   */
  NumberFieldDecrementDataAttributes["filled"] = "data-filled";
  /**
   * Present when the number field is focused (when wrapped in Field.Root).
   */
  NumberFieldDecrementDataAttributes["focused"] = "data-focused";
  return NumberFieldDecrementDataAttributes;
}({});
