// @base-ui/react/esm/accordion/panel/AccordionPanelDataAttributes.js
import { TransitionStatusDataAttributes } from "../../utils/stateAttributesMapping.js";
export let AccordionPanelDataAttributes = function (AccordionPanelDataAttributes) {
  /**
   * Indicates the index of the accordion item.
   * @type {number}
   */
  AccordionPanelDataAttributes["index"] = "data-index";
  /**
   * Present when the accordion panel is open.
   */
  AccordionPanelDataAttributes["open"] = "data-open";
  /**
   * Indicates the orientation of the accordion.
   */
  AccordionPanelDataAttributes["orientation"] = "data-orientation";
  /**
   * Present when the accordion item is disabled.
   */
  AccordionPanelDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the panel is animating in.
   */
  AccordionPanelDataAttributes[AccordionPanelDataAttributes["startingStyle"] = TransitionStatusDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the panel is animating out.
   */
  AccordionPanelDataAttributes[AccordionPanelDataAttributes["endingStyle"] = TransitionStatusDataAttributes.endingStyle] = "endingStyle";
  return AccordionPanelDataAttributes;
}({});
