// @base-ui/react/scroll-area/index.parts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Content", {
  enumerable: true,
  get: function () {
    return _ScrollAreaContent.ScrollAreaContent;
  }
});
Object.defineProperty(exports, "Corner", {
  enumerable: true,
  get: function () {
    return _ScrollAreaCorner.ScrollAreaCorner;
  }
});
Object.defineProperty(exports, "Root", {
  enumerable: true,
  get: function () {
    return _ScrollAreaRoot.ScrollAreaRoot;
  }
});
Object.defineProperty(exports, "Scrollbar", {
  enumerable: true,
  get: function () {
    return _ScrollAreaScrollbar.ScrollAreaScrollbar;
  }
});
Object.defineProperty(exports, "Thumb", {
  enumerable: true,
  get: function () {
    return _ScrollAreaThumb.ScrollAreaThumb;
  }
});
Object.defineProperty(exports, "Viewport", {
  enumerable: true,
  get: function () {
    return _ScrollAreaViewport.ScrollAreaViewport;
  }
});
var _ScrollAreaRoot = require("./root/ScrollAreaRoot");
var _ScrollAreaViewport = require("./viewport/ScrollAreaViewport");
var _ScrollAreaScrollbar = require("./scrollbar/ScrollAreaScrollbar");
var _ScrollAreaContent = require("./content/ScrollAreaContent");
var _ScrollAreaThumb = require("./thumb/ScrollAreaThumb");
var _ScrollAreaCorner = require("./corner/ScrollAreaCorner");
