// @base-ui/react/esm/tooltip/arrow/TooltipArrowDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let TooltipArrowDataAttributes = function (TooltipArrowDataAttributes) {
  /**
   * Present when the tooltip is open.
   */
  TooltipArrowDataAttributes[TooltipArrowDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the tooltip is closed.
   */
  TooltipArrowDataAttributes[TooltipArrowDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  TooltipArrowDataAttributes[TooltipArrowDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  TooltipArrowDataAttributes[TooltipArrowDataAttributes["align"] = CommonPopupDataAttributes.align] = "align";
  /**
   * Present when the tooltip arrow is uncentered.
   */
  TooltipArrowDataAttributes["uncentered"] = "data-uncentered";
  /**
   * Present if animations should be instant.
   * @type {'delay' | 'dismiss' | 'focus'}
   */
  TooltipArrowDataAttributes["instant"] = "data-instant";
  return TooltipArrowDataAttributes;
}({});
