// @base-ui/react/esm/field/error/FieldErrorDataAttributes.js
import { TransitionStatusDataAttributes } from "../../utils/stateAttributesMapping.js";
export let FieldErrorDataAttributes = function (FieldErrorDataAttributes) {
  /**
   * Present when the field is disabled.
   */
  FieldErrorDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the field is in valid state.
   */
  FieldErrorDataAttributes["valid"] = "data-valid";
  /**
   * Present when the field is in invalid state.
   */
  FieldErrorDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the field has been touched.
   */
  FieldErrorDataAttributes["touched"] = "data-touched";
  /**
   * Present when the field's value has changed.
   */
  FieldErrorDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the field is filled.
   */
  FieldErrorDataAttributes["filled"] = "data-filled";
  /**
   * Present when the field control is focused.
   */
  FieldErrorDataAttributes["focused"] = "data-focused";
  /**
   * Present when the error message is animating in.
   */
  FieldErrorDataAttributes[FieldErrorDataAttributes["startingStyle"] = TransitionStatusDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the error message is animating out.
   */
  FieldErrorDataAttributes[FieldErrorDataAttributes["endingStyle"] = TransitionStatusDataAttributes.endingStyle] = "endingStyle";
  return FieldErrorDataAttributes;
}({});
