// @base-ui/react/esm/menu/positioner/MenuPositionerDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let MenuPositionerDataAttributes = function (MenuPositionerDataAttributes) {
  /**
   * Present when the menu popup is open.
   */
  MenuPositionerDataAttributes[MenuPositionerDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the menu popup is closed.
   */
  MenuPositionerDataAttributes[MenuPositionerDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  MenuPositionerDataAttributes[MenuPositionerDataAttributes["anchorHidden"] = CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  MenuPositionerDataAttributes[MenuPositionerDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  MenuPositionerDataAttributes[MenuPositionerDataAttributes["align"] = CommonPopupDataAttributes.align] = "align";
  return MenuPositionerDataAttributes;
}({});
