// @base-ui/react/esm/preview-card/positioner/PreviewCardPositionerDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let PreviewCardPositionerDataAttributes = function (PreviewCardPositionerDataAttributes) {
  /**
   * Present when the preview card is open.
   */
  PreviewCardPositionerDataAttributes[PreviewCardPositionerDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the preview card is closed.
   */
  PreviewCardPositionerDataAttributes[PreviewCardPositionerDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  PreviewCardPositionerDataAttributes[PreviewCardPositionerDataAttributes["anchorHidden"] = CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  PreviewCardPositionerDataAttributes[PreviewCardPositionerDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  PreviewCardPositionerDataAttributes[PreviewCardPositionerDataAttributes["align"] = CommonPopupDataAttributes.align] = "align";
  return PreviewCardPositionerDataAttributes;
}({});
