// @base-ui/react/select/scroll-down-arrow/SelectScrollDownArrowDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectScrollDownArrowDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let SelectScrollDownArrowDataAttributes = exports.SelectScrollDownArrowDataAttributes = function (SelectScrollDownArrowDataAttributes) {
  /**
   * Present when the scroll arrow is animating in.
   */
  SelectScrollDownArrowDataAttributes[SelectScrollDownArrowDataAttributes["startingStyle"] = _popupStateMapping.CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the scroll arrow is animating out.
   */
  SelectScrollDownArrowDataAttributes[SelectScrollDownArrowDataAttributes["endingStyle"] = _popupStateMapping.CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates the direction of the scroll arrow.
   * @type {'down'}
   */
  SelectScrollDownArrowDataAttributes["direction"] = "data-direction";
  /**
   * Present when the scroll arrow is visible.
   */
  SelectScrollDownArrowDataAttributes["visible"] = "data-visible";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'none' | 'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  SelectScrollDownArrowDataAttributes[SelectScrollDownArrowDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  return SelectScrollDownArrowDataAttributes;
}({});
