// @base-ui/react/select/scroll-up-arrow/SelectScrollUpArrowDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectScrollUpArrowDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let SelectScrollUpArrowDataAttributes = exports.SelectScrollUpArrowDataAttributes = function (SelectScrollUpArrowDataAttributes) {
  /**
   * Present when the scroll arrow is animating in.
   */
  SelectScrollUpArrowDataAttributes[SelectScrollUpArrowDataAttributes["startingStyle"] = _popupStateMapping.CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the scroll arrow is animating out.
   */
  SelectScrollUpArrowDataAttributes[SelectScrollUpArrowDataAttributes["endingStyle"] = _popupStateMapping.CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates the direction of the scroll arrow.
   * @type {'up'}
   */
  SelectScrollUpArrowDataAttributes["direction"] = "data-direction";
  /**
   * Present when the scroll arrow is visible.
   */
  SelectScrollUpArrowDataAttributes["visible"] = "data-visible";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'none' | 'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  SelectScrollUpArrowDataAttributes[SelectScrollUpArrowDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  return SelectScrollUpArrowDataAttributes;
}({});
