// @base-ui/react/esm/autocomplete/index.parts.js
export { AutocompleteRoot as Root } from "./root/AutocompleteRoot.js";
export { AutocompleteValue as Value } from "./value/AutocompleteValue.js";
export { ComboboxTrigger as Trigger } from "../combobox/trigger/ComboboxTrigger.js";
export { ComboboxInput as Input } from "../combobox/input/ComboboxInput.js";
export { ComboboxInputGroup as InputGroup } from "../combobox/input-group/ComboboxInputGroup.js";
export { ComboboxIcon as Icon } from "../combobox/icon/ComboboxIcon.js";
export { ComboboxClear as Clear } from "../combobox/clear/ComboboxClear.js";
export { ComboboxList as List } from "../combobox/list/ComboboxList.js";
export { ComboboxStatus as Status } from "../combobox/status/ComboboxStatus.js";
export { ComboboxPortal as Portal } from "../combobox/portal/ComboboxPortal.js";
export { ComboboxBackdrop as Backdrop } from "../combobox/backdrop/ComboboxBackdrop.js";
export { ComboboxPositioner as Positioner } from "../combobox/positioner/ComboboxPositioner.js";
export { ComboboxPopup as Popup } from "../combobox/popup/ComboboxPopup.js";
export { ComboboxArrow as Arrow } from "../combobox/arrow/ComboboxArrow.js";
export { ComboboxGroup as Group } from "../combobox/group/ComboboxGroup.js";
export { ComboboxGroupLabel as GroupLabel } from "../combobox/group-label/ComboboxGroupLabel.js";
export { ComboboxItem as Item } from "../combobox/item/ComboboxItem.js";
export { ComboboxRow as Row } from "../combobox/row/ComboboxRow.js";
export { ComboboxCollection as Collection } from "../combobox/collection/ComboboxCollection.js";
export { ComboboxEmpty as Empty } from "../combobox/empty/ComboboxEmpty.js";
export { Separator } from "../separator/Separator.js";
export { useCoreFilter as useFilter } from "../combobox/root/utils/useFilter.js";
export { useFilteredItems } from "../combobox/root/utils/useFilteredItems.js";
