// @base-ui/react/navigation-menu/popup/NavigationMenuPopupDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NavigationMenuPopupDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let NavigationMenuPopupDataAttributes = exports.NavigationMenuPopupDataAttributes = function (NavigationMenuPopupDataAttributes) {
  /**
   * Present when the popup is open.
   */
  NavigationMenuPopupDataAttributes[NavigationMenuPopupDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  NavigationMenuPopupDataAttributes[NavigationMenuPopupDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the popup is animating in.
   */
  NavigationMenuPopupDataAttributes[NavigationMenuPopupDataAttributes["startingStyle"] = _popupStateMapping.CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the popup is animating out.
   */
  NavigationMenuPopupDataAttributes[NavigationMenuPopupDataAttributes["endingStyle"] = _popupStateMapping.CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  NavigationMenuPopupDataAttributes[NavigationMenuPopupDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to the specified side.
   * @type {'start' | 'center' | 'end'}
   */
  NavigationMenuPopupDataAttributes[NavigationMenuPopupDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  return NavigationMenuPopupDataAttributes;
}({});
