// @base-ui/react/esm/tooltip/positioner/TooltipPositionerDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let TooltipPositionerDataAttributes = function (TooltipPositionerDataAttributes) {
  /**
   * Present when the tooltip is open.
   */
  TooltipPositionerDataAttributes[TooltipPositionerDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the tooltip is closed.
   */
  TooltipPositionerDataAttributes[TooltipPositionerDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  TooltipPositionerDataAttributes[TooltipPositionerDataAttributes["anchorHidden"] = CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  TooltipPositionerDataAttributes[TooltipPositionerDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  TooltipPositionerDataAttributes[TooltipPositionerDataAttributes["align"] = CommonPopupDataAttributes.align] = "align";
  return TooltipPositionerDataAttributes;
}({});
