// @base-ui/react/esm/slider/thumb/SliderThumbDataAttributes.js
export let SliderThumbDataAttributes = /*#__PURE__*/function (SliderThumbDataAttributes) {
  /**
   * Indicates the index of the thumb in range sliders.
   */
  SliderThumbDataAttributes["index"] = "data-index";
  /**
   * Present while the user is dragging.
   */
  SliderThumbDataAttributes["dragging"] = "data-dragging";
  /**
   * Indicates the orientation of the slider.
   * @type {'horizontal' | 'vertical'}
   */
  SliderThumbDataAttributes["orientation"] = "data-orientation";
  /**
   * Present when the slider is disabled.
   */
  SliderThumbDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the slider is in valid state (when wrapped in Field.Root).
   */
  SliderThumbDataAttributes["valid"] = "data-valid";
  /**
   * Present when the slider is in invalid state (when wrapped in Field.Root).
   */
  SliderThumbDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the slider has been touched (when wrapped in Field.Root).
   */
  SliderThumbDataAttributes["touched"] = "data-touched";
  /**
   * Present when the slider's value has changed (when wrapped in Field.Root).
   */
  SliderThumbDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the slider is focused (when wrapped in Field.Root).
   */
  SliderThumbDataAttributes["focused"] = "data-focused";
  return SliderThumbDataAttributes;
}({});
