// @vis.gl/react-google-maps/dist/index.modern.mjs
import * as React from 'react';
import React__default, { useMemo, useState, useReducer, useCallback, useEffect, useRef as useRef$1, useContext, useLayoutEffect as useLayoutEffect$1, forwardRef, useImperativeHandle, Children, createContext } from 'react';
import { createPortal } from 'react-dom';
import isDeepEqual from 'fast-deep-equal';

// This file is automatically updated by the build process.
const VERSION = '1.8.3';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol, Iterator */


function __rest(s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
}

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

/*
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
function setScriptSrc(script, src) {
    script.src = src;
}

/*
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */


const bootstrap = bootstrapParams => {
  var bootstrapPromise;
  var script;
  var bootstrapParamsKey;
  var PRODUCT_NAME = "The Google Maps JavaScript API";
  var GOOGLE = "google";
  var IMPORT_API_NAME = "importLibrary";
  var PENDING_BOOTSTRAP_KEY = "__ib__";
  var doc = document;
  var global_ = window;
  var google_ = global_[GOOGLE] || (global_[GOOGLE] = {});
  var namespace = google_.maps || (google_.maps = {});
  var libraries = new Set();
  var searchParams = new URLSearchParams();
  var triggerBootstrap = () => bootstrapPromise || (bootstrapPromise = new Promise(async(resolve, reject) => {
    await (script = doc.createElement("script"));
    searchParams.set("libraries", [...libraries] + "");
    for (bootstrapParamsKey in bootstrapParams) {
      searchParams.set(bootstrapParamsKey.replace(/[A-Z]/g, g => "_" + g[0].toLowerCase()), bootstrapParams[bootstrapParamsKey]);
    }
    searchParams.set("callback", GOOGLE + ".maps." + PENDING_BOOTSTRAP_KEY);
    setScriptSrc(script, "https://maps.googleapis.com/maps/api/js?" + searchParams);
    namespace[PENDING_BOOTSTRAP_KEY] = resolve;
    script.onerror = () => bootstrapPromise = reject(Error(PRODUCT_NAME + " could not load."));
    script.nonce = doc.querySelector("script[nonce]")?.nonce || "";
    doc.head.append(script);
  }));
  namespace[IMPORT_API_NAME] ? console.warn(PRODUCT_NAME + " only loads once. Ignoring:", bootstrapParams) : namespace[IMPORT_API_NAME] = (libraryName, ...args) => libraries.add(libraryName) && triggerBootstrap().then(() => namespace[IMPORT_API_NAME](libraryName, ...args));
};
const MSG_REPEATED_SET_OPTIONS = (options) => `The setOptions() function should only be called once. The options passed ` +
    `to the additional call (${JSON.stringify(options)}) will be ignored.`;
const MSG_IMPORT_LIBRARY_EXISTS = (options) => `The google.maps.importLibrary() function is already defined, and ` +
    `@googlemaps/js-api-loader will use the existing function instead of ` +
    `overwriting it. The options passed to setOptions ` +
    `(${JSON.stringify(options)}) will be ignored.`;
const MSG_SET_OPTIONS_NOT_CALLED = "No options were set before calling importLibrary. Make sure to configure " +
    "the loader using setOptions().";
const MSG_SCRIPT_ELEMENT_EXISTS = "There already is a script loading the Google Maps JavaScript " +
    "API, and no google.maps.importLibrary function is defined. " +
    "@googlemaps/js-api-loader will proceed to bootstrap the API " +
    "with the specified options, but the existing script might cause " +
    "problems using the API. Make sure to remove the script " +
    "loading the API.";
const __DEV__$1 = process.env.NODE_ENV !== 'production';
const logDevWarning = __DEV__$1
    ? (message) => {
        console.warn(`[@googlemaps/js-api-loader] ${message}`);
    }
    : () => { };
const logDevNotice = __DEV__$1
    ? (message) => {
        console.info(`[@googlemaps/js-api-loader] ${message}`);
    }
    : () => { };

/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const __DEV__ = process.env.NODE_ENV !== 'production';
let setOptionsWasCalled_ = false;
/**
 * Sets the options for the Maps JavaScript API.
 *
 * Has to be called before any library is loaded.
 *
 * See https://developers.google.com/maps/documentation/javascript/load-maps-js-api#required_parameters
 * for the full documentation of available options.
 *
 * @param options The options to set.
 */
function setOptions(options) {
    if (setOptionsWasCalled_) {
        logDevWarning(MSG_REPEATED_SET_OPTIONS(options));
        return;
    }
    installImportLibrary_(options);
    setOptionsWasCalled_ = true;
}
async function importLibrary(libraryName) {
    if (!setOptionsWasCalled_) {
        logDevWarning(MSG_SET_OPTIONS_NOT_CALLED);
    }
    if (!window?.google?.maps?.importLibrary) {
        throw new Error("google.maps.importLibrary is not installed.");
    }
    return (await google.maps.importLibrary(libraryName));
}
/**
 * The installImportLibrary_ function makes sure that a usable version of the
 * `google.maps.importLibrary` function exists.
 */
function installImportLibrary_(options) {
    const importLibraryExists = Boolean(window.google?.maps?.importLibrary);
    if (importLibraryExists) {
        logDevNotice(MSG_IMPORT_LIBRARY_EXISTS(options));
    }
    else if (__DEV__) {
        const scriptEl = document.querySelector('script[src*="maps.googleapis.com/maps/api/js"]');
        if (scriptEl) {
            logDevWarning(MSG_SCRIPT_ELEMENT_EXISTS);
        }
    }
    // If the google.maps.importLibrary function already exists, bootstrap()
    // won't do anything, so we won't call it
    if (!importLibraryExists) {
        bootstrap(options);
    }
}

const APILoadingStatus = {
    NOT_LOADED: 'NOT_LOADED',
    LOADING: 'LOADING',
    LOADED: 'LOADED',
    FAILED: 'FAILED',
    AUTH_FAILURE: 'AUTH_FAILURE'
};

const DEFAULT_SOLUTION_CHANNEL = 'GMP_visgl_rgmlibrary_v1_default';
const DEFAULT_INTERNAL_USAGE_ATTRIBUTION_IDS = [
    `gmp_visgl_reactgooglemaps_v${VERSION}`
];
const APIProviderContext = React__default.createContext(null);
// loading the Maps JavaScript API can only happen once in the runtime, so these
// variables are kept at the module level.
let loadingStatus = APILoadingStatus.NOT_LOADED;
let serializedApiParams;
const listeners = new Set();
/**
 * Called to update the local status and notify the listeners for any mounted
 * components.
 * @internal
 */
function updateLoadingStatus(status) {
    if (status === loadingStatus) {
        return;
    }
    loadingStatus = status;
    listeners.forEach(listener => listener(loadingStatus));
}
/**
 * Local hook to set up the map-instance management context.
 * @internal
 */
function useMapInstances() {
    const [mapInstances, setMapInstances] = useState({});
    const addMapInstance = (mapInstance, id = 'default') => {
        setMapInstances(instances => (Object.assign(Object.assign({}, instances), { [id]: mapInstance })));
    };
    const removeMapInstance = (id = 'default') => {
        setMapInstances((_a) => {
            var _b = id; _a[_b]; var remaining = __rest(_a, [typeof _b === "symbol" ? _b : _b + ""]);
            return remaining;
        });
    };
    const clearMapInstances = () => {
        setMapInstances({});
    };
    return { mapInstances, addMapInstance, removeMapInstance, clearMapInstances };
}
/**
 * local hook to set up the 3D map-instance management context.
 */
function useMap3DInstances() {
    const [map3dInstances, setMap3DInstances] = useState({});
    const addMap3DInstance = (map3dInstance, id = 'default') => {
        setMap3DInstances(instances => (Object.assign(Object.assign({}, instances), { [id]: map3dInstance })));
    };
    const removeMap3DInstance = (id = 'default') => {
        setMap3DInstances((_a) => {
            var _b = id; _a[_b]; var remaining = __rest(_a, [typeof _b === "symbol" ? _b : _b + ""]);
            return remaining;
        });
    };
    const clearMap3DInstances = () => {
        setMap3DInstances({});
    };
    return {
        map3dInstances,
        addMap3DInstance,
        removeMap3DInstance,
        clearMap3DInstances
    };
}
/**
 * Local hook to handle the loading of the maps API.
 * @internal
 */
function useGoogleMapsApiLoader(props) {
    const { onLoad, onError, apiKey, version, libraries = [], region, language, authReferrerPolicy, channel, solutionChannel, fetchAppCheckToken } = props;
    const [status, setStatus] = useState(loadingStatus);
    const [loadedLibraries, addLoadedLibrary] = useReducer((loadedLibraries, action) => {
        return loadedLibraries[action.name]
            ? loadedLibraries
            : Object.assign(Object.assign({}, loadedLibraries), { [action.name]: action.value });
    }, {});
    const currentSerializedParams = useMemo(() => {
        const params = {
            apiKey,
            version,
            libraries: libraries.join(','),
            region,
            language,
            authReferrerPolicy,
            channel,
            solutionChannel
        };
        return JSON.stringify(params);
    }, [
        apiKey,
        version,
        libraries,
        region,
        language,
        authReferrerPolicy,
        channel,
        solutionChannel
    ]);
    const importLibraryCallback = useCallback((name) => __awaiter(this, void 0, void 0, function* () {
        if (loadedLibraries[name]) {
            return loadedLibraries[name];
        }
        const res = yield importLibrary(name);
        addLoadedLibrary({ name, value: res });
        return res;
    }), [loadedLibraries]);
    // effect: we want to get notified of global loading-status changes
    useEffect(() => {
        listeners.add(setStatus);
        // sync component state on mount (shouldn't be different from the initial state)
        setStatus(loadingStatus);
        return () => {
            listeners.delete(setStatus);
        };
    }, []);
    // effect: set and store options
    useEffect(() => {
        (() => __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            try {
                // This indicates that the API has been loaded with a different set of parameters.
                // While this is not blocking, it's not recommended and we should warn the user.
                if (serializedApiParams &&
                    serializedApiParams !== currentSerializedParams) {
                    console.warn(`The Google Maps JavaScript API has already been loaded with different parameters. ` +
                        `The new parameters will be ignored. If you need to use different parameters, ` +
                        `please refresh the page.`);
                }
                const librariesToLoad = ['core', 'maps', ...libraries];
                // If the google.maps namespace is already available, the API has been loaded externally.
                if ((_b = (_a = window.google) === null || _a === void 0 ? void 0 : _a.maps) === null || _b === void 0 ? void 0 : _b.importLibrary) {
                    if (!serializedApiParams) {
                        updateLoadingStatus(APILoadingStatus.LOADED);
                    }
                    yield Promise.all(librariesToLoad.map(name => importLibraryCallback(name)));
                    if (onLoad)
                        onLoad();
                    return;
                }
                // Abort if the API is already loading or has been loaded.
                if (loadingStatus === APILoadingStatus.LOADING ||
                    loadingStatus === APILoadingStatus.LOADED) {
                    if (loadingStatus === APILoadingStatus.LOADED && onLoad)
                        onLoad();
                    return;
                }
                serializedApiParams = currentSerializedParams;
                updateLoadingStatus(APILoadingStatus.LOADING);
                const options = Object.fromEntries(Object.entries({
                    key: apiKey,
                    v: version,
                    libraries,
                    region,
                    language,
                    authReferrerPolicy
                }).filter(([, value]) => value !== undefined));
                if (channel !== undefined && channel >= 0 && channel <= 999) {
                    options.channel = String(channel);
                }
                // solution-channel: when undefined, use the default; otherwise use
                // an explicit value.
                if (solutionChannel === undefined) {
                    options.solutionChannel = DEFAULT_SOLUTION_CHANNEL;
                }
                else if (solutionChannel !== '') {
                    options.solutionChannel = solutionChannel;
                }
                // this will actually trigger loading the maps API
                setOptions(options);
                // wait for all requested libraries (inluding 'core' and 'maps') to
                // finish loading
                yield Promise.all(librariesToLoad.map(name => importLibraryCallback(name)));
                updateLoadingStatus(APILoadingStatus.LOADED);
                if (onLoad) {
                    onLoad();
                }
            }
            catch (error) {
                updateLoadingStatus(APILoadingStatus.FAILED);
                if (onError) {
                    onError(error);
                }
                else {
                    console.error('The Google Maps JavaScript API failed to load.', error);
                }
            }
        }))();
    }, 
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [currentSerializedParams, onLoad, onError, importLibraryCallback, libraries]);
    // set the fetchAppCheckToken if provided
    useEffect(() => {
        if (status !== APILoadingStatus.LOADED)
            return;
        const settings = google.maps.Settings.getInstance();
        if (fetchAppCheckToken) {
            settings.fetchAppCheckToken = fetchAppCheckToken;
        }
        else if (settings.fetchAppCheckToken) {
            settings.fetchAppCheckToken = null;
        }
    }, [status, fetchAppCheckToken]);
    return {
        status,
        loadedLibraries,
        importLibrary: importLibraryCallback
    };
}
function useInternalUsageAttributionIds(props) {
    return useMemo(() => props.disableUsageAttribution
        ? null
        : DEFAULT_INTERNAL_USAGE_ATTRIBUTION_IDS, [props.disableUsageAttribution]);
}
/**
 * Component to wrap the components from this library and load the Google Maps JavaScript API
 */
const APIProvider = props => {
    const { children } = props, loaderProps = __rest(props, ["children"]);
    const { mapInstances, addMapInstance, removeMapInstance, clearMapInstances } = useMapInstances();
    const { map3dInstances, addMap3DInstance, removeMap3DInstance, clearMap3DInstances } = useMap3DInstances();
    const { status, loadedLibraries, importLibrary } = useGoogleMapsApiLoader(loaderProps);
    const internalUsageAttributionIds = useInternalUsageAttributionIds(loaderProps);
    const contextValue = useMemo(() => ({
        mapInstances,
        addMapInstance,
        removeMapInstance,
        clearMapInstances,
        map3dInstances,
        addMap3DInstance,
        removeMap3DInstance,
        clearMap3DInstances,
        status,
        loadedLibraries,
        importLibrary,
        internalUsageAttributionIds
    }), [
        mapInstances,
        addMapInstance,
        removeMapInstance,
        clearMapInstances,
        map3dInstances,
        addMap3DInstance,
        removeMap3DInstance,
        clearMap3DInstances,
        status,
        loadedLibraries,
        importLibrary,
        internalUsageAttributionIds
    ]);
    return (React__default.createElement(APIProviderContext.Provider, { value: contextValue }, children));
};
/**
 * @internal
 * Resets module-level state for testing purposes only.
 * This should never be used in production code.
 */
function __resetModuleState() {
    loadingStatus = APILoadingStatus.NOT_LOADED;
    serializedApiParams = undefined;
    listeners.clear();
}

/**
 * Sets up effects to bind event-handlers for all event-props in MapEventProps.
 * @internal
 */
function useMapEvents(map, props) {
    // note: calling a useEffect hook from within a loop is prohibited by the
    // rules of hooks, but it's ok here since it's unconditional and the number
    // and order of iterations is always strictly the same.
    // (see https://legacy.reactjs.org/docs/hooks-rules.html)
    for (const propName of eventPropNames) {
        // fixme: this cast is essentially a 'trust me, bro' for typescript, but
        //   a proper solution seems way too complicated right now
        const handler = props[propName];
        const eventType = propNameToEventType[propName];
        // eslint-disable-next-line react-hooks/rules-of-hooks
        useEffect(() => {
            if (!map)
                return;
            if (!handler)
                return;
            const listener = google.maps.event.addListener(map, eventType, (ev) => {
                handler(createMapEvent(eventType, map, ev));
            });
            return () => listener.remove();
        }, [map, eventType, handler]);
    }
}
/**
 * Create the wrapped map-events used for the event-props.
 * @param type the event type as it is specified to the maps api
 * @param map the map instance the event originates from
 * @param srcEvent the source-event if there is one.
 */
function createMapEvent(type, map, srcEvent) {
    var _a;
    const ev = {
        type,
        map,
        detail: {},
        stoppable: false,
        stop: () => { }
    };
    if (cameraEventTypes.includes(type)) {
        const camEvent = ev;
        const center = map.getCenter();
        const zoom = map.getZoom();
        const heading = map.getHeading() || 0;
        const tilt = map.getTilt() || 0;
        const bounds = map.getBounds();
        if (!center || !bounds || !Number.isFinite(zoom)) {
            console.warn('[createEvent] at least one of the values from the map ' +
                'returned undefined. This is not expected to happen. Please ' +
                'report an issue at https://github.com/visgl/react-google-maps/issues/new');
        }
        camEvent.detail = {
            center: (center === null || center === void 0 ? void 0 : center.toJSON()) || { lat: 0, lng: 0 },
            zoom: zoom || 0,
            heading: heading,
            tilt: tilt,
            bounds: (bounds === null || bounds === void 0 ? void 0 : bounds.toJSON()) || {
                north: 90,
                east: 180,
                south: -90,
                west: -180
            }
        };
        return camEvent;
    }
    else if (mouseEventTypes.includes(type)) {
        if (!srcEvent)
            throw new Error('[createEvent] mouse events must provide a srcEvent');
        const mouseEvent = ev;
        mouseEvent.domEvent = srcEvent.domEvent;
        mouseEvent.stoppable = true;
        mouseEvent.stop = () => srcEvent.stop();
        mouseEvent.detail = {
            latLng: ((_a = srcEvent.latLng) === null || _a === void 0 ? void 0 : _a.toJSON()) || null,
            placeId: srcEvent.placeId
        };
        return mouseEvent;
    }
    return ev;
}
/**
 * maps the camelCased names of event-props to the corresponding event-types
 * used in the maps API.
 */
const propNameToEventType = {
    onBoundsChanged: 'bounds_changed',
    onCenterChanged: 'center_changed',
    onClick: 'click',
    onContextmenu: 'contextmenu',
    onDblclick: 'dblclick',
    onDrag: 'drag',
    onDragend: 'dragend',
    onDragstart: 'dragstart',
    onHeadingChanged: 'heading_changed',
    onIdle: 'idle',
    onIsFractionalZoomEnabledChanged: 'isfractionalzoomenabled_changed',
    onMapCapabilitiesChanged: 'mapcapabilities_changed',
    onMapTypeIdChanged: 'maptypeid_changed',
    onMousemove: 'mousemove',
    onMouseout: 'mouseout',
    onMouseover: 'mouseover',
    onProjectionChanged: 'projection_changed',
    onRenderingTypeChanged: 'renderingtype_changed',
    onTilesLoaded: 'tilesloaded',
    onTiltChanged: 'tilt_changed',
    onZoomChanged: 'zoom_changed',
    // note: onCameraChanged is an alias for the bounds_changed event,
    // since that is going to be fired in every situation where the camera is
    // updated.
    onCameraChanged: 'bounds_changed'
};
const cameraEventTypes = [
    'bounds_changed',
    'center_changed',
    'heading_changed',
    'tilt_changed',
    'zoom_changed'
];
const mouseEventTypes = [
    'click',
    'contextmenu',
    'dblclick',
    'mousemove',
    'mouseout',
    'mouseover'
];
const eventPropNames = Object.keys(propNameToEventType);

/* eslint-disable react-hooks/refs */
// refs should not be used in render because changes to refs won't
// trigger a re-render, making them unreliable for holding state.
// In this case though, that is exactly what we want.
function useMemoized(value, isEqual) {
    const ref = useRef$1(value);
    if (!isEqual(value, ref.current)) {
        ref.current = value;
    }
    return ref.current;
}

function useCustomCompareEffect(effect, dependencies, isEqual) {
    // eslint-disable-next-line react-hooks/exhaustive-deps
    useEffect(effect, [useMemoized(dependencies, isEqual)]);
}

function useDeepCompareEffect(effect, dependencies) {
    useCustomCompareEffect(effect, dependencies, isDeepEqual);
}

const mapOptionKeys = new Set([
    'backgroundColor',
    'clickableIcons',
    'controlSize',
    'disableDefaultUI',
    'disableDoubleClickZoom',
    'draggable',
    'draggableCursor',
    'draggingCursor',
    'fullscreenControl',
    'fullscreenControlOptions',
    'gestureHandling',
    'headingInteractionEnabled',
    'isFractionalZoomEnabled',
    'keyboardShortcuts',
    'mapTypeControl',
    'mapTypeControlOptions',
    'mapTypeId',
    'maxZoom',
    'minZoom',
    'noClear',
    'panControl',
    'panControlOptions',
    'restriction',
    'rotateControl',
    'rotateControlOptions',
    'scaleControl',
    'scaleControlOptions',
    'scrollwheel',
    'streetView',
    'streetViewControl',
    'streetViewControlOptions',
    'styles',
    'tiltInteractionEnabled',
    'zoomControl',
    'zoomControlOptions'
]);
/**
 * Internal hook to update the map-options when props are changed.
 *
 * @param map the map instance
 * @param mapProps the props to update the map-instance with
 * @internal
 */
function useMapOptions(map, mapProps) {
    /* eslint-disable react-hooks/exhaustive-deps --
     *
     * The following effects aren't triggered when the map is changed.
     * In that case, the values will be or have been passed to the map
     * constructor via mapOptions.
     */
    const mapOptions = {};
    const keys = Object.keys(mapProps);
    for (const key of keys) {
        if (!mapOptionKeys.has(key))
            continue;
        mapOptions[key] = mapProps[key];
    }
    // update the map options when mapOptions is changed
    // Note: due to the destructuring above, mapOptions will be seen as changed
    //   with every re-render, so we're assuming the maps-api will properly
    //   deal with unchanged option-values passed into setOptions.
    useDeepCompareEffect(() => {
        if (!map)
            return;
        map.setOptions(mapOptions);
    }, [mapOptions]);
    /* eslint-enable react-hooks/exhaustive-deps */
}

function useApiLoadingStatus() {
    var _a;
    return ((_a = useContext(APIProviderContext)) === null || _a === void 0 ? void 0 : _a.status) || APILoadingStatus.NOT_LOADED;
}

/**
 * Internal hook that updates the camera when deck.gl viewState changes.
 * @internal
 */
function useDeckGLCameraUpdate(map, props) {
    const { viewport, viewState } = props;
    const isDeckGlControlled = !!viewport;
    useLayoutEffect$1(() => {
        if (!map || !viewState)
            return;
        const { latitude, longitude, bearing: heading, pitch: tilt, zoom } = viewState;
        map.moveCamera({
            center: { lat: latitude, lng: longitude },
            heading,
            tilt,
            zoom: zoom + 1
        });
    }, [map, viewState]);
    return isDeckGlControlled;
}

function isLatLngLiteral(obj) {
    if (!obj || typeof obj !== 'object')
        return false;
    if (!('lat' in obj && 'lng' in obj))
        return false;
    return Number.isFinite(obj.lat) && Number.isFinite(obj.lng);
}
function latLngEquals(a, b) {
    if (!a || !b)
        return false;
    const A = toLatLngLiteral(a);
    const B = toLatLngLiteral(b);
    if (A.lat !== B.lat || A.lng !== B.lng)
        return false;
    return true;
}
function toLatLngLiteral(obj) {
    if (isLatLngLiteral(obj))
        return obj;
    return obj.toJSON();
}
function toLatLngBoundsLiteral(obj) {
    if ('north' in obj && 'south' in obj && 'east' in obj && 'west' in obj) {
        return obj;
    }
    const ne = obj.getNorthEast().toJSON();
    const sw = obj.getSouthWest().toJSON();
    return {
        north: ne.lat,
        east: ne.lng,
        south: sw.lat,
        west: sw.lng
    };
}
function boundsEquals(a, b) {
    if (!a || !b)
        return false;
    const A = toLatLngBoundsLiteral(a);
    const B = toLatLngBoundsLiteral(b);
    return (A.north === B.north &&
        A.south === B.south &&
        A.east === B.east &&
        A.west === B.west);
}
/**
 * Compares two paths (arrays of LatLng points) for equality.
 */
function pathEquals(a, b) {
    if (!a || !b)
        return a === b;
    const arrayB = 'getArray' in b ? b.getArray() : b;
    if (a.length !== arrayB.length)
        return false;
    for (let i = 0; i < a.length; i++) {
        if (!latLngEquals(a[i], arrayB[i]))
            return false;
    }
    return true;
}
/**
 * Compares two arrays of paths (for Polygon) for equality.
 */
function pathsEquals(a, b) {
    if (!a || !b)
        return a === b;
    const arrayB = 'getArray' in b ? b.getArray().map(inner => inner.getArray()) : b;
    if (a.length !== arrayB.length)
        return false;
    for (let i = 0; i < a.length; i++) {
        if (!pathEquals(a[i], arrayB[i]))
            return false;
    }
    return true;
}

function useMapCameraParams(map, cameraStateRef, mapProps) {
    const center = mapProps.center ? toLatLngLiteral(mapProps.center) : null;
    let lat = null;
    let lng = null;
    if (center && Number.isFinite(center.lat) && Number.isFinite(center.lng)) {
        lat = center.lat;
        lng = center.lng;
    }
    const zoom = Number.isFinite(mapProps.zoom)
        ? mapProps.zoom
        : null;
    const heading = Number.isFinite(mapProps.heading)
        ? mapProps.heading
        : null;
    const tilt = Number.isFinite(mapProps.tilt)
        ? mapProps.tilt
        : null;
    // the following effect runs for every render of the map component and checks
    // if there are differences between the known state of the map instance
    // (cameraStateRef, which is updated by all bounds_changed events) and the
    // desired state in the props.
    useLayoutEffect$1(() => {
        if (!map)
            return;
        const nextCamera = {};
        let needsUpdate = false;
        if (lat !== null &&
            lng !== null &&
            (cameraStateRef.current.center.lat !== lat ||
                cameraStateRef.current.center.lng !== lng)) {
            nextCamera.center = { lat, lng };
            needsUpdate = true;
        }
        if (zoom !== null && cameraStateRef.current.zoom !== zoom) {
            nextCamera.zoom = zoom;
            needsUpdate = true;
        }
        if (heading !== null && cameraStateRef.current.heading !== heading) {
            nextCamera.heading = heading;
            needsUpdate = true;
        }
        if (tilt !== null && cameraStateRef.current.tilt !== tilt) {
            nextCamera.tilt = tilt;
            needsUpdate = true;
        }
        if (needsUpdate) {
            map.moveCamera(nextCamera);
        }
    });
}

const AuthFailureMessage = () => {
    const style = {
        position: 'absolute',
        top: 0,
        left: 0,
        bottom: 0,
        right: 0,
        zIndex: 999,
        display: 'flex',
        flexFlow: 'column nowrap',
        textAlign: 'center',
        justifyContent: 'center',
        fontSize: '.8rem',
        color: 'rgba(0,0,0,0.6)',
        background: '#dddddd',
        padding: '1rem 1.5rem'
    };
    return (React__default.createElement("div", { style: style },
        React__default.createElement("h2", null, "Error: AuthFailure"),
        React__default.createElement("p", null,
            "A problem with your API key prevents the map from rendering correctly. Please make sure the value of the ",
            React__default.createElement("code", null, "APIProvider.apiKey"),
            " prop is correct. Check the error-message in the console for further details.")));
};

function useCallbackRef() {
    const [el, setEl] = useState(null);
    const ref = useCallback((value) => setEl(value), [setEl]);
    return [el, ref];
}

/**
 * Hook to check if the Maps JavaScript API is loaded
 */
function useApiIsLoaded() {
    const status = useApiLoadingStatus();
    return status === APILoadingStatus.LOADED;
}

function useForceUpdate() {
    const [, forceUpdate] = useReducer(x => x + 1, 0);
    return forceUpdate;
}

function handleBoundsChange(map, ref) {
    const center = map.getCenter();
    const zoom = map.getZoom();
    const heading = map.getHeading() || 0;
    const tilt = map.getTilt() || 0;
    const bounds = map.getBounds();
    if (!center || !bounds || !Number.isFinite(zoom)) {
        console.warn('[useTrackedCameraState] at least one of the values from the map ' +
            'returned undefined. This is not expected to happen. Please ' +
            'report an issue at https://github.com/visgl/react-google-maps/issues/new');
    }
    // fixme: do we need the `undefined` cases for the camera-params? When are they used in the maps API?
    Object.assign(ref.current, {
        center: (center === null || center === void 0 ? void 0 : center.toJSON()) || { lat: 0, lng: 0 },
        zoom: zoom || 0,
        heading: heading,
        tilt: tilt
    });
}
/**
 * Creates a mutable ref object to track the last known state of the map camera.
 * This is used in `useMapCameraParams` to reduce stuttering in normal operation
 * by avoiding updates of the map camera with values that have already been processed.
 */
function useTrackedCameraStateRef(map) {
    const forceUpdate = useForceUpdate();
    const ref = useRef$1({
        center: { lat: 0, lng: 0 },
        heading: 0,
        tilt: 0,
        zoom: 0
    });
    // Record camera state with every bounds_changed event dispatched by the map.
    // This data is used to prevent feeding these values back to the
    // map-instance when a typical "controlled component" setup (state variable is
    // fed into and updated by the map).
    useEffect(() => {
        if (!map)
            return;
        const listener = google.maps.event.addListener(map, 'bounds_changed', () => {
            handleBoundsChange(map, ref);
            // When an event is occured, we have to update during the next cycle.
            // The application could decide to ignore the event and not update any
            // camera props of the map, meaning that in that case we will have to
            // 'undo' the change to the camera.
            forceUpdate();
        });
        return () => listener.remove();
    }, [map, forceUpdate]);
    return ref;
}

/**
 * Stores a stack of map-instances for each mapId. Whenever an
 * instance is used, it is removed from the stack while in use,
 * and returned to the stack when the component unmounts.
 * This allows us to correctly implement caching for multiple
 * maps om the same page, while reusing as much as possible.
 *
 * FIXME: while it should in theory be possible to reuse maps solely
 *   based on mapId/renderingType/colorScheme (as all other parameters can be
 *   changed at runtime), we don't yet have good enough tracking of options to
 *   reliably unset all the options that have been set.
 */
class CachedMapStack {
    static has(key) {
        return this.entries[key] && this.entries[key].length > 0;
    }
    static pop(key) {
        if (!this.entries[key])
            return null;
        return this.entries[key].pop() || null;
    }
    static push(key, value) {
        if (!this.entries[key])
            this.entries[key] = [];
        this.entries[key].push(value);
    }
}
CachedMapStack.entries = {};
/**
 * The main hook takes care of creating map-instances and registering them in
 * the api-provider context.
 * @return a tuple of the map-instance created (or null) and the callback
 *   ref that will be used to pass the map-container into this hook.
 * @internal
 */
function useMapInstance(props, context) {
    const apiIsLoaded = useApiIsLoaded();
    const [map, setMap] = useState(null);
    const [container, containerRef] = useCallbackRef();
    const cameraStateRef = useTrackedCameraStateRef(map);
    const { id, defaultBounds, defaultCenter, defaultZoom, defaultHeading, defaultTilt, reuseMaps, renderingType, colorScheme } = props, mapOptions = __rest(props, ["id", "defaultBounds", "defaultCenter", "defaultZoom", "defaultHeading", "defaultTilt", "reuseMaps", "renderingType", "colorScheme"]);
    const hasZoom = props.zoom !== undefined || props.defaultZoom !== undefined;
    const hasCenter = props.center !== undefined || props.defaultCenter !== undefined;
    if (!defaultBounds && (!hasZoom || !hasCenter)) {
        console.warn('<Map> component is missing configuration. ' +
            'You have to provide zoom and center (via the `zoom`/`defaultZoom` and ' +
            '`center`/`defaultCenter` props) or specify the region to show using ' +
            '`defaultBounds`. See ' +
            'https://visgl.github.io/react-google-maps/docs/api-reference/components/map#required');
    }
    // apply default camera props if available and not overwritten by controlled props
    if (!mapOptions.center && defaultCenter)
        mapOptions.center = defaultCenter;
    if (!mapOptions.zoom && Number.isFinite(defaultZoom))
        mapOptions.zoom = defaultZoom;
    if (!mapOptions.heading && Number.isFinite(defaultHeading))
        mapOptions.heading = defaultHeading;
    if (!mapOptions.tilt && Number.isFinite(defaultTilt))
        mapOptions.tilt = defaultTilt;
    // Handle internalUsageAttributionIds
    const customIds = mapOptions.internalUsageAttributionIds;
    if (customIds == null) {
        // Not specified - use context default (which may be null if disabled)
        mapOptions.internalUsageAttributionIds =
            context.internalUsageAttributionIds;
    }
    else {
        // Merge context defaults with custom IDs
        mapOptions.internalUsageAttributionIds = [
            ...(context.internalUsageAttributionIds || []),
            ...customIds
        ];
    }
    for (const key of Object.keys(mapOptions))
        if (mapOptions[key] === undefined)
            delete mapOptions[key];
    const savedMapStateRef = useRef$1(undefined);
    // create the map instance and register it in the context
    useEffect(() => {
        if (!container || !apiIsLoaded)
            return;
        const { addMapInstance, removeMapInstance } = context;
        // note: colorScheme (upcoming feature) isn't yet in the typings, remove once that is fixed:
        const { mapId } = props;
        const cacheKey = `${mapId || 'default'}:${renderingType || 'default'}:${colorScheme || 'LIGHT'}`;
        let mapDiv;
        let map;
        if (reuseMaps && CachedMapStack.has(cacheKey)) {
            map = CachedMapStack.pop(cacheKey);
            mapDiv = map.getDiv();
            container.appendChild(mapDiv);
            map.setOptions(mapOptions);
            // detaching the element from the DOM sometimes causes the map to collapse
            // and no longer render tiles that should be in view after re-attaching it.
            // Triggering moveCamera after remounting should trigger a re-layout of
            // the map.
            setTimeout(() => map.moveCamera({}), 0);
        }
        else {
            mapDiv = document.createElement('div');
            mapDiv.style.height = '100%';
            container.appendChild(mapDiv);
            map = new google.maps.Map(mapDiv, Object.assign(Object.assign(Object.assign({}, mapOptions), (renderingType
                ? { renderingType: renderingType }
                : {})), (colorScheme
                ? { colorScheme: colorScheme }
                : {})));
        }
        setMap(map);
        addMapInstance(map, id);
        if (defaultBounds) {
            const { padding } = defaultBounds, defBounds = __rest(defaultBounds, ["padding"]);
            map.fitBounds(defBounds, padding);
        }
        // prevent map not rendering due to missing configuration
        else if (!hasZoom || !hasCenter) {
            map.fitBounds({ east: 180, west: -180, south: -90, north: 90 });
        }
        // the savedMapState is used to restore the camera parameters when the mapId is changed
        if (savedMapStateRef.current) {
            const { mapId: savedMapId, cameraState: savedCameraState } = savedMapStateRef.current;
            if (savedMapId !== mapId) {
                map.moveCamera(savedCameraState);
            }
        }
        return () => {
            savedMapStateRef.current = {
                mapId,
                // eslint-disable-next-line react-hooks/exhaustive-deps
                cameraState: cameraStateRef.current
            };
            // detach the map-div from the dom
            mapDiv.remove();
            if (reuseMaps) {
                // push back on the stack
                CachedMapStack.push(cacheKey, map);
            }
            else {
                // remove all event-listeners to minimize the possibility of memory-leaks
                google.maps.event.clearInstanceListeners(map);
            }
            setMap(null);
            removeMapInstance(id);
        };
    }, 
    // some dependencies are ignored in the list below:
    //  - defaultBounds and the default* camera props will only be used once, and
    //    changes should be ignored
    //  - mapOptions has special hooks that take care of updating the options
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [
        container,
        apiIsLoaded,
        id,
        // these props can't be changed after initialization and require a new
        // instance to be created
        props.mapId,
        props.renderingType,
        props.colorScheme
    ]);
    return [map, containerRef, cameraStateRef];
}

const GoogleMapsContext = React__default.createContext(null);
// ColorScheme and RenderingType are redefined here to make them usable before the
// maps API has been fully loaded.
const ColorScheme = {
    DARK: 'DARK',
    LIGHT: 'LIGHT',
    FOLLOW_SYSTEM: 'FOLLOW_SYSTEM'
};
const RenderingType = {
    VECTOR: 'VECTOR',
    RASTER: 'RASTER',
    UNINITIALIZED: 'UNINITIALIZED'
};
const Map = (props) => {
    const { children, id, className, style } = props;
    const context = useContext(APIProviderContext);
    const loadingStatus = useApiLoadingStatus();
    if (!context) {
        throw new Error('<Map> can only be used inside an <ApiProvider> component.');
    }
    const [map, mapRef, cameraStateRef] = useMapInstance(props, context);
    useMapCameraParams(map, cameraStateRef, props);
    useMapEvents(map, props);
    useMapOptions(map, props);
    const isDeckGlControlled = useDeckGLCameraUpdate(map, props);
    const isControlledExternally = !!props.controlled;
    // disable interactions with the map for externally controlled maps
    useEffect(() => {
        if (!map)
            return;
        // fixme: this doesn't seem to belong here (and it's mostly there for convenience anyway).
        //   The reasoning is that a deck.gl canvas will be put on top of the map, rendering
        //   any default map controls pretty much useless
        if (isDeckGlControlled) {
            map.setOptions({ disableDefaultUI: true });
        }
        // disable all control-inputs when the map is controlled externally
        if (isDeckGlControlled || isControlledExternally) {
            map.setOptions({
                gestureHandling: 'none',
                keyboardShortcuts: false
            });
        }
        return () => {
            map.setOptions({
                gestureHandling: props.gestureHandling,
                keyboardShortcuts: props.keyboardShortcuts
            });
        };
    }, [
        map,
        isDeckGlControlled,
        isControlledExternally,
        props.gestureHandling,
        props.keyboardShortcuts
    ]);
    // setup a stable cameraOptions object that can be used as dependency
    const center = props.center ? toLatLngLiteral(props.center) : null;
    let lat = null;
    let lng = null;
    if (center && Number.isFinite(center.lat) && Number.isFinite(center.lng)) {
        lat = center.lat;
        lng = center.lng;
    }
    const cameraOptions = useMemo(() => {
        var _a, _b, _c;
        return {
            center: { lat: lat !== null && lat !== void 0 ? lat : 0, lng: lng !== null && lng !== void 0 ? lng : 0 },
            zoom: (_a = props.zoom) !== null && _a !== void 0 ? _a : 0,
            heading: (_b = props.heading) !== null && _b !== void 0 ? _b : 0,
            tilt: (_c = props.tilt) !== null && _c !== void 0 ? _c : 0
        };
    }, [lat, lng, props.zoom, props.heading, props.tilt]);
    // externally controlled mode: reject all camera changes that don't correspond to changes in props
    useLayoutEffect$1(() => {
        if (!map || !isControlledExternally)
            return;
        map.moveCamera(cameraOptions);
        const listener = map.addListener('bounds_changed', () => {
            map.moveCamera(cameraOptions);
        });
        return () => listener.remove();
    }, [map, isControlledExternally, cameraOptions]);
    const combinedStyle = useMemo(() => (Object.assign({ width: '100%', height: '100%', position: 'relative', 
        // when using deckgl, the map should be sent to the back
        zIndex: isDeckGlControlled ? -1 : 0 }, style)), [style, isDeckGlControlled]);
    const contextValue = useMemo(() => ({ map }), [map]);
    if (loadingStatus === APILoadingStatus.AUTH_FAILURE) {
        return (React__default.createElement("div", { style: Object.assign({ position: 'relative' }, (className ? {} : combinedStyle)), className: className },
            React__default.createElement(AuthFailureMessage, null)));
    }
    return (React__default.createElement("div", Object.assign({ ref: mapRef, "data-testid": 'map', style: className ? undefined : combinedStyle, className: className }, (id ? { id } : {})), map ? (React__default.createElement(GoogleMapsContext.Provider, { value: contextValue }, children)) : null));
};
// The deckGLViewProps flag here indicates to deck.gl that the Map component is
// able to handle viewProps from deck.gl when deck.gl is used to control the map.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
Map.deckGLViewProps = true;

const shownMessages = new Set();
function logErrorOnce(...args) {
    const key = JSON.stringify(args);
    if (!shownMessages.has(key)) {
        shownMessages.add(key);
        console.error(...args);
    }
}

/**
 * Retrieves a map-instance from the context. This is either an instance
 * identified by id or the parent map instance if no id is specified.
 * Returns null if neither can be found.
 */
const useMap = (id = null) => {
    const ctx = useContext(APIProviderContext);
    const { map } = useContext(GoogleMapsContext) || {};
    if (ctx === null) {
        logErrorOnce('useMap(): failed to retrieve APIProviderContext. ' +
            'Make sure that the <APIProvider> component exists and that the ' +
            'component you are calling `useMap()` from is a sibling of the ' +
            '<APIProvider>.');
        return null;
    }
    const { mapInstances } = ctx;
    // if an id is specified, the corresponding map or null is returned
    if (id !== null)
        return mapInstances[id] || null;
    // otherwise, return the closest ancestor
    if (map)
        return map;
    // finally, return the default map instance
    return mapInstances['default'] || null;
};

function useMapsLibrary(name) {
    const apiIsLoaded = useApiIsLoaded();
    const ctx = useContext(APIProviderContext);
    useEffect(() => {
        if (!apiIsLoaded || !ctx)
            return;
        // Trigger loading the libraries via our proxy-method.
        // The returned promise is ignored, since importLibrary will update loadedLibraries
        // list in the context, triggering a re-render.
        void ctx.importLibrary(name);
    }, [apiIsLoaded, ctx, name]);
    return (ctx === null || ctx === void 0 ? void 0 : ctx.loadedLibraries[name]) || null;
}

// Copyright (c) 2025 Sanity
// This file is derived from sanity-io/use-effect-event (MIT Licensed)
//
// https://github.com/sanity-io/use-effect-event
var _a;
const { useLayoutEffect, useRef } = React;
// useInsertionEffect was added in React 18; fall back to useLayoutEffect for
// React 16/17. Both run before useEffect, so ref.current is always up-to-date
// by the time any passive effect (or real event) reads it.
const useBeforeEffect = (_a = React.useInsertionEffect) !== null && _a !== void 0 ? _a : useLayoutEffect;
function forbiddenInRender() {
    throw new Error('useEffectEvent: invalid call during rendering.');
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function useEffectEventPolyfill(fn) {
    /**
     * Initialize the ref with `forbiddenInRender`, to catch illegal calls during
     * rendering. After the insertion effect ran, the ref will contain the actual
     * function, so all effects can see the actual value.
     */
    const ref = useRef(forbiddenInRender);
    useBeforeEffect(() => {
        ref.current = fn;
    }, [fn]);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return ((...args) => ref.current(...args));
}
/**
 * Uses a polyfill implementation of `useEffectEvent`. The native useEffectEvent
 * implementation was causing issues that we do not fully understand yet.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const useEffectEvent = useEffectEventPolyfill;

/* eslint-disable @typescript-eslint/no-explicit-any */
const noop$1 = () => { };
/**
 * Internally used to bind events to Maps JavaScript API objects.
 * @internal
 */
function useMapsEventListener(target, name, callback) {
    const eventFn = useEffectEvent(callback !== null && callback !== void 0 ? callback : noop$1);
    const isCallbackDefined = Boolean(callback);
    useEffect(() => {
        if (!target || !name || !isCallbackDefined)
            return;
        const listener = google.maps.event.addListener(target, name, eventFn);
        return () => listener.remove();
    }, [target, name, isCallbackDefined]);
}

/* eslint-disable react-hooks/immutability -- Google Maps API objects are designed to be mutated */
/**
 * Internally used to copy values from props into API-Objects
 * whenever they change.
 *
 * @example
 *   usePropBinding(marker, 'position', position);
 *
 * @internal
 */
function usePropBinding(object, prop, value) {
    useEffect(() => {
        if (!object)
            return;
        object[prop] = value;
    }, [object, prop, value]);
}

/* eslint-disable @typescript-eslint/no-explicit-any */
const noop = () => { };
/**
 * Internally used to bind events to DOM nodes.
 * @internal
 */
function useDomEventListener(target, name, callback) {
    const eventFn = useEffectEvent(callback !== null && callback !== void 0 ? callback : noop);
    const isCallbackDefined = Boolean(callback);
    useEffect(() => {
        if (!target || !name || !isCallbackDefined)
            return;
        // Note: eventFn is not guaranteed to be stable across renders, so we need
        // to use a local variable to be sure to remove the very same listener
        // function that has been added
        const listenerCallback = eventFn;
        target.addEventListener(name, listenerCallback);
        return () => target.removeEventListener(name, listenerCallback);
    }, [target, name, isCallbackDefined]);
}

// Global style manager to track rendered styles and avoid duplicates
class GlobalStyleManager {
    constructor() {
        this.renderedStyles = new Set();
        this.styleElement = null;
    }
    getStyleElement() {
        if (!this.styleElement) {
            this.styleElement = document.createElement('style');
            this.styleElement.setAttribute('data-rgm-anchor-styles', '');
            document.head.appendChild(this.styleElement);
        }
        return this.styleElement;
    }
    addAdvancedMarkerPointerEventsOverwrite() {
        if (this.renderedStyles.has('marker-pointer-events')) {
            return;
        }
        const styleElement = this.getStyleElement();
        styleElement.textContent += `
      gmp-advanced-marker[data-origin='rgm'] {
        pointer-events: none !important;
      }
    `;
        this.renderedStyles.add('marker-pointer-events');
    }
    cleanup() {
        if (this.styleElement) {
            this.styleElement.remove();
            this.styleElement = null;
            this.renderedStyles.clear();
        }
    }
}
const globalStyleManager = new GlobalStyleManager();

function isVersionGreaterEqual(major, minor) {
    var _a;
    if (!((_a = google === null || google === void 0 ? void 0 : google.maps) === null || _a === void 0 ? void 0 : _a.version))
        return undefined;
    const version = google.maps.version.split('.');
    const currentMajor = parseInt(version[0], 10);
    const currentMinor = parseInt(version[1], 10);
    return (currentMajor > major || (currentMajor === major && currentMinor >= minor));
}

/**
 * Copy of the `google.maps.CollisionBehavior` constants.
 * They have to be duplicated here since we can't wait for the maps API to load to be able to use them.
 */
const CollisionBehavior = {
    REQUIRED: 'REQUIRED',
    REQUIRED_AND_HIDES_OPTIONAL: 'REQUIRED_AND_HIDES_OPTIONAL',
    OPTIONAL_AND_HIDES_LOWER_PRIORITY: 'OPTIONAL_AND_HIDES_LOWER_PRIORITY'
};
const AdvancedMarkerContext = React__default.createContext(null);
// [xPosition, yPosition] when the top left corner is [0, 0]
/**
 * @deprecated Using `anchorPosition` is deprecated.
 *   Use `anchorLeft` and `anchorTop` instead.
 */
const AdvancedMarkerAnchorPoint = {
    TOP_LEFT: ['0%', '0%'],
    TOP_CENTER: ['50%', '0%'],
    TOP: ['50%', '0%'],
    TOP_RIGHT: ['100%', '0%'],
    LEFT_CENTER: ['0%', '50%'],
    LEFT_TOP: ['0%', '0%'],
    LEFT: ['0%', '50%'],
    LEFT_BOTTOM: ['0%', '100%'],
    RIGHT_TOP: ['100%', '0%'],
    RIGHT: ['100%', '50%'],
    RIGHT_CENTER: ['100%', '50%'],
    RIGHT_BOTTOM: ['100%', '100%'],
    BOTTOM_LEFT: ['0%', '100%'],
    BOTTOM_CENTER: ['50%', '100%'],
    BOTTOM: ['50%', '100%'],
    BOTTOM_RIGHT: ['100%', '100%'],
    CENTER: ['50%', '50%']
};
const AdvancedMarker = forwardRef((props, ref) => {
    const { children, style, className, anchorPoint } = props;
    const [marker, contentContainer] = useAdvancedMarker(props);
    const advancedMarkerContextValue = useMemo(() => (marker ? { marker } : null), [marker]);
    useImperativeHandle(ref, () => marker, [marker]);
    if (!contentContainer)
        return null;
    return (React__default.createElement(AdvancedMarkerContext.Provider, { value: advancedMarkerContextValue }, createPortal(React__default.createElement(MarkerContent, { anchorPoint: anchorPoint, styles: style, className: className }, children), contentContainer)));
});
AdvancedMarker.displayName = 'AdvancedMarker';
function useAdvancedMarkerRef() {
    const [marker, setMarker] = useState(null);
    const refCallback = useCallback((m) => {
        setMarker(m);
    }, []);
    return [refCallback, marker];
}
function isAdvancedMarker(marker) {
    return (marker.content !== undefined);
}
function isElementNode(node) {
    return node.nodeType === Node.ELEMENT_NODE;
}
const MarkerContent = ({ children, styles, className }) => {
    /* AdvancedMarker div that user can give styles and classes */
    return (React__default.createElement("div", { className: className, style: styles }, children));
};
function useAdvancedMarker(props) {
    const [marker, setMarker] = useState(null);
    const [contentContainer, setContentContainer] = useState(null);
    const map = useMap();
    const markerLibrary = useMapsLibrary('marker');
    const { children, onClick, className, onMouseEnter, onMouseLeave, onDrag, onDragStart, onDragEnd, collisionBehavior, clickable, draggable, position, title, zIndex, anchorPoint, anchorLeft, anchorTop } = props;
    const numChildren = Children.count(children);
    // create an AdvancedMarkerElement instance and add it to the map once available
    useEffect(() => {
        if (!map || !markerLibrary)
            return;
        const newMarker = new markerLibrary.AdvancedMarkerElement();
        newMarker.map = map;
        setMarker(newMarker);
        // create the container for marker content if there are children
        let contentElement = null;
        if (numChildren > 0) {
            contentElement = document.createElement('div');
            newMarker.content = contentElement;
            setContentContainer(contentElement);
        }
        return () => {
            newMarker.map = null;
            contentElement === null || contentElement === void 0 ? void 0 : contentElement.remove();
            setMarker(null);
            setContentContainer(null);
        };
    }, [map, markerLibrary, numChildren]);
    // When no children are present we don't have our own wrapper div
    // which usually gets the user provided className. In this case
    // we set the className directly on the marker.content element that comes
    // with the AdvancedMarker.
    useEffect(() => {
        if (!(marker === null || marker === void 0 ? void 0 : marker.content) || !isElementNode(marker.content) || numChildren > 0)
            return;
        marker.content.className = className !== null && className !== void 0 ? className : '';
    }, [marker, className, numChildren]);
    useAdvancedMarkerAnchoring(marker, anchorPoint, anchorLeft, anchorTop, numChildren > 0);
    // copy other props
    usePropBinding(marker, 'position', position);
    usePropBinding(marker, 'title', title !== null && title !== void 0 ? title : '');
    usePropBinding(marker, 'zIndex', zIndex);
    usePropBinding(marker, 'collisionBehavior', collisionBehavior);
    // set gmpDraggable from props (when unspecified, it's true if any drag-event
    // callbacks are specified)
    useEffect(() => {
        if (!marker)
            return;
        if (draggable !== undefined)
            marker.gmpDraggable = draggable;
        else if (onDrag || onDragStart || onDragEnd)
            marker.gmpDraggable = true;
        else
            marker.gmpDraggable = false;
    }, [marker, draggable, onDrag, onDragEnd, onDragStart]);
    // set gmpClickable from props (when unspecified, it's true if the onClick or one of
    // the hover events callbacks are specified)
    useEffect(() => {
        if (!marker)
            return;
        // when clickable is defined, we will always use its value for gmpClickable.
        // otherwise we auto-detect based on existing event-handlers.
        const gmpClickable = clickable !== undefined
            ? clickable
            : Boolean(onClick) || Boolean(onMouseEnter) || Boolean(onMouseLeave);
        // gmpClickable is only available in beta version of the
        // maps api (as of 2024-10-10)
        marker.gmpClickable = gmpClickable;
        // enable pointer events for the markers with custom content
        if (gmpClickable && (marker === null || marker === void 0 ? void 0 : marker.content) && isElementNode(marker.content)) {
            marker.content.style.pointerEvents = 'all';
            if (onClick) {
                marker.content.style.cursor = 'pointer';
            }
        }
    }, [marker, clickable, onClick, onMouseEnter, onMouseLeave]);
    useMapsEventListener(marker, 'click', onClick);
    useMapsEventListener(marker, 'drag', onDrag);
    useMapsEventListener(marker, 'dragstart', onDragStart);
    useMapsEventListener(marker, 'dragend', onDragEnd);
    useDomEventListener(marker === null || marker === void 0 ? void 0 : marker.element, 'mouseenter', onMouseEnter);
    useDomEventListener(marker === null || marker === void 0 ? void 0 : marker.element, 'mouseleave', onMouseLeave);
    return [marker, contentContainer];
}
function useAdvancedMarkerAnchoring(marker, anchorPoint, anchorLeft, anchorTop, hasChildren) {
    useEffect(() => {
        if (!marker || !hasChildren)
            return;
        // The anchorLeft and anchorTop options are available since version 3.62.9c
        // With the release of 3.65 (~May 2026) there will no longer be a version
        // that doesn't support it.
        const anchorOptionsSupported = isVersionGreaterEqual(3, 62);
        const contentElement = marker.content;
        if (!contentElement || !isElementNode(contentElement))
            return;
        if (anchorLeft !== undefined || anchorTop !== undefined) {
            if (!anchorOptionsSupported) {
                console.warn('AdvancedMarker: The anchorLeft and anchorTop props are only supported ' +
                    'in Google Maps API version 3.62 and above. ' +
                    `The current version is ${google.maps.version}.`);
            }
            marker.anchorLeft = anchorLeft;
            marker.anchorTop = anchorTop;
            // when anchorLeft and/or anchorTop are set, we'll ignore the anchorPoint
            if (anchorPoint !== undefined) {
                console.warn('AdvancedMarker: the anchorPoint prop is ignored when anchorLeft ' +
                    'and/or anchorTop are set.');
            }
            return;
        }
        if (anchorPoint !== undefined) {
            // TODO: add console.warn in a future version to inform about deprecation
            const [x, y] = anchorPoint !== null && anchorPoint !== void 0 ? anchorPoint : AdvancedMarkerAnchorPoint['BOTTOM'];
            // NOTE: since x and y can be any valid CSS length-percentage
            //   value, we need to use calc() to negate them.
            const translateX = `calc(-1 * ${x})`;
            const translateY = `calc(-1 * ${y})`;
            if (anchorOptionsSupported) {
                // implement anchorPoint using the new anchorLeft and anchorTop options
                marker.anchorLeft = translateX;
                marker.anchorTop = translateY;
                // reset transform from legacy implementation
                contentElement.style.transform = '';
            }
            else {
                // The "translate(50%, 100%)" counters and resets the default
                // anchoring of the advanced marker element from the api
                contentElement.style.transform = `translate(50%, 100%) translate(${translateX}, ${translateY})`;
                // data-origin is needed to identify the custom marker content in the
                // InfoWindow component as well as in the global CSS used to disable
                // the pointer event when anchor points are used in older Google Maps
                // versions.
                marker.dataset.origin = 'rgm';
                globalStyleManager.addAdvancedMarkerPointerEventsOverwrite();
            }
        }
    }, [marker, anchorPoint, anchorLeft, anchorTop, hasChildren]);
}

function useCircle(props) {
    var _a, _b, _c;
    const { onClick, onDrag, onDragStart, onDragEnd, onMouseOver, onMouseOut, onRadiusChanged, onCenterChanged, center, defaultCenter, radius, defaultRadius } = props, destructuredOptions = __rest(props, ["onClick", "onDrag", "onDragStart", "onDragEnd", "onMouseOver", "onMouseOut", "onRadiusChanged", "onCenterChanged", "center", "defaultCenter", "radius", "defaultRadius"]);
    const [circle, setCircle] = useState(null);
    const map = useMap();
    // Memoize options with automatic inference of clickable/draggable/editable
    const circleOptions = useMemoized(Object.assign(Object.assign({}, destructuredOptions), { clickable: (_a = destructuredOptions.clickable) !== null && _a !== void 0 ? _a : Boolean(onClick), draggable: (_b = destructuredOptions.draggable) !== null && _b !== void 0 ? _b : Boolean(onDrag || onDragStart || onDragEnd || onCenterChanged), editable: (_c = destructuredOptions.editable) !== null && _c !== void 0 ? _c : Boolean(onRadiusChanged) }), isDeepEqual);
    useEffect(() => {
        if (!map) {
            if (map === undefined)
                console.error('<Circle> has to be inside a Map component.');
            return;
        }
        const newCircle = new google.maps.Circle(Object.assign(Object.assign({}, circleOptions), { center: center !== null && center !== void 0 ? center : defaultCenter, radius: radius !== null && radius !== void 0 ? radius : defaultRadius }));
        newCircle.setMap(map);
        setCircle(newCircle);
        return () => {
            newCircle.setMap(null);
            setCircle(null);
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps -- circle options are handled separately to avoid recreating the instance
    }, [map]);
    useMapsEventListener(circle, 'click', onClick);
    useMapsEventListener(circle, 'drag', onDrag);
    useMapsEventListener(circle, 'dragstart', onDragStart);
    useMapsEventListener(circle, 'dragend', onDragEnd);
    useMapsEventListener(circle, 'mouseover', onMouseOver);
    useMapsEventListener(circle, 'mouseout', onMouseOut);
    useMapsEventListener(circle, 'radius_changed', onRadiusChanged
        ? () => {
            const newRadius = circle === null || circle === void 0 ? void 0 : circle.getRadius();
            if (newRadius !== undefined)
                onRadiusChanged(newRadius);
        }
        : null);
    useMapsEventListener(circle, 'center_changed', onCenterChanged
        ? () => {
            onCenterChanged(circle === null || circle === void 0 ? void 0 : circle.getCenter());
        }
        : null);
    useEffect(() => {
        if (!circle)
            return;
        circle.setOptions(circleOptions);
    }, [circle, circleOptions]);
    // Sync controlled center prop with the circle instance
    useEffect(() => {
        if (!circle || !center)
            return;
        if (!latLngEquals(center, circle.getCenter())) {
            circle.setCenter(center);
        }
    }, [circle, center]);
    // Sync controlled radius prop with the circle instance
    useEffect(() => {
        if (!circle || radius === undefined)
            return;
        if (radius !== circle.getRadius()) {
            circle.setRadius(radius);
        }
    }, [circle, radius]);
    return circle;
}
const Circle = forwardRef((props, ref) => {
    const circle = useCircle(props);
    useImperativeHandle(ref, () => circle, [circle]);
    return React__default.createElement(React__default.Fragment, null);
});
Circle.displayName = 'Circle';

function setValueForStyles(element, styles, prevStyles) {
    if (styles != null && typeof styles !== 'object') {
        throw new Error('The `style` prop expects a mapping from style properties to values, ' +
            "not a string. For example, style={{marginRight: spacing + 'em'}} when " +
            'using JSX.');
    }
    const elementStyle = element.style;
    // without `prevStyles`, just set all values
    if (prevStyles == null) {
        if (styles == null)
            return;
        for (const styleName in styles) {
            if (!styles.hasOwnProperty(styleName))
                continue;
            setValueForStyle(elementStyle, styleName, styles[styleName]);
        }
        return;
    }
    // unset all styles in `prevStyles` that aren't in `styles`
    for (const styleName in prevStyles) {
        if (prevStyles.hasOwnProperty(styleName) &&
            (styles == null || !styles.hasOwnProperty(styleName))) {
            // Clear style
            const isCustomProperty = styleName.indexOf('--') === 0;
            if (isCustomProperty) {
                elementStyle.setProperty(styleName, '');
            }
            else if (styleName === 'float') {
                elementStyle.cssFloat = '';
            }
            else {
                elementStyle[styleName] = '';
            }
        }
    }
    // only assign values from `styles` that are different from `prevStyles`
    if (styles == null)
        return;
    for (const styleName in styles) {
        const value = styles[styleName];
        if (styles.hasOwnProperty(styleName) &&
            prevStyles[styleName] !== value) {
            setValueForStyle(elementStyle, styleName, value);
        }
    }
}
function setValueForStyle(elementStyle, styleName, value) {
    const isCustomProperty = styleName.indexOf('--') === 0;
    // falsy values will unset the style property
    if (value == null || typeof value === 'boolean' || value === '') {
        if (isCustomProperty) {
            elementStyle.setProperty(styleName, '');
        }
        else if (styleName === 'float') {
            elementStyle.cssFloat = '';
        }
        else {
            elementStyle[styleName] = '';
        }
    }
    // custom properties can't be directly assigned
    else if (isCustomProperty) {
        elementStyle.setProperty(styleName, value);
    }
    // numeric values are treated as 'px' unless the style property expects unitless numbers
    else if (typeof value === 'number' &&
        value !== 0 &&
        !isUnitlessNumber(styleName)) {
        elementStyle[styleName] = value + 'px'; // Presumes implicit 'px' suffix for unitless numbers
    }
    // everything else can just be assigned
    else {
        if (styleName === 'float') {
            elementStyle.cssFloat = value;
        }
        else {
            elementStyle[styleName] = ('' + value).trim();
        }
    }
}
// CSS properties which accept numbers but are not in units of "px".
const unitlessNumbers = new Set([
    'animationIterationCount',
    'aspectRatio',
    'borderImageOutset',
    'borderImageSlice',
    'borderImageWidth',
    'boxFlex',
    'boxFlexGroup',
    'boxOrdinalGroup',
    'columnCount',
    'columns',
    'flex',
    'flexGrow',
    'flexPositive',
    'flexShrink',
    'flexNegative',
    'flexOrder',
    'gridArea',
    'gridRow',
    'gridRowEnd',
    'gridRowSpan',
    'gridRowStart',
    'gridColumn',
    'gridColumnEnd',
    'gridColumnSpan',
    'gridColumnStart',
    'fontWeight',
    'lineClamp',
    'lineHeight',
    'opacity',
    'order',
    'orphans',
    'scale',
    'tabSize',
    'widows',
    'zIndex',
    'zoom',
    'fillOpacity', // SVG-related properties
    'floodOpacity',
    'stopOpacity',
    'strokeDasharray',
    'strokeDashoffset',
    'strokeMiterlimit',
    'strokeOpacity',
    'strokeWidth'
]);
function isUnitlessNumber(name) {
    return unitlessNumbers.has(name);
}

/**
 * Component to render an Info Window with the Maps JavaScript API
 */
const InfoWindow = props => {
    const { 
    // content options
    children, headerContent, style, className, pixelOffset, 
    // open options
    anchor, shouldFocus, 
    // events
    onClose, onCloseClick } = props, 
    // other options
    volatileInfoWindowOptions = __rest(props, ["children", "headerContent", "style", "className", "pixelOffset", "anchor", "shouldFocus", "onClose", "onCloseClick"]);
    // ## create infowindow instance once the mapsLibrary is available.
    const mapsLibrary = useMapsLibrary('maps');
    const [infoWindow, setInfoWindow] = useState(null);
    const contentContainerRef = useRef$1(null);
    const headerContainerRef = useRef$1(null);
    const infoWindowOptions = useMemoized(volatileInfoWindowOptions, isDeepEqual);
    // ---- initial mount: create content- and header container, create and
    // configure the InfoWindow instance
    useEffect(() => {
        if (!mapsLibrary)
            return;
        contentContainerRef.current = document.createElement('div');
        headerContainerRef.current = document.createElement('div');
        const opts = Object.assign({}, infoWindowOptions);
        if (pixelOffset) {
            opts.pixelOffset = new google.maps.Size(pixelOffset[0], pixelOffset[1]);
        }
        if (headerContent) {
            // if headerContent is specified as string we can directly forward it,
            // otherwise we'll pass the element the portal will render into
            opts.headerContent =
                typeof headerContent === 'string'
                    ? headerContent
                    : headerContainerRef.current;
        }
        // intentionally shadowing the state variables here
        const infoWindow = new google.maps.InfoWindow(opts);
        infoWindow.setContent(contentContainerRef.current);
        setInfoWindow(infoWindow);
        // unmount: remove infoWindow and content elements (note: close is called in a different effect-cleanup)
        return () => {
            var _a, _b;
            infoWindow.setContent(null);
            (_a = contentContainerRef.current) === null || _a === void 0 ? void 0 : _a.remove();
            (_b = headerContainerRef.current) === null || _b === void 0 ? void 0 : _b.remove();
            contentContainerRef.current = null;
            headerContainerRef.current = null;
            setInfoWindow(null);
        };
    }, 
    // `infoWindowOptions` and other props are missing from dependencies:
    //
    // We don't want to re-create the infowindow instance
    // when the options change.
    // Updating the options is handled in the useEffect below.
    //
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [mapsLibrary]);
    // ---- update className and styles for `contentContainer`
    // prevStyleRef stores previously applied style properties, so they can be
    // removed when unset
    const prevStyleRef = useRef$1(null);
    useLayoutEffect$1(() => {
        if (!infoWindow || !contentContainerRef.current)
            return;
        setValueForStyles(contentContainerRef.current, style || null, prevStyleRef.current);
        prevStyleRef.current = style || null;
        if (className !== contentContainerRef.current.className)
            contentContainerRef.current.className = className || '';
    }, [infoWindow, className, style]);
    // ---- update options
    useEffect(() => {
        if (!infoWindow)
            return;
        const opts = Object.assign({}, infoWindowOptions);
        if (!pixelOffset) {
            opts.pixelOffset = null;
        }
        else {
            opts.pixelOffset = new google.maps.Size(pixelOffset[0], pixelOffset[1]);
        }
        if (!headerContent) {
            opts.headerContent = null;
        }
        else {
            opts.headerContent =
                typeof headerContent === 'string'
                    ? headerContent
                    : headerContainerRef.current;
        }
        infoWindow.setOptions(opts);
    }, 
    // dependency `infoWindow` isn't needed since options are also passed
    // to the constructor when a new infoWindow is created.
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [infoWindowOptions, pixelOffset, headerContent]);
    // ---- bind event handlers
    useMapsEventListener(infoWindow, 'close', onClose);
    useMapsEventListener(infoWindow, 'closeclick', onCloseClick);
    // ---- open info window when content and map are available
    const map = useMap();
    useEffect(() => {
        var _a;
        // `anchor === null` means an anchor is defined but not ready yet.
        if (!map || !infoWindow || anchor === null)
            return;
        const isOpenedWithAnchor = !!anchor;
        const openOptions = { map };
        if (anchor) {
            openOptions.anchor = anchor;
            // Only do the infowindow adjusting when dealing with an AdvancedMarker
            if (isAdvancedMarker(anchor) && anchor.content instanceof Element) {
                const anchorBcr = anchor === null || anchor === void 0 ? void 0 : anchor.getBoundingClientRect();
                // This checks whether or not the anchor has custom content with our own
                // div wrapper. If not, that means we have a regular AdvancedMarker without
                // children, or an AdvancedMarker that uses the anchorLeft/anchorTop props.
                // In that case we do not want to adjust the infowindow since it is all handled correctly
                // by the Google Maps API.
                if (anchorBcr && anchor.dataset.origin === 'rgm') {
                    // We can safely typecast here since we control that element and we know that
                    // it is a div
                    const anchorDomContent = (_a = anchor.content.firstElementChild) === null || _a === void 0 ? void 0 : _a.firstElementChild;
                    const contentBcr = anchorDomContent === null || anchorDomContent === void 0 ? void 0 : anchorDomContent.getBoundingClientRect();
                    // center infowindow above marker
                    const anchorOffsetX = contentBcr.x -
                        anchorBcr.x +
                        (contentBcr.width - anchorBcr.width) / 2;
                    const anchorOffsetY = contentBcr.y - anchorBcr.y;
                    const opts = Object.assign({}, infoWindowOptions);
                    opts.pixelOffset = new google.maps.Size(pixelOffset ? pixelOffset[0] + anchorOffsetX : anchorOffsetX, pixelOffset ? pixelOffset[1] + anchorOffsetY : anchorOffsetY);
                    infoWindow.setOptions(opts);
                }
            }
        }
        if (shouldFocus !== undefined) {
            openOptions.shouldFocus = shouldFocus;
        }
        infoWindow.open(openOptions);
        return () => {
            // Note: when the infowindow has an anchor, it will automatically show up again when the
            // anchor was removed from the map before infoWindow.close() is called but the it gets
            // added back to the map after that.
            // More information here: https://issuetracker.google.com/issues/343750849
            if (isOpenedWithAnchor)
                infoWindow.set('anchor', null);
            infoWindow.close();
        };
    }, [infoWindow, anchor, map, shouldFocus, infoWindowOptions, pixelOffset]);
    return (React__default.createElement(React__default.Fragment, null,
        contentContainerRef.current &&
            createPortal(children, contentContainerRef.current),
        headerContainerRef.current !== null &&
            createPortal(headerContent, headerContainerRef.current)));
};

/**
 * Extracts paths as a nested array from a Polygon instance.
 */
function getPathsArray(polygon) {
    const mvcPaths = polygon.getPaths();
    const result = [];
    for (let i = 0; i < mvcPaths.getLength(); i++) {
        result.push(mvcPaths.getAt(i).getArray());
    }
    return result;
}
function usePolygon(props) {
    var _a, _b, _c;
    const { onClick, onDrag, onDragStart, onDragEnd, onMouseOver, onMouseOut, onPathsChanged, polygon: externalPolygon, encodedPaths, paths, defaultPaths } = props, destructuredOptions = __rest(props, ["onClick", "onDrag", "onDragStart", "onDragEnd", "onMouseOver", "onMouseOut", "onPathsChanged", "polygon", "encodedPaths", "paths", "defaultPaths"]);
    const [polygon, setPolygon] = useState(null);
    const map = useMap();
    const geometryLibrary = useMapsLibrary('geometry');
    // Track if we're programmatically updating to avoid firing onPathsChanged
    const isUpdatingRef = useRef$1(false);
    // Memoize options with automatic inference of clickable/draggable/editable
    const polygonOptions = useMemoized(Object.assign(Object.assign({}, destructuredOptions), { clickable: (_a = destructuredOptions.clickable) !== null && _a !== void 0 ? _a : Boolean(onClick), draggable: (_b = destructuredOptions.draggable) !== null && _b !== void 0 ? _b : Boolean(onDrag || onDragStart || onDragEnd || onPathsChanged), editable: (_c = destructuredOptions.editable) !== null && _c !== void 0 ? _c : Boolean(onPathsChanged) }), isDeepEqual);
    useEffect(() => {
        if (!map) {
            if (map === undefined)
                console.error('<Polygon> has to be inside a Map component.');
            return;
        }
        // Use provided instance or create a new one
        let instance;
        if (externalPolygon) {
            instance = externalPolygon;
            // Apply initial paths and options to the existing instance
            const initialPaths = paths !== null && paths !== void 0 ? paths : defaultPaths;
            if (initialPaths && Array.isArray(initialPaths)) {
                instance.setPaths(initialPaths);
            }
            instance.setOptions(polygonOptions);
        }
        else {
            const initialPaths = paths !== null && paths !== void 0 ? paths : defaultPaths;
            const polygonOptionsWithPaths = Object.assign({}, polygonOptions);
            // Google Maps throws "not an Array" error if paths is undefined
            if (initialPaths && Array.isArray(initialPaths)) {
                polygonOptionsWithPaths.paths = initialPaths;
            }
            instance = new google.maps.Polygon(polygonOptionsWithPaths);
        }
        instance.setMap(map);
        setPolygon(instance);
        return () => {
            instance.setMap(null);
            setPolygon(null);
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps -- options are handled separately to avoid recreating the instance
    }, [map, externalPolygon]);
    useMapsEventListener(polygon, 'click', onClick);
    useMapsEventListener(polygon, 'drag', onDrag);
    useMapsEventListener(polygon, 'dragstart', onDragStart);
    useMapsEventListener(polygon, 'mouseover', onMouseOver);
    useMapsEventListener(polygon, 'mouseout', onMouseOut);
    // Fire onPathsChanged on dragend (when whole polygon is dragged)
    useMapsEventListener(polygon, 'dragend', (e) => {
        onDragEnd === null || onDragEnd === void 0 ? void 0 : onDragEnd(e);
        if (onPathsChanged && polygon && !isUpdatingRef.current) {
            onPathsChanged(getPathsArray(polygon));
        }
    });
    // Subscribe to MVCArray events for vertex-level edits
    useEffect(() => {
        if (!polygon || !onPathsChanged)
            return;
        const listeners = [];
        const mvcPaths = polygon.getPaths();
        if (typeof mvcPaths.getLength !==
            'function' ||
            typeof mvcPaths.getAt !== 'function') {
            return;
        }
        const handlePathsChange = () => {
            if (!isUpdatingRef.current) {
                onPathsChanged(getPathsArray(polygon));
            }
        };
        // Subscribe to each inner path's events
        const subscribeToInnerPath = (innerPath) => {
            listeners.push(google.maps.event.addListener(innerPath, 'insert_at', handlePathsChange));
            listeners.push(google.maps.event.addListener(innerPath, 'remove_at', handlePathsChange));
            listeners.push(google.maps.event.addListener(innerPath, 'set_at', handlePathsChange));
        };
        // Subscribe to existing inner paths
        for (let i = 0; i < mvcPaths.getLength(); i++) {
            subscribeToInnerPath(mvcPaths.getAt(i));
        }
        // Subscribe to outer array changes (paths added/removed)
        listeners.push(google.maps.event.addListener(mvcPaths, 'insert_at', (index) => {
            subscribeToInnerPath(mvcPaths.getAt(index));
            handlePathsChange();
        }));
        listeners.push(google.maps.event.addListener(mvcPaths, 'set_at', (index) => {
            subscribeToInnerPath(mvcPaths.getAt(index));
            handlePathsChange();
        }));
        listeners.push(google.maps.event.addListener(mvcPaths, 'remove_at', handlePathsChange));
        return () => {
            listeners.forEach(listener => listener.remove());
        };
    }, [
        polygon,
        onPathsChanged,
        paths,
        encodedPaths,
        polygonOptions.editable,
        polygonOptions.draggable
    ]);
    useEffect(() => {
        if (!polygon)
            return;
        polygon.setOptions(polygonOptions);
    }, [polygon, polygonOptions]);
    // Sync controlled paths prop with the polygon instance
    useEffect(() => {
        if (!polygon || !paths)
            return;
        if (!Array.isArray(paths))
            return;
        // Normalize to nested array for comparison
        const firstPath = paths[0];
        const normalizedPaths = Array.isArray(firstPath) ? paths : [paths];
        const currentPaths = polygon.getPaths();
        if (!pathsEquals(normalizedPaths, currentPaths)) {
            isUpdatingRef.current = true;
            polygon.setPaths(paths);
            isUpdatingRef.current = false;
        }
    }, [polygon, paths]);
    // Handle encoded paths
    useEffect(() => {
        if (!polygon || !encodedPaths || !geometryLibrary)
            return;
        isUpdatingRef.current = true;
        const decodedPaths = encodedPaths.map(encodedPath => geometryLibrary.encoding.decodePath(encodedPath));
        polygon.setPaths(decodedPaths);
        isUpdatingRef.current = false;
    }, [polygon, encodedPaths, geometryLibrary]);
    return polygon;
}
const Polygon = forwardRef((props, ref) => {
    const polygon = usePolygon(props);
    useImperativeHandle(ref, () => polygon, [polygon]);
    return React__default.createElement(React__default.Fragment, null);
});
Polygon.displayName = 'Polygon';

function usePolyline(props) {
    var _a, _b, _c;
    const { onClick, onDrag, onDragStart, onDragEnd, onMouseOver, onMouseOut, onPathChanged, polyline: externalPolyline, encodedPath, path, defaultPath } = props, destructuredOptions = __rest(props, ["onClick", "onDrag", "onDragStart", "onDragEnd", "onMouseOver", "onMouseOut", "onPathChanged", "polyline", "encodedPath", "path", "defaultPath"]);
    const [polyline, setPolyline] = useState(null);
    const map = useMap();
    const geometryLibrary = useMapsLibrary('geometry');
    // Track if we're programmatically updating to avoid firing onPathChanged
    const isUpdatingRef = useRef$1(false);
    // Memoize options with automatic inference of clickable/draggable/editable
    const polylineOptions = useMemoized(Object.assign(Object.assign({}, destructuredOptions), { clickable: (_a = destructuredOptions.clickable) !== null && _a !== void 0 ? _a : Boolean(onClick), draggable: (_b = destructuredOptions.draggable) !== null && _b !== void 0 ? _b : Boolean(onDrag || onDragStart || onDragEnd || onPathChanged), editable: (_c = destructuredOptions.editable) !== null && _c !== void 0 ? _c : Boolean(onPathChanged) }), isDeepEqual);
    useEffect(() => {
        if (!map) {
            if (map === undefined)
                console.error('<Polyline> has to be inside a Map component.');
            return;
        }
        // Use provided instance or create a new one
        let instance;
        if (externalPolyline) {
            instance = externalPolyline;
            // Apply initial path and options to the existing instance
            const initialPath = path !== null && path !== void 0 ? path : defaultPath;
            if (initialPath && Array.isArray(initialPath)) {
                instance.setPath(initialPath);
            }
            instance.setOptions(polylineOptions);
        }
        else {
            const initialPath = path !== null && path !== void 0 ? path : defaultPath;
            const polylineOptionsWithPath = Object.assign({}, polylineOptions);
            // Google Maps throws "not an Array" error if path is undefined
            if (initialPath && Array.isArray(initialPath)) {
                polylineOptionsWithPath.path = initialPath;
            }
            instance = new google.maps.Polyline(polylineOptionsWithPath);
        }
        instance.setMap(map);
        setPolyline(instance);
        return () => {
            instance.setMap(null);
            setPolyline(null);
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps -- options are handled separately to avoid recreating the instance
    }, [map, externalPolyline]);
    useMapsEventListener(polyline, 'click', onClick);
    useMapsEventListener(polyline, 'drag', onDrag);
    useMapsEventListener(polyline, 'dragstart', onDragStart);
    useMapsEventListener(polyline, 'mouseover', onMouseOver);
    useMapsEventListener(polyline, 'mouseout', onMouseOut);
    // Fire onPathChanged on dragend (when whole polyline is dragged)
    useMapsEventListener(polyline, 'dragend', (e) => {
        onDragEnd === null || onDragEnd === void 0 ? void 0 : onDragEnd(e);
        if (onPathChanged && polyline && !isUpdatingRef.current) {
            onPathChanged(polyline.getPath().getArray());
        }
    });
    // Subscribe to MVCArray events for vertex-level edits
    useEffect(() => {
        if (!polyline || !onPathChanged)
            return;
        const mvcPath = polyline.getPath();
        if (!mvcPath)
            return;
        const handlePathChange = () => {
            if (!isUpdatingRef.current) {
                onPathChanged(mvcPath.getArray());
            }
        };
        const listeners = [
            google.maps.event.addListener(mvcPath, 'insert_at', handlePathChange),
            google.maps.event.addListener(mvcPath, 'remove_at', handlePathChange),
            google.maps.event.addListener(mvcPath, 'set_at', handlePathChange)
        ];
        return () => {
            listeners.forEach(listener => listener.remove());
        };
    }, [
        polyline,
        onPathChanged,
        path,
        encodedPath,
        polylineOptions.editable,
        polylineOptions.draggable
    ]);
    useEffect(() => {
        if (!polyline)
            return;
        polyline.setOptions(polylineOptions);
    }, [polyline, polylineOptions]);
    // Sync controlled path prop with the polyline instance
    useEffect(() => {
        if (!polyline || !path)
            return;
        const currentPath = polyline.getPath();
        if (!pathEquals(path, currentPath)) {
            isUpdatingRef.current = true;
            polyline.setPath(path);
            isUpdatingRef.current = false;
        }
    }, [polyline, path]);
    // Handle encoded path
    useEffect(() => {
        if (!polyline || !encodedPath || !geometryLibrary)
            return;
        isUpdatingRef.current = true;
        const decodedPath = geometryLibrary.encoding.decodePath(encodedPath);
        polyline.setPath(decodedPath);
        isUpdatingRef.current = false;
    }, [polyline, encodedPath, geometryLibrary]);
    return polyline;
}
const Polyline = forwardRef((props, ref) => {
    const polyline = usePolyline(props);
    useImperativeHandle(ref, () => polyline, [polyline]);
    return React__default.createElement(React__default.Fragment, null);
});
Polyline.displayName = 'Polyline';

const DEFAULT_CAMERA_STATE = {
    center: { lat: 0, lng: 0, altitude: 0 },
    range: 0,
    heading: 0,
    tilt: 0,
    roll: 0
};
/**
 * Camera property names that correspond to gmp-*change events.
 */
const CAMERA_PROPS = ['center', 'range', 'heading', 'tilt', 'roll'];
/**
 * Updates the camera state ref with values from the map element.
 */
function updateCameraState(map3d, ref, prop) {
    const value = map3d[prop];
    if (value == null)
        return;
    if (prop === 'center') {
        // The center property returns a LatLngAltitude object, convert to literal
        const center = value;
        ref.current.center = center.toJSON
            ? center.toJSON()
            : center;
    }
    else {
        ref.current[prop] = value;
    }
}
/**
 * Creates a mutable ref object to track the last known state of the 3D map camera.
 * This is used in `useMap3DCameraParams` to reduce stuttering by avoiding updates
 * of the map camera with values that have already been processed.
 *
 * @internal
 */
function useTrackedCameraStateRef3D(map3d) {
    const forceUpdate = useForceUpdate();
    const ref = useRef$1(Object.assign({}, DEFAULT_CAMERA_STATE));
    useEffect(() => {
        if (!map3d)
            return;
        const listeners = [];
        for (const prop of CAMERA_PROPS) {
            const eventName = `gmp-${prop}change`;
            const handler = () => {
                updateCameraState(map3d, ref, prop);
                forceUpdate();
            };
            map3d.addEventListener(eventName, handler);
            listeners.push(() => map3d.removeEventListener(eventName, handler));
        }
        return () => {
            for (const removeListener of listeners) {
                removeListener();
            }
        };
    }, [map3d, forceUpdate]);
    return ref;
}

/**
 * Hook to manage the Map3DElement instance lifecycle.
 *
 * Handles:
 * - Waiting for the 'maps3d' library to load
 * - Waiting for the 'gmp-map-3d' custom element to be defined
 * - Creating a callback ref for the element
 * - Applying initial options when the element is ready
 * - Tracking camera state
 *
 * @internal
 */
function useMap3DInstance(props) {
    const maps3dLib = useMapsLibrary('maps3d');
    const [customElementReady, setCustomElementReady] = useState(false);
    const [, containerRef] = useCallbackRef();
    const [map3d, map3dRef] = useCallbackRef();
    const cameraStateRef = useTrackedCameraStateRef3D(map3d);
    useEffect(() => {
        customElements.whenDefined('gmp-map-3d').then(() => {
            setCustomElementReady(true);
        });
    }, []);
    // Apply initial options once when the element is first available
    useEffect(() => {
        if (!map3d)
            return;
        const { center, heading, tilt, range, roll, defaultCenter, defaultHeading, defaultTilt, defaultRange, defaultRoll, 
        // Non-element props to exclude
        id, style, className, children, onCenterChanged, onHeadingChanged, onTiltChanged, onRangeChanged, onRollChanged, onCameraChanged, onClick, onSteadyChange, onAnimationEnd, onError, mode, gestureHandling } = props, elementOptions = __rest(props, ["center", "heading", "tilt", "range", "roll", "defaultCenter", "defaultHeading", "defaultTilt", "defaultRange", "defaultRoll", "id", "style", "className", "children", "onCenterChanged", "onHeadingChanged", "onTiltChanged", "onRangeChanged", "onRollChanged", "onCameraChanged", "onClick", "onSteadyChange", "onAnimationEnd", "onError", "mode", "gestureHandling"]);
        const initialCenter = center !== null && center !== void 0 ? center : defaultCenter;
        const initialHeading = heading !== null && heading !== void 0 ? heading : defaultHeading;
        const initialTilt = tilt !== null && tilt !== void 0 ? tilt : defaultTilt;
        const initialRange = range !== null && range !== void 0 ? range : defaultRange;
        const initialRoll = roll !== null && roll !== void 0 ? roll : defaultRoll;
        const initialOptions = Object.assign({}, elementOptions);
        if (initialCenter)
            initialOptions.center = initialCenter;
        if (initialHeading !== undefined)
            initialOptions.heading = initialHeading;
        if (initialTilt !== undefined)
            initialOptions.tilt = initialTilt;
        if (initialRange !== undefined)
            initialOptions.range = initialRange;
        if (initialRoll !== undefined)
            initialOptions.roll = initialRoll;
        Object.assign(map3d, initialOptions);
    }, 
    // this effect should only run when the map3d element first becomes
    // available, so we skip re-running it when other props change.
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [map3d]);
    const isReady = !!maps3dLib && customElementReady;
    return [map3d, containerRef, map3dRef, cameraStateRef, isReady];
}

/* eslint-disable react-hooks/immutability -- Google Maps API objects are designed to be mutated */
/**
 * Converts a LatLngAltitude or LatLngAltitudeLiteral to a literal object.
 */
function toLatLngAltitudeLiteral(value) {
    if (!value)
        return null;
    // Check if it's a LatLngAltitude object with toJSON method
    if ('toJSON' in value && typeof value.toJSON === 'function') {
        return value.toJSON();
    }
    return value;
}
/**
 * Hook to update Map3D camera parameters when props change.
 * Compares the current camera state with props and updates only when there are differences.
 *
 * @internal
 */
function useMap3DCameraParams(map3d, cameraStateRef, props) {
    var _a, _b, _c, _d, _e, _f, _g;
    const centerLiteral = toLatLngAltitudeLiteral(props.center);
    const lat = (_a = centerLiteral === null || centerLiteral === void 0 ? void 0 : centerLiteral.lat) !== null && _a !== void 0 ? _a : null;
    const lng = (_b = centerLiteral === null || centerLiteral === void 0 ? void 0 : centerLiteral.lng) !== null && _b !== void 0 ? _b : null;
    const altitude = (_c = centerLiteral === null || centerLiteral === void 0 ? void 0 : centerLiteral.altitude) !== null && _c !== void 0 ? _c : null;
    const range = (_d = props.range) !== null && _d !== void 0 ? _d : null;
    const heading = (_e = props.heading) !== null && _e !== void 0 ? _e : null;
    const tilt = (_f = props.tilt) !== null && _f !== void 0 ? _f : null;
    const roll = (_g = props.roll) !== null && _g !== void 0 ? _g : null;
    // Runs on every render to sync controlled camera props with the map element
    useLayoutEffect$1(() => {
        var _a;
        if (!map3d)
            return;
        const currentState = cameraStateRef.current;
        if (lat !== null &&
            lng !== null &&
            (currentState.center.lat !== lat ||
                currentState.center.lng !== lng ||
                (altitude !== null && currentState.center.altitude !== altitude))) {
            map3d.center = {
                lat,
                lng,
                altitude: (_a = altitude !== null && altitude !== void 0 ? altitude : currentState.center.altitude) !== null && _a !== void 0 ? _a : 0
            };
        }
        if (range !== null && currentState.range !== range) {
            map3d.range = range;
        }
        if (heading !== null && currentState.heading !== heading) {
            map3d.heading = heading;
        }
        if (tilt !== null && currentState.tilt !== tilt) {
            map3d.tilt = tilt;
        }
        if (roll !== null && currentState.roll !== roll) {
            map3d.roll = roll;
        }
    });
}

/**
 * Camera-related event types for the aggregated onCameraChanged handler.
 */
const CAMERA_EVENTS = [
    'gmp-centerchange',
    'gmp-headingchange',
    'gmp-tiltchange',
    'gmp-rangechange',
    'gmp-rollchange'
];
/**
 * Creates a camera changed event with current camera state.
 */
function createCameraEvent(map3d, type) {
    const center = map3d.center;
    // Normalize center to LatLngAltitudeLiteral
    // If center is a LatLngAltitude class instance, it has a toJSON method
    // Otherwise it's already a literal object
    let centerLiteral;
    if (center && 'toJSON' in center && typeof center.toJSON === 'function') {
        centerLiteral = center.toJSON();
    }
    else if (center) {
        centerLiteral = center;
    }
    else {
        centerLiteral = { lat: 0, lng: 0, altitude: 0 };
    }
    return {
        type,
        map3d,
        detail: {
            center: centerLiteral,
            range: map3d.range || 0,
            heading: map3d.heading || 0,
            tilt: map3d.tilt || 0,
            roll: map3d.roll || 0
        }
    };
}
/**
 * Creates a click event from a LocationClickEvent or PlaceClickEvent.
 */
function createClickEvent(map3d, srcEvent) {
    const placeClickEvent = srcEvent;
    return {
        type: 'gmp-click',
        map3d,
        detail: {
            position: srcEvent.position || null,
            placeId: placeClickEvent.placeId
        }
    };
}
/**
 * Creates a steady change event.
 */
function createSteadyChangeEvent(map3d, srcEvent) {
    return {
        type: 'gmp-steadychange',
        map3d,
        detail: {
            isSteady: srcEvent.isSteady
        }
    };
}
/**
 * Hook to set up event handlers for Map3D events.
 *
 * @internal
 */
function useMap3DEvents(map3d, props) {
    const { onCenterChanged, onHeadingChanged, onTiltChanged, onRangeChanged, onRollChanged, onCameraChanged, onClick, onSteadyChange, onAnimationEnd, onError } = props;
    useMap3DEvent(map3d, 'gmp-centerchange', onCenterChanged, createCameraEvent);
    useMap3DEvent(map3d, 'gmp-headingchange', onHeadingChanged, createCameraEvent);
    useMap3DEvent(map3d, 'gmp-tiltchange', onTiltChanged, createCameraEvent);
    useMap3DEvent(map3d, 'gmp-rangechange', onRangeChanged, createCameraEvent);
    useMap3DEvent(map3d, 'gmp-rollchange', onRollChanged, createCameraEvent);
    // onCameraChanged aggregates all camera property change events into one handler
    useEffect(() => {
        if (!map3d || !onCameraChanged)
            return;
        const handler = () => {
            onCameraChanged(createCameraEvent(map3d, 'camerachange'));
        };
        for (const eventName of CAMERA_EVENTS) {
            map3d.addEventListener(eventName, handler);
        }
        return () => {
            for (const eventName of CAMERA_EVENTS) {
                map3d.removeEventListener(eventName, handler);
            }
        };
    }, [map3d, onCameraChanged]);
    useEffect(() => {
        if (!map3d || !onClick)
            return;
        const handler = (ev) => {
            onClick(createClickEvent(map3d, ev));
        };
        map3d.addEventListener('gmp-click', handler);
        return () => map3d.removeEventListener('gmp-click', handler);
    }, [map3d, onClick]);
    useEffect(() => {
        if (!map3d || !onSteadyChange)
            return;
        const handler = (ev) => {
            onSteadyChange(createSteadyChangeEvent(map3d, ev));
        };
        map3d.addEventListener('gmp-steadychange', handler);
        return () => map3d.removeEventListener('gmp-steadychange', handler);
    }, [map3d, onSteadyChange]);
    useMap3DEvent(map3d, 'gmp-animationend', onAnimationEnd, (map3d, type) => ({
        type,
        map3d
    }));
    useMap3DEvent(map3d, 'gmp-error', onError, (map3d, type) => ({
        type,
        map3d
    }));
}
/**
 * Helper hook for individual events.
 */
function useMap3DEvent(map3d, eventName, handler, createEvent) {
    useEffect(() => {
        if (!map3d || !handler)
            return;
        const listener = () => {
            handler(createEvent(map3d, eventName));
        };
        map3d.addEventListener(eventName, listener);
        return () => map3d.removeEventListener(eventName, listener);
    }, [map3d, eventName, handler, createEvent]);
}

/**
 * Set of option keys that can be updated on Map3DElement.
 * Camera props (center, heading, tilt, range, roll) are handled separately.
 */
const MAP_3D_OPTION_KEYS = new Set([
    'bounds',
    'defaultUIHidden',
    'gestureHandling',
    'internalUsageAttributionIds',
    'maxAltitude',
    'maxHeading',
    'maxTilt',
    'minAltitude',
    'minHeading',
    'minTilt',
    'mode'
]);
/**
 * Hook to update Map3D options when props change.
 *
 * @internal
 */
function useMap3DOptions(map3d, props) {
    // Filter props to only include valid option keys, memoized to avoid
    // creating a new object on every render
    const options = useMemo(() => {
        const result = {};
        const keys = Object.keys(props);
        for (const key of keys) {
            if (!MAP_3D_OPTION_KEYS.has(key))
                continue;
            const value = props[key];
            if (value === undefined)
                continue;
            result[key] = value;
        }
        return result;
    }, [props]);
    useDeepCompareEffect(() => {
        if (!map3d)
            return;
        Object.assign(map3d, options);
    }, [map3d, options]);
}

/**
 * MapMode for specifying how the 3D map should be rendered.
 * This mirrors google.maps.maps3d.MapMode but is available without waiting for the API to load.
 */
const MapMode = {
    /** This map mode displays a transparent layer of major streets on satellite imagery. */
    HYBRID: 'HYBRID',
    /** This map mode displays satellite or photorealistic imagery. */
    SATELLITE: 'SATELLITE'
};
/**
 * GestureHandling for specifying how gesture events should be handled on the map.
 * This mirrors google.maps.maps3d.GestureHandling but is available without waiting for the API to load.
 */
const GestureHandling = {
    /**
     * This lets the map choose whether to use cooperative or greedy gesture handling.
     * This is the default behavior if not specified.
     */
    AUTO: 'AUTO',
    /**
     * This forces cooperative mode, where modifier keys or two-finger gestures
     * are required to scroll the map.
     */
    COOPERATIVE: 'COOPERATIVE',
    /**
     * This forces greedy mode, where the host page cannot be scrolled from user
     * events on the map element.
     */
    GREEDY: 'GREEDY'
};
/**
 * React context for accessing the Map3D instance from child components.
 */
const GoogleMaps3DContext = React__default.createContext(null);
/**
 * Default styles for the map container.
 */
const DEFAULT_CONTAINER_STYLE = {
    width: '100%',
    height: '100%',
    position: 'relative'
};
/**
 * A React component that renders a 3D map using the Google Maps JavaScript API.
 *
 * @example
 * ```tsx
 * <APIProvider apiKey={API_KEY}>
 *   <Map3D
 *     defaultCenter={{ lat: 37.7749, lng: -122.4194, altitude: 1000 }}
 *     defaultRange={5000}
 *     defaultHeading={0}
 *     defaultTilt={45}
 *   />
 * </APIProvider>
 * ```
 */
const Map3D = forwardRef((props, ref) => {
    const { children, id, className, style } = props;
    // Verify we're inside an APIProvider
    const context = useContext(APIProviderContext);
    if (!context) {
        throw new Error('<Map3D> can only be used inside an <APIProvider> component.');
    }
    const { addMap3DInstance, removeMap3DInstance } = context;
    const [map3d, containerRef, map3dRef, cameraStateRef, isReady] = useMap3DInstance(props);
    useMap3DCameraParams(map3d, cameraStateRef, props);
    useMap3DEvents(map3d, props);
    useMap3DOptions(map3d, props);
    useEffect(() => {
        if (!map3d)
            return;
        const instanceId = id !== null && id !== void 0 ? id : 'default';
        addMap3DInstance(map3d, instanceId);
        return () => {
            removeMap3DInstance(instanceId);
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [map3d, id]);
    useImperativeHandle(ref, () => ({
        map3d,
        flyCameraAround: (options) => {
            map3d === null || map3d === void 0 ? void 0 : map3d.flyCameraAround(options);
        },
        flyCameraTo: (options) => {
            map3d === null || map3d === void 0 ? void 0 : map3d.flyCameraTo(options);
        },
        stopCameraAnimation: () => {
            map3d === null || map3d === void 0 ? void 0 : map3d.stopCameraAnimation();
        }
    }), [map3d]);
    const combinedStyle = useMemo(() => (Object.assign(Object.assign({}, DEFAULT_CONTAINER_STYLE), style)), [style]);
    const contextValue = useMemo(() => ({ map3d }), [map3d]);
    if (!isReady) {
        return (React__default.createElement("div", Object.assign({ ref: containerRef, "data-testid": "map-3d", style: className ? undefined : combinedStyle, className: className }, (id ? { id } : {}))));
    }
    return (React__default.createElement("div", Object.assign({ ref: containerRef, "data-testid": "map-3d", style: className ? undefined : combinedStyle, className: className }, (id ? { id } : {})),
        React__default.createElement("gmp-map-3d", { ref: map3dRef, style: { width: '100%', height: '100%' } }, map3d && (React__default.createElement(GoogleMaps3DContext.Provider, { value: contextValue }, children)))));
});
Map3D.displayName = 'Map3D';

const Marker3DContext = createContext(null);
/**
 * Hook to access the Marker3D context.
 */
function useMarker3D() {
    return useContext(Marker3DContext);
}
/**
 * AltitudeMode for specifying how altitude is interpreted for 3D elements.
 * This mirrors google.maps.maps3d.AltitudeMode but is available without waiting for the API to load.
 */
const AltitudeMode = {
    /** Allows to express objects relative to the average mean sea level. */
    ABSOLUTE: 'ABSOLUTE',
    /** Allows to express objects placed on the ground. */
    CLAMP_TO_GROUND: 'CLAMP_TO_GROUND',
    /** Allows to express objects relative to the ground surface. */
    RELATIVE_TO_GROUND: 'RELATIVE_TO_GROUND',
    /** Allows to express objects relative to the highest of ground+building+water surface. */
    RELATIVE_TO_MESH: 'RELATIVE_TO_MESH'
};
/**
 * Marker3D component for displaying markers on a Map3D.
 *
 * Automatically uses Marker3DInteractiveElement when onClick is provided,
 * otherwise uses Marker3DElement.
 *
 * Children can include:
 * - `<img>` elements (automatically wrapped in <template>)
 * - `<svg>` elements (automatically wrapped in <template>)
 * - PinElement instances (passed through directly)
 *
 * @example
 * ```tsx
 * // Basic marker
 * <Marker3D position={{ lat: 37.7749, lng: -122.4194 }} label="SF" />
 *
 * // Interactive marker
 * <Marker3D
 *   position={{ lat: 37.7749, lng: -122.4194 }}
 *   onClick={() => console.log('clicked')}
 *   title="Click me"
 * />
 *
 * // Custom marker with image
 * <Marker3D position={{ lat: 37.7749, lng: -122.4194 }}>
 *   <img src="/icon.png" width={32} height={32} />
 * </Marker3D>
 * ```
 */
const Marker3D = forwardRef(function Marker3D(props, ref) {
    const { children, onClick, position, altitudeMode, collisionBehavior, drawsWhenOccluded, extruded, label, sizePreserved, zIndex, title } = props;
    const isInteractive = Boolean(onClick);
    const [marker, setMarker] = useState(null);
    // Track if a child component (like Pin) is handling its own content
    const [contentHandledExternally, setContentHandledExternally] = useState(false);
    // Create a container for rendering React children to be wrapped and relocated
    // into the parent gmp-marker-3d element.
    const contentContainer = useMemo(() => {
        const container = document.createElement('div');
        container.style.display = 'none';
        document.body.appendChild(container);
        return container;
    }, []);
    // Remove the container on unmount
    useEffect(() => {
        return () => contentContainer.remove();
    }, [contentContainer]);
    // Callback ref that sets both internal state and forwards the ref
    const markerRef = useCallback((node) => {
        setMarker(node);
        if (typeof ref === 'function') {
            ref(node);
        }
        else if (ref) {
            ref.current = node;
        }
    }, [ref]);
    useDomEventListener(marker, 'gmp-click', onClick);
    // Move React children to marker's slot, wrapping img/svg in <template> as required by the API
    useLayoutEffect$1(() => {
        if (contentHandledExternally)
            return;
        if (!marker || !contentContainer)
            return;
        while (marker.firstChild) {
            marker.removeChild(marker.firstChild);
        }
        const childNodes = Array.from(contentContainer.childNodes);
        for (const node of childNodes) {
            if (node.nodeType !== Node.ELEMENT_NODE)
                continue;
            const element = node;
            const tagName = element.tagName.toLowerCase();
            if (tagName === 'img' || tagName === 'svg') {
                const template = document.createElement('template');
                template.content.appendChild(element.cloneNode(true));
                marker.appendChild(template);
            }
            else {
                marker.appendChild(element.cloneNode(true));
            }
        }
    }, [marker, contentContainer, children, contentHandledExternally]);
    const contextValue = useMemo(() => ({ marker, setContentHandledExternally }), [marker]);
    usePropBinding(marker, 'position', position);
    usePropBinding(marker, 'altitudeMode', altitudeMode);
    usePropBinding(marker, 'collisionBehavior', collisionBehavior);
    usePropBinding(marker, 'drawsWhenOccluded', drawsWhenOccluded);
    usePropBinding(marker, 'extruded', extruded);
    usePropBinding(marker, 'label', label);
    usePropBinding(marker, 'sizePreserved', sizePreserved);
    usePropBinding(marker, 'zIndex', zIndex);
    usePropBinding(marker, 'title', title !== null && title !== void 0 ? title : '');
    return (React__default.createElement(Marker3DContext.Provider, { value: contextValue },
        isInteractive ? (React__default.createElement("gmp-marker-3d-interactive", { ref: markerRef })) : (React__default.createElement("gmp-marker-3d", { ref: markerRef })),
        createPortal(children, contentContainer)));
});
Marker3D.displayName = 'Marker3D';

/**
 * Popover component for displaying info windows on a Map3D.
 *
 * Similar to InfoWindow for 2D maps, Popover provides a way to show
 * contextual information at a specific location or attached to a marker
 * on a 3D map.
 *
 * @example
 * ```tsx
 * // Basic popover at position
 * <Popover
 *   position={{ lat: 37.7749, lng: -122.4194 }}
 *   open={isOpen}
 * >
 *   <div>Hello from San Francisco!</div>
 * </Popover>
 *
 * // Popover anchored to a marker (place as sibling, use anchor prop)
 * <Marker3D
 *   ref={markerRef}
 *   position={{ lat: 37.7749, lng: -122.4194 }}
 *   onClick={() => setOpen(true)}
 * />
 * <Popover
 *   anchor={markerRef.current}
 *   open={isOpen}
 *   onClose={() => setOpen(false)}
 * >
 *   <div>Marker info</div>
 * </Popover>
 * ```
 */
const Popover = forwardRef(function Popover(props, ref) {
    var _a;
    const { children, headerContent, style, className, open = true, position, anchor, anchorId, altitudeMode, lightDismissDisabled, autoPanDisabled, onClose } = props;
    const [popover, setPopover] = useState(null);
    const prevStyleRef = useRef$1(null);
    // Forward the ref to the parent
    useImperativeHandle(ref, () => popover, [popover]);
    // Observe the open attribute and call onClose when popover is automatically
    // closed by the Maps API (light dismiss)
    usePopoverCloseObserver(popover, open, onClose);
    // Set properties on the popover element
    usePropBinding(popover, 'open', open !== null && open !== void 0 ? open : false);
    usePropBinding(popover, 'altitudeMode', altitudeMode);
    usePropBinding(popover, 'lightDismissDisabled', lightDismissDisabled);
    usePropBinding(popover, 'autoPanDisabled', autoPanDisabled);
    // positionAnchor accepts a position, marker element, or marker ID string
    const positionAnchor = (_a = anchor !== null && anchor !== void 0 ? anchor : anchorId) !== null && _a !== void 0 ? _a : position;
    usePropBinding(popover, 'positionAnchor', positionAnchor);
    // Set styles via ref for compatibility with older React versions
    useLayoutEffect$1(() => {
        if (!popover)
            return;
        setValueForStyles(popover, style || null, prevStyleRef.current);
        prevStyleRef.current = style || null;
    }, [popover, style]);
    return (React__default.createElement("gmp-popover", { ref: setPopover, className: className },
        headerContent && React__default.createElement("div", { slot: "header" }, headerContent),
        children));
});
Popover.displayName = 'Popover';
/**
 * Custom hook to observe the open attribute of a popover element
 * and call onClose when it transitions from open to closed due to light dismiss.
 * Does not call onClose when the open prop changes programmatically.
 */
function usePopoverCloseObserver(popover, open, onClose) {
    const previousOpenState = useRef$1(undefined);
    const openPropRef = useRef$1(open);
    // Track the open prop value
    useEffect(() => {
        openPropRef.current = open;
    }, [open]);
    useEffect(() => {
        if (!popover || !onClose)
            return;
        const observer = new MutationObserver(mutations => {
            for (const mutation of mutations) {
                if (mutation.type === 'attributes' &&
                    mutation.attributeName === 'open') {
                    const isOpen = popover.hasAttribute('open');
                    // Only call onClose when:
                    // 1. Transitioning from open to closed
                    // 2. The prop hasn't changed to false (meaning this was light dismiss, not programmatic)
                    if (previousOpenState.current === true &&
                        !isOpen &&
                        openPropRef.current !== false) {
                        onClose();
                    }
                    previousOpenState.current = isOpen;
                }
            }
        });
        observer.observe(popover, {
            attributes: true,
            attributeFilter: ['open']
        });
        // Initialize the previous state
        previousOpenState.current = popover.hasAttribute('open');
        return () => {
            observer.disconnect();
        };
    }, [popover, onClose]);
}

/**
 * Formats a location into a string representation suitable for Google Static Maps API.
 *
 * @param location - The location to format, can be either a string or an object with lat/lng properties
 * @returns A string representation of the location in the format "lat,lng" or the original string
 *
 * @example
 * // Returns "40.714728,-73.998672"
 * formatLocation({ lat: 40.714728, lng: -73.998672 })
 *
 * @example
 * // Returns "New York, NY"
 * formatLocation("New York, NY")
 */
function formatLocation(location) {
    return typeof location === 'string'
        ? location
        : `${location.lat},${location.lng}`;
}
// Used for removing the leading pipe from the param string
function formatParam(string) {
    return string.slice(1);
}

/**
 * Assembles marker parameters for static maps.
 *
 * This function takes an array of markers and groups them by their style properties.
 * It then creates a string representation of these markers, including their styles and locations,
 * which can be used as parameters for static map APIs.
 *
 * @param {StaticMapsMarker[]} [markers=[]] - An array of markers to be processed. Each marker can have properties such as color, label, size, scale, icon, anchor, and location.
 * @returns {string[]} An array of strings, each representing a group of markers with their styles and locations.
 *
 * @example
 * const markers = [
 *   { color: 'blue', label: 'A', size: 'mid', location: '40.714728,-73.998672' },
 *   { color: 'blue', label: 'B', size: 'mid', location: '40.714728,-73.998672' },
 *   { icon: 'http://example.com/icon.png', location: { lat: 40.714728, lng: -73.998672 } }
 * ];
 * const params = assembleMarkerParams(markers);
 * // Params will be an array of strings representing the marker parameters
 * Example output: [
 *   "color:blue|label:A|size:mid|40.714728,-73.998672|40.714728,-73.998672",
 *   "color:blue|label:B|size:mid|40.714728,-73.998672|40.714728,-73.998672",
 *   "icon:http://example.com/icon.png|40.714728,-73.998672"
 * ]
 */
function assembleMarkerParams(markers = []) {
    const markerParams = [];
    // Group markers by style
    const markersByStyle = markers === null || markers === void 0 ? void 0 : markers.reduce((styles, marker) => {
        const { color = 'red', label, size, scale, icon, anchor } = marker;
        // Create a unique style key based on either icon properties or standard marker properties
        const relevantProps = icon ? [icon, anchor, scale] : [color, label, size];
        const key = relevantProps.filter(Boolean).join('-');
        styles[key] = styles[key] || [];
        styles[key].push(marker);
        return styles;
    }, {});
    Object.values(markersByStyle !== null && markersByStyle !== void 0 ? markersByStyle : {}).forEach(markers => {
        let markerParam = '';
        const { icon } = markers[0];
        // Create marker style from first marker in group since all markers share the same style.
        Object.entries(markers[0]).forEach(([key, value]) => {
            // Determine which properties to include based on whether marker uses custom icon
            const relevantKeys = icon
                ? ['icon', 'anchor', 'scale']
                : ['color', 'label', 'size'];
            if (relevantKeys.includes(key)) {
                markerParam += `|${key}:${value}`;
            }
        });
        // Add location coordinates for each marker in the style group
        // Handles both string locations and lat/lng object formats.
        for (const marker of markers) {
            const location = typeof marker.location === 'string'
                ? marker.location
                : `${marker.location.lat},${marker.location.lng}`;
            markerParam += `|${location}`;
        }
        markerParams.push(markerParam);
    });
    return markerParams.map(formatParam);
}

// Style properties that can be applied to paths in the Static Maps API
const PATH_STYLE_KEYS = ['color', 'weight', 'fillcolor', 'geodesic'];
/**
 * Builds the style portion of a path parameter string.
 * @param path - The path object containing style properties
 * @returns A string with style parameters in the format "|key:value"
 */
function buildStyleParams(path) {
    let styleParams = '';
    PATH_STYLE_KEYS.forEach(key => {
        if (path[key] !== undefined) {
            styleParams += `|${key}:${path[key]}`;
        }
    });
    return styleParams;
}
/**
 * Builds the coordinates portion of a path parameter string.
 * @param coordinates - Either a string or array of location objects
 * @returns A string with coordinates in the format "|lat,lng|lat,lng"
 */
function buildCoordinateParams(coordinates) {
    if (typeof coordinates === 'string') {
        return `|${decodeURIComponent(coordinates)}`;
    }
    return coordinates.map(location => `|${formatLocation(location)}`).join('');
}
/**
 * Assembles path parameters for the Static Maps API from an array of paths.
 *
 * This function constructs a string of path parameters for each path. Each path parameter string
 * includes the style properties and the coordinates of the paths.
 *
 * @param {Array<StaticMapsPath>} [paths=[]] - An array of paths to be assembled into path parameters.
 * @returns {Array<string>} An array of path parameter strings.
 *
 * @example
 * ```typescript
 * const paths = [
 *   {
 *     color: 'red',
 *     weight: 5,
 *     coordinates: [
 *       { lat: 40.714728, lng: -73.998672 },
 *       { lat: 40.718217, lng: -73.998284 }
 *     ]
 *   }
 * ];
 *
 * const pathParams = assemblePathParams(paths);
 * // Output: ['color:red|weight:5|40.714728,-73.998672|40.718217,-73.998284']
 * ```
 */
function assemblePathParams(paths = []) {
    return paths.map(path => {
        const styleParams = buildStyleParams(path);
        const coordinateParams = buildCoordinateParams(path.coordinates);
        const pathParam = styleParams + coordinateParams;
        return formatParam(pathParam);
    });
}

/**
 * Converts an array of Google Maps style objects into an array of style strings
 * compatible with the Google Static Maps API.
 *
 * @param styles - An array of Google Maps MapTypeStyle objects that define the styling rules
 * @returns An array of formatted style strings ready to be used with the Static Maps API
 *
 * @example
 * const styles = [{
 *   featureType: "road",
 *   elementType: "geometry",
 *   stylers: [{color: "#ff0000"}, {weight: 1}]
 * }];
 *
 * const styleStrings = assembleMapTypeStyles(styles);
 * // Returns: ["|feature:road|element:geometry|color:0xff0000|weight:1"]
 *
 * Each style string follows the format:
 * "feature:{featureType}|element:{elementType}|{stylerName}:{stylerValue}"
 *
 * Note: Color values with hexadecimal notation (#) are automatically converted
 * to the required 0x format for the Static Maps API.
 */
function assembleMapTypeStyles(styles) {
    return styles
        .map((mapTypeStyle) => {
        const { featureType, elementType, stylers = [] } = mapTypeStyle;
        let styleString = '';
        if (featureType) {
            styleString += `|feature:${featureType}`;
        }
        if (elementType) {
            styleString += `|element:${elementType}`;
        }
        for (const styler of stylers) {
            Object.entries(styler).forEach(([name, value]) => {
                styleString += `|${name}:${String(value).replace('#', '0x')}`;
            });
        }
        return styleString;
    })
        .map(formatParam);
}

const STATIC_MAPS_BASE = 'https://maps.googleapis.com/maps/api/staticmap';
/**
 * Creates a URL for the Google Static Maps API with the specified parameters.
 *
 * @param {Object} options - The configuration options for the static map
 * @param {string} options.apiKey - Your Google Maps API key (required)
 * @param {number} options.width - The width of the map image in pixels (required)
 * @param {number} options.height - The height of the map image in pixels (required)
 * @param {StaticMapsLocation} [options.center] - The center point of the map (lat/lng or address).
 *  Required if no markers or paths or "visible locations" are provided.
 * @param {number} [options.zoom] - The zoom level of the map. Required if no markers or paths or "visible locations" are provided.
 * @param {1|2|4} [options.scale] - The resolution of the map (1, 2, or 4)
 * @param {string} [options.format] - The image format (png, png8, png32, gif, jpg, jpg-baseline)
 * @param {string} [options.mapType] - The type of map (roadmap, satellite, terrain, hybrid)
 * @param {string} [options.language] - The language of the map labels
 * @param {string} [options.region] - The region code for the map
 * @param {string} [options.map_id] - The Cloud-based map style ID
 * @param {StaticMapsMarker[]} [options.markers=[]] - Array of markers to display on the map
 * @param {StaticMapsPath[]} [options.paths=[]] - Array of paths to display on the map
 * @param {StaticMapsLocation[]} [options.visible=[]] - Array of locations that should be visible on the map
 * @param {MapTypeStyle[]} [options.style=[]] - Array of style objects to customize the map appearance
 *
 * @returns {string} The complete Google Static Maps API URL
 *
 * @throws {Error} If API key is not provided
 * @throws {Error} If width or height is not provided
 *
 * @example
 * const url = createStaticMapsUrl({
 *   apiKey: 'YOUR_API_KEY',
 *   width: 600,
 *   height: 400,
 *   center: { lat: 40.714728, lng: -73.998672 },
 *   zoom: 12,
 *   markers: [
 *     {
 *       location: { lat: 40.714728, lng: -73.998672 },
 *       color: 'red',
 *       label: 'A'
 *     }
 *   ],
 *   paths: [
 *     {
 *       coordinates: [
 *         { lat: 40.714728, lng: -73.998672 },
 *         { lat: 40.719728, lng: -73.991672 }
 *       ],
 *       color: '0x0000ff',
 *       weight: 5
 *     }
 *   ],
 *   style: [
 *     {
 *       featureType: 'road',
 *       elementType: 'geometry',
 *       stylers: [{color: '#00ff00'}]
 *     }
 *   ]
 * });
 *
 * // Results in URL similar to:
 * // https://maps.googleapis.com/maps/api/staticmap?key=YOUR_API_KEY
 * // &size=600x400
 * // &center=40.714728,-73.998672&zoom=12
 * // &markers=color:red|label:A|40.714728,-73.998672
 * // &path=color:0x0000ff|weight:5|40.714728,-73.998672|40.719728,-73.991672
 * // &style=feature:road|element:geometry|color:0x00ff00
 */
function createStaticMapsUrl({ apiKey, width, height, center, zoom, scale, format, mapType, language, region, mapId, markers = [], paths = [], visible = [], style = [] }) {
    if (!apiKey) {
        console.warn('API key is required');
    }
    if (!width || !height) {
        console.warn('Width and height are required');
    }
    const params = Object.assign(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({ key: apiKey, size: `${width}x${height}` }, (center && { center: formatLocation(center) })), (zoom && { zoom })), (scale && { scale })), (format && { format })), (mapType && { maptype: mapType })), (language && { language })), (region && { region })), (mapId && { map_id: mapId }));
    const url = new URL(STATIC_MAPS_BASE);
    // Params that don't need special handling
    Object.entries(params).forEach(([key, value]) => {
        url.searchParams.append(key, String(value));
    });
    // Assemble Markers
    for (const markerParam of assembleMarkerParams(markers)) {
        url.searchParams.append('markers', markerParam);
    }
    // Assemble Paths
    for (const pathParam of assemblePathParams(paths)) {
        url.searchParams.append('path', pathParam);
    }
    // Assemble visible locations
    if (visible.length) {
        url.searchParams.append('visible', visible.map(location => formatLocation(location)).join('|'));
    }
    // Assemble Map Type Styles
    for (const styleString of assembleMapTypeStyles(style)) {
        url.searchParams.append('style', styleString);
    }
    return url.toString();
}

const StaticMap = (props) => {
    const { url, className } = props;
    if (!url)
        throw new Error('URL is required');
    return React__default.createElement("img", { className: className, src: url, width: "100%" });
};

/**
 * Copy of the `google.maps.ControlPosition` constants.
 * They have to be duplicated here since we can't wait for the maps API to load to be able to use them.
 */
const ControlPosition = {
    TOP_LEFT: 1,
    TOP_CENTER: 2,
    TOP: 2,
    TOP_RIGHT: 3,
    LEFT_CENTER: 4,
    LEFT_TOP: 5,
    LEFT: 5,
    LEFT_BOTTOM: 6,
    RIGHT_TOP: 7,
    RIGHT: 7,
    RIGHT_CENTER: 8,
    RIGHT_BOTTOM: 9,
    BOTTOM_LEFT: 10,
    BOTTOM_CENTER: 11,
    BOTTOM: 11,
    BOTTOM_RIGHT: 12,
    CENTER: 13,
    BLOCK_START_INLINE_START: 14,
    BLOCK_START_INLINE_CENTER: 15,
    BLOCK_START_INLINE_END: 16,
    INLINE_START_BLOCK_CENTER: 17,
    INLINE_START_BLOCK_START: 18,
    INLINE_START_BLOCK_END: 19,
    INLINE_END_BLOCK_START: 20,
    INLINE_END_BLOCK_CENTER: 21,
    INLINE_END_BLOCK_END: 22,
    BLOCK_END_INLINE_START: 23,
    BLOCK_END_INLINE_CENTER: 24,
    BLOCK_END_INLINE_END: 25
};
const MapControl = ({ children, position }) => {
    const controlContainer = useMemo(() => document.createElement('div'), []);
    const map = useMap();
    useEffect(() => {
        if (!map)
            return;
        const controls = map.controls[position];
        controls.push(controlContainer);
        return () => {
            const controlsArray = controls.getArray();
            // controlsArray could be undefined if the map is in an undefined state (e.g. invalid API-key, see #276
            if (!controlsArray)
                return;
            const index = controlsArray.indexOf(controlContainer);
            controls.removeAt(index);
        };
    }, [controlContainer, map, position]);
    return createPortal(children, controlContainer);
};

function useMarker(props) {
    const [marker, setMarker] = useState(null);
    const map = useMap();
    const { onClick, onDrag, onDragStart, onDragEnd, onMouseOver, onMouseOut } = props, markerOptions = __rest(props, ["onClick", "onDrag", "onDragStart", "onDragEnd", "onMouseOver", "onMouseOut"]);
    const { position, draggable } = markerOptions;
    // create marker instance and add to the map once the map is available
    useEffect(() => {
        if (!map) {
            if (map === undefined)
                console.error('<Marker> has to be inside a Map component.');
            return;
        }
        const newMarker = new google.maps.Marker(markerOptions);
        newMarker.setMap(map);
        setMarker(newMarker);
        return () => {
            newMarker.setMap(null);
            setMarker(null);
        };
        // We do not want to re-render the whole marker when the options change.
        // Marker options update is handled in a useEffect below.
        // Excluding markerOptions from dependency array on purpose here.
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [map]);
    // attach and re-attach event-handlers when any of the properties change
    useEffect(() => {
        if (!marker)
            return;
        const m = marker;
        // Add event listeners
        const gme = google.maps.event;
        if (onClick)
            gme.addListener(m, 'click', onClick);
        if (onDrag)
            gme.addListener(m, 'drag', onDrag);
        if (onDragStart)
            gme.addListener(m, 'dragstart', onDragStart);
        if (onDragEnd)
            gme.addListener(m, 'dragend', onDragEnd);
        if (onMouseOver)
            gme.addListener(m, 'mouseover', onMouseOver);
        if (onMouseOut)
            gme.addListener(m, 'mouseout', onMouseOut);
        marker.setDraggable(Boolean(draggable));
        return () => {
            gme.clearInstanceListeners(m);
        };
    }, [
        marker,
        draggable,
        onClick,
        onDrag,
        onDragStart,
        onDragEnd,
        onMouseOver,
        onMouseOut
    ]);
    // update markerOptions (note the dependencies aren't properly checked
    // here, we just assume that setOptions is smart enough to not waste a
    // lot of time updating values that didn't change)
    useEffect(() => {
        if (!marker)
            return;
        if (markerOptions)
            marker.setOptions(markerOptions);
    }, [marker, markerOptions]);
    // update position when changed
    useEffect(() => {
        // Should not update position when draggable
        if (draggable || !position || !marker)
            return;
        marker.setPosition(position);
    }, [draggable, position, marker]);
    return marker;
}
/**
 * Component to render a marker on a map
 */
const Marker = forwardRef((props, ref) => {
    const marker = useMarker(props);
    useImperativeHandle(ref, () => marker, [marker]);
    return React__default.createElement(React__default.Fragment, null);
});
Marker.displayName = 'Marker';
function useMarkerRef() {
    const [marker, setMarker] = useState(null);
    const refCallback = useCallback((m) => {
        setMarker(m);
    }, []);
    return [refCallback, marker];
}

/**
 * Legacy implementation of Pin component for Maps API < 3.62
 * Component to configure the appearance of an AdvancedMarker or Marker3D
 */
const PinLegacy = props => {
    const advancedMarkerContext = useContext(AdvancedMarkerContext);
    const marker3dContext = useContext(Marker3DContext);
    const advancedMarker = advancedMarkerContext === null || advancedMarkerContext === void 0 ? void 0 : advancedMarkerContext.marker;
    const marker3d = marker3dContext === null || marker3dContext === void 0 ? void 0 : marker3dContext.marker;
    const setContentHandledExternally = marker3dContext === null || marker3dContext === void 0 ? void 0 : marker3dContext.setContentHandledExternally;
    const glyphContainer = useMemo(() => document.createElement('div'), []);
    // Load marker library for Marker3D case (AdvancedMarker already has it loaded)
    const markerLibrary = useMapsLibrary('marker');
    // Signal to Marker3D that we're handling content
    useEffect(() => {
        if (marker3d && setContentHandledExternally) {
            setContentHandledExternally(true);
            return () => setContentHandledExternally(false);
        }
    }, [marker3d, setContentHandledExternally]);
    // Handle Marker3D case - append PinElement directly
    useEffect(() => {
        if (!marker3d || !markerLibrary)
            return;
        const pinViewOptions = Object.assign({}, props);
        const pinElement = new markerLibrary.PinElement(pinViewOptions);
        // Set glyph to glyph container if children are present
        if (props.children) {
            pinElement.glyph = glyphContainer;
        }
        // Append PinElement directly to Marker3D (not in a template)
        // PinElement is a special element that Marker3D can handle natively
        while (marker3d.firstChild) {
            marker3d.removeChild(marker3d.firstChild);
        }
        marker3d.append(pinElement);
        return () => {
            // Cleanup handled by marker3d removal
        };
    }, [marker3d, markerLibrary, glyphContainer, props]);
    // Create Pin View instance for AdvancedMarker
    useEffect(() => {
        var _a;
        // Skip if we're in Marker3D context
        if (marker3d)
            return;
        if (!advancedMarker) {
            if (advancedMarker === undefined && marker3d === undefined) {
                console.error('The <Pin> component can only be used inside <AdvancedMarker> or <Marker3D>.');
            }
            return;
        }
        if (props.glyph && props.children) {
            logErrorOnce('The <Pin> component only uses children to render the glyph if both the glyph property and children are present.');
        }
        if (Children.count(props.children) > 1) {
            logErrorOnce('Passing multiple children to the <Pin> component might lead to unexpected results.');
        }
        const pinViewOptions = Object.assign({}, props);
        const pinElement = new google.maps.marker.PinElement(pinViewOptions);
        // Set glyph to glyph container if children are present (rendered via portal).
        // If both props.glyph and props.children are present, props.children takes priority.
        if (props.children) {
            pinElement.glyph = glyphContainer;
        }
        // Set content of Advanced Marker View to the Pin View element
        // Here we are selecting the anchor container.
        // The hierarchy is as follows:
        // "advancedMarker.content" (from google) -> "anchor container"
        const markerContent = (_a = advancedMarker.content) === null || _a === void 0 ? void 0 : _a.firstChild;
        while (markerContent === null || markerContent === void 0 ? void 0 : markerContent.firstChild) {
            markerContent.removeChild(markerContent.firstChild);
        }
        if (markerContent) {
            markerContent.appendChild(pinElement.element);
        }
    }, [advancedMarker, glyphContainer, marker3d, props]);
    return createPortal(props.children, glyphContainer);
};

/**
 * Modern implementation of Pin component for Maps API >= 3.62
 * Uses the new custom element API with glyphSrc/glyphText properties
 * Component to configure the appearance of an AdvancedMarker or Marker3D
 */
const PinModern = props => {
    const { children } = props;
    const advancedMarkerContext = useContext(AdvancedMarkerContext);
    const marker3dContext = useContext(Marker3DContext);
    const advancedMarker = advancedMarkerContext === null || advancedMarkerContext === void 0 ? void 0 : advancedMarkerContext.marker;
    const marker3d = marker3dContext === null || marker3dContext === void 0 ? void 0 : marker3dContext.marker;
    const setContentHandledExternally = marker3dContext === null || marker3dContext === void 0 ? void 0 : marker3dContext.setContentHandledExternally;
    const glyphContainer = useMemo(() => document.createElement('div'), []);
    const markerLibrary = useMapsLibrary('marker');
    // Signal to Marker3D that we're handling content
    useEffect(() => {
        if (marker3d && setContentHandledExternally) {
            setContentHandledExternally(true);
            return () => setContentHandledExternally(false);
        }
    }, [marker3d, setContentHandledExternally]);
    // Handle Marker3D case
    useEffect(() => {
        if (!marker3d || !markerLibrary)
            return;
        const pinOptions = Object.assign({}, props);
        const pinElement = new markerLibrary.PinElement(pinOptions);
        // Children still rendered via portal into a container
        if (children) {
            // In modern API, DOM children of PinElement serve as glyph
            pinElement.appendChild(glyphContainer);
        }
        // Append PinElement directly (not .element - that's deprecated)
        while (marker3d.firstChild) {
            marker3d.removeChild(marker3d.firstChild);
        }
        marker3d.appendChild(pinElement);
        return () => {
            // Cleanup handled by marker3d removal
        };
    }, [marker3d, markerLibrary, glyphContainer, children, props]);
    // Handle AdvancedMarker case
    useEffect(() => {
        var _a;
        if (marker3d)
            return; // Skip if in Marker3D context
        if (!advancedMarker || !markerLibrary) {
            if (advancedMarker === undefined && marker3d === undefined) {
                console.error('The <Pin> component can only be used inside <AdvancedMarker> or <Marker3D>.');
            }
            return;
        }
        if ((props.glyphSrc || props.glyphText) && children) {
            logErrorOnce('The <Pin> component only uses children to render the glyph if both glyphSrc/glyphText and children are present.');
        }
        if (Children.count(children) > 1) {
            logErrorOnce('Passing multiple children to the <Pin> component might lead to unexpected results.');
        }
        const pinOptions = Object.assign({}, props);
        const pinElement = new markerLibrary.PinElement(pinOptions);
        // Children rendered via portal into container
        if (children) {
            pinElement.appendChild(glyphContainer);
        }
        const markerContent = (_a = advancedMarker.content) === null || _a === void 0 ? void 0 : _a.firstChild;
        while (markerContent === null || markerContent === void 0 ? void 0 : markerContent.firstChild) {
            markerContent.removeChild(markerContent.firstChild);
        }
        if (markerContent) {
            // In modern API, append PinElement directly (not .element)
            markerContent.appendChild(pinElement);
        }
    }, [
        advancedMarker,
        markerLibrary,
        glyphContainer,
        marker3d,
        children,
        props
    ]);
    return createPortal(children, glyphContainer);
};

/**
 * Component to configure the appearance of an AdvancedMarker or Marker3D.
 *
 * Automatically detects the Maps API version and uses the appropriate implementation:
 * - Legacy implementation (< 3.62): Uses the original PinElement API with glyph property
 * - Modern implementation (≥ 3.62): Uses the new custom element API with glyphSrc/glyphText
 *
 * @example
 * ```tsx
 * // Basic usage
 * <AdvancedMarker position={{lat: 0, lng: 0}}>
 *   <Pin background="#FF0000" glyphColor="#FFFFFF" />
 * </AdvancedMarker>
 *
 * // With custom glyph (legacy)
 * <Pin glyph="📍" />
 *
 * // With custom glyph (modern, recommended for Maps API 3.62+)
 * <Pin glyphText="📍" />
 * <Pin glyphSrc="https://example.com/icon.png" />
 *
 * // With React children as glyph
 * <Pin>
 *   <CustomIcon />
 * </Pin>
 * ```
 */
const Pin = props => {
    var _a;
    const markerLibrary = useMapsLibrary('marker');
    // Detect capability after the marker library is loaded
    const isModern = useMemo(() => {
        if (!markerLibrary)
            return false;
        // Check if <gmp-pin> custom element exists
        // This happens while loading the marker library in Maps API 3.62+
        return (typeof customElements !== 'undefined' &&
            customElements.get('gmp-pin') !== undefined);
    }, [markerLibrary]);
    // Wait for marker library to load before rendering
    if (!markerLibrary)
        return null;
    // Map props based on which implementation we're using
    if (isModern) {
        // Modern: map old glyph prop to new glyphSrc/glyphText
        const { glyph, glyphSrc, glyphText } = props, restProps = __rest(props, ["glyph", "glyphSrc", "glyphText"]);
        const isUrl = glyph instanceof URL ||
            (typeof glyph === 'string' && glyph.startsWith('http'));
        const finalGlyphText = glyphText !== null && glyphText !== void 0 ? glyphText : (typeof glyph === 'string' && !isUrl ? glyph : undefined);
        const finalGlyphSrc = glyphSrc !== null && glyphSrc !== void 0 ? glyphSrc : (isUrl ? String(glyph) : undefined);
        return (React__default.createElement(PinModern, Object.assign({}, restProps, { glyphText: finalGlyphText, glyphSrc: finalGlyphSrc })));
    }
    else {
        // Legacy: map new glyphSrc/glyphText to old glyph prop
        const { glyph, glyphSrc, glyphText } = props, restProps = __rest(props, ["glyph", "glyphSrc", "glyphText"]);
        const finalGlyph = (_a = glyph !== null && glyph !== void 0 ? glyph : glyphSrc) !== null && _a !== void 0 ? _a : glyphText;
        return React__default.createElement(PinLegacy, Object.assign({}, restProps, { glyph: finalGlyph }));
    }
};

function useRectangle(props) {
    var _a, _b, _c;
    const { onClick, onDrag, onDragStart, onDragEnd, onMouseOver, onMouseOut, onBoundsChanged, bounds, defaultBounds } = props, destructuredOptions = __rest(props, ["onClick", "onDrag", "onDragStart", "onDragEnd", "onMouseOver", "onMouseOut", "onBoundsChanged", "bounds", "defaultBounds"]);
    const [rectangle, setRectangle] = useState(null);
    const map = useMap();
    // Memoize options with automatic inference of clickable/draggable/editable
    const rectangleOptions = useMemoized(Object.assign(Object.assign({}, destructuredOptions), { clickable: (_a = destructuredOptions.clickable) !== null && _a !== void 0 ? _a : Boolean(onClick), draggable: (_b = destructuredOptions.draggable) !== null && _b !== void 0 ? _b : Boolean(onDrag || onDragStart || onDragEnd || onBoundsChanged), editable: (_c = destructuredOptions.editable) !== null && _c !== void 0 ? _c : Boolean(onBoundsChanged) }), isDeepEqual);
    useEffect(() => {
        if (!map) {
            if (map === undefined)
                console.error('<Rectangle> has to be inside a Map component.');
            return;
        }
        const newRectangle = new google.maps.Rectangle(Object.assign(Object.assign({}, rectangleOptions), { bounds: bounds !== null && bounds !== void 0 ? bounds : defaultBounds }));
        newRectangle.setMap(map);
        setRectangle(newRectangle);
        return () => {
            newRectangle.setMap(null);
            setRectangle(null);
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps -- rectangle options are handled separately to avoid recreating the instance
    }, [map]);
    useMapsEventListener(rectangle, 'click', onClick);
    useMapsEventListener(rectangle, 'drag', onDrag);
    useMapsEventListener(rectangle, 'dragstart', onDragStart);
    useMapsEventListener(rectangle, 'dragend', onDragEnd);
    useMapsEventListener(rectangle, 'mouseover', onMouseOver);
    useMapsEventListener(rectangle, 'mouseout', onMouseOut);
    useMapsEventListener(rectangle, 'bounds_changed', onBoundsChanged
        ? () => {
            onBoundsChanged(rectangle === null || rectangle === void 0 ? void 0 : rectangle.getBounds());
        }
        : null);
    useEffect(() => {
        if (!rectangle)
            return;
        rectangle.setOptions(rectangleOptions);
    }, [rectangle, rectangleOptions]);
    // Sync controlled bounds prop with the rectangle instance
    useEffect(() => {
        if (!rectangle || !bounds)
            return;
        if (!boundsEquals(bounds, rectangle.getBounds())) {
            rectangle.setBounds(bounds);
        }
    }, [rectangle, bounds]);
    return rectangle;
}
const Rectangle = forwardRef((props, ref) => {
    const rectangle = useRectangle(props);
    useImperativeHandle(ref, () => rectangle, [
        rectangle
    ]);
    return React__default.createElement(React__default.Fragment, null);
});
Rectangle.displayName = 'Rectangle';

/**
 * Hook to retrieve a Map3DElement instance from context.
 *
 * When called without an id, it returns the map from the nearest parent Map3D
 * component. When called with an id, it retrieves the map with that id from
 * the APIProvider context.
 *
 * @param id - Optional id of the map to retrieve. If not specified, returns
 *   the parent map instance or the default map instance.
 *
 * @example
 * ```tsx
 * // Get the parent Map3D instance
 * function MyComponent() {
 *   const map3d = useMap3D();
 *   // ...
 * }
 *
 * // Get a specific Map3D instance by id
 * function ControlPanel() {
 *   const mainMap = useMap3D('main-map');
 *   const miniMap = useMap3D('mini-map');
 *   // ...
 * }
 * ```
 */
function useMap3D(id = null) {
    const apiContext = useContext(APIProviderContext);
    const map3dContext = useContext(GoogleMaps3DContext);
    if (apiContext === null) {
        logErrorOnce('useMap3D(): failed to retrieve APIProviderContext. ' +
            'Make sure that the <APIProvider> component exists and that the ' +
            'component you are calling `useMap3D()` from is a child of the ' +
            '<APIProvider>.');
        return null;
    }
    const { map3dInstances } = apiContext;
    // Lookup priority: explicit id > parent context > default instance
    if (id !== null)
        return map3dInstances[id] || null;
    if (map3dContext === null || map3dContext === void 0 ? void 0 : map3dContext.map3d)
        return map3dContext.map3d;
    return map3dInstances['default'] || null;
}

const mapLinear = (x, a1, a2, b1, b2) => b1 + ((x - a1) * (b2 - b1)) / (a2 - a1);
const getMapMaxTilt = (zoom) => {
    if (zoom <= 10) {
        return 30;
    }
    if (zoom >= 15.5) {
        return 67.5;
    }
    // range [10...14]
    if (zoom <= 14) {
        return mapLinear(zoom, 10, 14, 30, 45);
    }
    // range [14...15.5]
    return mapLinear(zoom, 14, 15.5, 45, 67.5);
};
/**
 * Function to limit the tilt range of the Google map when updating the view state
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const limitTiltRange = ({ viewState }) => {
    const pitch = viewState.pitch;
    const gmZoom = viewState.zoom + 1;
    const maxTilt = getMapMaxTilt(gmZoom);
    return Object.assign(Object.assign({}, viewState), { fovy: 25, pitch: Math.min(maxTilt, pitch) });
};

export { APILoadingStatus, APIProvider, APIProviderContext, AdvancedMarker, AdvancedMarkerAnchorPoint, AdvancedMarkerContext, AltitudeMode, Circle, CollisionBehavior, ColorScheme, ControlPosition, GestureHandling, GoogleMaps3DContext, GoogleMapsContext, InfoWindow, Map, Map3D, MapControl, MapMode, Marker, Marker3D, Marker3DContext, Pin, Polygon, Polyline, Popover, Rectangle, RenderingType, StaticMap, VERSION, __resetModuleState, boundsEquals, createStaticMapsUrl, isAdvancedMarker, isLatLngLiteral, latLngEquals, limitTiltRange, pathEquals, pathsEquals, toLatLngBoundsLiteral, toLatLngLiteral, useAdvancedMarkerRef, useApiIsLoaded, useApiLoadingStatus, useMap, useMap3D, useMapsLibrary, useMarker3D, useMarkerRef };
