// @base-ui/react/preview-card/popup/PreviewCardPopupDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PreviewCardPopupDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let PreviewCardPopupDataAttributes = exports.PreviewCardPopupDataAttributes = function (PreviewCardPopupDataAttributes) {
  /**
   * Present when the preview card is open.
   */
  PreviewCardPopupDataAttributes[PreviewCardPopupDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the preview card is closed.
   */
  PreviewCardPopupDataAttributes[PreviewCardPopupDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the preview card is animating in.
   */
  PreviewCardPopupDataAttributes[PreviewCardPopupDataAttributes["startingStyle"] = _popupStateMapping.CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the preview card is animating out.
   */
  PreviewCardPopupDataAttributes[PreviewCardPopupDataAttributes["endingStyle"] = _popupStateMapping.CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  PreviewCardPopupDataAttributes[PreviewCardPopupDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  PreviewCardPopupDataAttributes[PreviewCardPopupDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  return PreviewCardPopupDataAttributes;
}({});
