// @base-ui/react/scroll-area/root/stateAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.scrollAreaStateAttributesMapping = void 0;
var _ScrollAreaRootDataAttributes = require("./ScrollAreaRootDataAttributes");
const scrollAreaStateAttributesMapping = exports.scrollAreaStateAttributesMapping = {
  hasOverflowX: value => value ? {
    [_ScrollAreaRootDataAttributes.ScrollAreaRootDataAttributes.hasOverflowX]: ''
  } : null,
  hasOverflowY: value => value ? {
    [_ScrollAreaRootDataAttributes.ScrollAreaRootDataAttributes.hasOverflowY]: ''
  } : null,
  overflowXStart: value => value ? {
    [_ScrollAreaRootDataAttributes.ScrollAreaRootDataAttributes.overflowXStart]: ''
  } : null,
  overflowXEnd: value => value ? {
    [_ScrollAreaRootDataAttributes.ScrollAreaRootDataAttributes.overflowXEnd]: ''
  } : null,
  overflowYStart: value => value ? {
    [_ScrollAreaRootDataAttributes.ScrollAreaRootDataAttributes.overflowYStart]: ''
  } : null,
  overflowYEnd: value => value ? {
    [_ScrollAreaRootDataAttributes.ScrollAreaRootDataAttributes.overflowYEnd]: ''
  } : null,
  cornerHidden: () => null
};
