// @base-ui/react/esm/popover/arrow/PopoverArrowDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let PopoverArrowDataAttributes = function (PopoverArrowDataAttributes) {
  /**
   * Present when the popup is open.
   */
  PopoverArrowDataAttributes[PopoverArrowDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  PopoverArrowDataAttributes[PopoverArrowDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  PopoverArrowDataAttributes[PopoverArrowDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  PopoverArrowDataAttributes[PopoverArrowDataAttributes["align"] = CommonPopupDataAttributes.align] = "align";
  /**
   * Present when the popover arrow is uncentered.
   */
  PopoverArrowDataAttributes["uncentered"] = "data-uncentered";
  return PopoverArrowDataAttributes;
}({});
