// @base-ui/react/esm/menu/radio-item-indicator/MenuRadioItemIndicatorDataAttributes.js
import { TransitionStatusDataAttributes } from "../../utils/stateAttributesMapping.js";
export let MenuRadioItemIndicatorDataAttributes = function (MenuRadioItemIndicatorDataAttributes) {
  /**
   * Present when the menu radio item is selected.
   */
  MenuRadioItemIndicatorDataAttributes["checked"] = "data-checked";
  /**
   * Present when the menu radio item is not selected.
   */
  MenuRadioItemIndicatorDataAttributes["unchecked"] = "data-unchecked";
  /**
   * Present when the menu radio item is disabled.
   */
  MenuRadioItemIndicatorDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the radio indicator is animating in.
   */
  MenuRadioItemIndicatorDataAttributes[MenuRadioItemIndicatorDataAttributes["startingStyle"] = TransitionStatusDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the radio indicator is animating out.
   */
  MenuRadioItemIndicatorDataAttributes[MenuRadioItemIndicatorDataAttributes["endingStyle"] = TransitionStatusDataAttributes.endingStyle] = "endingStyle";
  return MenuRadioItemIndicatorDataAttributes;
}({});
