// @base-ui/react/esm/preview-card/popup/PreviewCardPopupDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let PreviewCardPopupDataAttributes = function (PreviewCardPopupDataAttributes) {
  /**
   * Present when the preview card is open.
   */
  PreviewCardPopupDataAttributes[PreviewCardPopupDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the preview card is closed.
   */
  PreviewCardPopupDataAttributes[PreviewCardPopupDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the preview card is animating in.
   */
  PreviewCardPopupDataAttributes[PreviewCardPopupDataAttributes["startingStyle"] = CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the preview card is animating out.
   */
  PreviewCardPopupDataAttributes[PreviewCardPopupDataAttributes["endingStyle"] = CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  PreviewCardPopupDataAttributes[PreviewCardPopupDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  PreviewCardPopupDataAttributes[PreviewCardPopupDataAttributes["align"] = CommonPopupDataAttributes.align] = "align";
  return PreviewCardPopupDataAttributes;
}({});
