// @base-ui/react/combobox/arrow/ComboboxArrowDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComboboxArrowDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let ComboboxArrowDataAttributes = exports.ComboboxArrowDataAttributes = function (ComboboxArrowDataAttributes) {
  /**
   * Present when the popup is open.
   */
  ComboboxArrowDataAttributes[ComboboxArrowDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  ComboboxArrowDataAttributes[ComboboxArrowDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  ComboboxArrowDataAttributes[ComboboxArrowDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  ComboboxArrowDataAttributes[ComboboxArrowDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  /**
   * Present when the arrow is uncentered.
   */
  ComboboxArrowDataAttributes["uncentered"] = "data-uncentered";
  return ComboboxArrowDataAttributes;
}({});
