// @base-ui/react/floating-ui-react/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "FloatingDelayGroup", {
  enumerable: true,
  get: function () {
    return _FloatingDelayGroup.FloatingDelayGroup;
  }
});
Object.defineProperty(exports, "FloatingFocusManager", {
  enumerable: true,
  get: function () {
    return _FloatingFocusManager.FloatingFocusManager;
  }
});
Object.defineProperty(exports, "FloatingNode", {
  enumerable: true,
  get: function () {
    return _FloatingTree.FloatingNode;
  }
});
Object.defineProperty(exports, "FloatingPortal", {
  enumerable: true,
  get: function () {
    return _FloatingPortal.FloatingPortal;
  }
});
Object.defineProperty(exports, "FloatingTree", {
  enumerable: true,
  get: function () {
    return _FloatingTree.FloatingTree;
  }
});
Object.defineProperty(exports, "FloatingTreeStore", {
  enumerable: true,
  get: function () {
    return _FloatingTreeStore.FloatingTreeStore;
  }
});
Object.defineProperty(exports, "arrow", {
  enumerable: true,
  get: function () {
    return _reactDom.arrow;
  }
});
Object.defineProperty(exports, "autoPlacement", {
  enumerable: true,
  get: function () {
    return _reactDom.autoPlacement;
  }
});
Object.defineProperty(exports, "autoUpdate", {
  enumerable: true,
  get: function () {
    return _reactDom.autoUpdate;
  }
});
Object.defineProperty(exports, "computePosition", {
  enumerable: true,
  get: function () {
    return _reactDom.computePosition;
  }
});
Object.defineProperty(exports, "detectOverflow", {
  enumerable: true,
  get: function () {
    return _reactDom.detectOverflow;
  }
});
Object.defineProperty(exports, "flip", {
  enumerable: true,
  get: function () {
    return _reactDom.flip;
  }
});
Object.defineProperty(exports, "getOverflowAncestors", {
  enumerable: true,
  get: function () {
    return _reactDom.getOverflowAncestors;
  }
});
Object.defineProperty(exports, "hide", {
  enumerable: true,
  get: function () {
    return _reactDom.hide;
  }
});
Object.defineProperty(exports, "inline", {
  enumerable: true,
  get: function () {
    return _reactDom.inline;
  }
});
Object.defineProperty(exports, "limitShift", {
  enumerable: true,
  get: function () {
    return _reactDom.limitShift;
  }
});
Object.defineProperty(exports, "offset", {
  enumerable: true,
  get: function () {
    return _reactDom.offset;
  }
});
Object.defineProperty(exports, "platform", {
  enumerable: true,
  get: function () {
    return _reactDom.platform;
  }
});
Object.defineProperty(exports, "safePolygon", {
  enumerable: true,
  get: function () {
    return _safePolygon.safePolygon;
  }
});
Object.defineProperty(exports, "shift", {
  enumerable: true,
  get: function () {
    return _reactDom.shift;
  }
});
Object.defineProperty(exports, "size", {
  enumerable: true,
  get: function () {
    return _reactDom.size;
  }
});
Object.defineProperty(exports, "useClick", {
  enumerable: true,
  get: function () {
    return _useClick.useClick;
  }
});
Object.defineProperty(exports, "useClientPoint", {
  enumerable: true,
  get: function () {
    return _useClientPoint.useClientPoint;
  }
});
Object.defineProperty(exports, "useDelayGroup", {
  enumerable: true,
  get: function () {
    return _FloatingDelayGroup.useDelayGroup;
  }
});
Object.defineProperty(exports, "useDismiss", {
  enumerable: true,
  get: function () {
    return _useDismiss.useDismiss;
  }
});
Object.defineProperty(exports, "useFloating", {
  enumerable: true,
  get: function () {
    return _useFloating.useFloating;
  }
});
Object.defineProperty(exports, "useFloatingNodeId", {
  enumerable: true,
  get: function () {
    return _FloatingTree.useFloatingNodeId;
  }
});
Object.defineProperty(exports, "useFloatingParentNodeId", {
  enumerable: true,
  get: function () {
    return _FloatingTree.useFloatingParentNodeId;
  }
});
Object.defineProperty(exports, "useFloatingPortalNode", {
  enumerable: true,
  get: function () {
    return _FloatingPortal.useFloatingPortalNode;
  }
});
Object.defineProperty(exports, "useFloatingRootContext", {
  enumerable: true,
  get: function () {
    return _useFloatingRootContext.useFloatingRootContext;
  }
});
Object.defineProperty(exports, "useFloatingTree", {
  enumerable: true,
  get: function () {
    return _FloatingTree.useFloatingTree;
  }
});
Object.defineProperty(exports, "useFocus", {
  enumerable: true,
  get: function () {
    return _useFocus.useFocus;
  }
});
Object.defineProperty(exports, "useHover", {
  enumerable: true,
  get: function () {
    return _useHover.useHover;
  }
});
Object.defineProperty(exports, "useHoverFloatingInteraction", {
  enumerable: true,
  get: function () {
    return _useHoverFloatingInteraction.useHoverFloatingInteraction;
  }
});
Object.defineProperty(exports, "useHoverReferenceInteraction", {
  enumerable: true,
  get: function () {
    return _useHoverReferenceInteraction.useHoverReferenceInteraction;
  }
});
Object.defineProperty(exports, "useInteractions", {
  enumerable: true,
  get: function () {
    return _useInteractions.useInteractions;
  }
});
Object.defineProperty(exports, "useListNavigation", {
  enumerable: true,
  get: function () {
    return _useListNavigation.useListNavigation;
  }
});
Object.defineProperty(exports, "useRole", {
  enumerable: true,
  get: function () {
    return _useRole.useRole;
  }
});
Object.defineProperty(exports, "useSyncedFloatingRootContext", {
  enumerable: true,
  get: function () {
    return _useSyncedFloatingRootContext.useSyncedFloatingRootContext;
  }
});
Object.defineProperty(exports, "useTypeahead", {
  enumerable: true,
  get: function () {
    return _useTypeahead.useTypeahead;
  }
});
var _FloatingDelayGroup = require("./components/FloatingDelayGroup");
var _FloatingFocusManager = require("./components/FloatingFocusManager");
var _FloatingPortal = require("./components/FloatingPortal");
var _FloatingTree = require("./components/FloatingTree");
var _FloatingTreeStore = require("./components/FloatingTreeStore");
var _useClick = require("./hooks/useClick");
var _useClientPoint = require("./hooks/useClientPoint");
var _useDismiss = require("./hooks/useDismiss");
var _useFloating = require("./hooks/useFloating");
var _useFloatingRootContext = require("./hooks/useFloatingRootContext");
var _useSyncedFloatingRootContext = require("./hooks/useSyncedFloatingRootContext");
var _useFocus = require("./hooks/useFocus");
var _useHoverFloatingInteraction = require("./hooks/useHoverFloatingInteraction");
var _useHoverReferenceInteraction = require("./hooks/useHoverReferenceInteraction");
var _useHover = require("./hooks/useHover");
var _useInteractions = require("./hooks/useInteractions");
var _useListNavigation = require("./hooks/useListNavigation");
var _useRole = require("./hooks/useRole");
var _useTypeahead = require("./hooks/useTypeahead");
var _safePolygon = require("./safePolygon");
var _reactDom = require("@floating-ui/react-dom");
