// @base-ui/react/select/popup/SelectPopupDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectPopupDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let SelectPopupDataAttributes = exports.SelectPopupDataAttributes = function (SelectPopupDataAttributes) {
  /**
   * Present when the select is open.
   */
  SelectPopupDataAttributes[SelectPopupDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the select is closed.
   */
  SelectPopupDataAttributes[SelectPopupDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the select is animating in.
   */
  SelectPopupDataAttributes[SelectPopupDataAttributes["startingStyle"] = _popupStateMapping.CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the select is animating out.
   */
  SelectPopupDataAttributes[SelectPopupDataAttributes["endingStyle"] = _popupStateMapping.CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'none' | 'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  SelectPopupDataAttributes[SelectPopupDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  SelectPopupDataAttributes[SelectPopupDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  return SelectPopupDataAttributes;
}({});
