// @base-ui/react/esm/popover/popup/PopoverPopupDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let PopoverPopupDataAttributes = function (PopoverPopupDataAttributes) {
  /**
   * Present when the popup is open.
   */
  PopoverPopupDataAttributes[PopoverPopupDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  PopoverPopupDataAttributes[PopoverPopupDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the popup is animating in.
   */
  PopoverPopupDataAttributes[PopoverPopupDataAttributes["startingStyle"] = CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the popup is animating out.
   */
  PopoverPopupDataAttributes[PopoverPopupDataAttributes["endingStyle"] = CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  PopoverPopupDataAttributes["side"] = "data-side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  PopoverPopupDataAttributes["align"] = "data-align";
  /**
   * Present if animations should be instant.
   * @type {'click' | 'dismiss'}
   */
  PopoverPopupDataAttributes["instant"] = "data-instant";
  return PopoverPopupDataAttributes;
}({});
