// @base-ui/react/slider/value/SliderValueDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SliderValueDataAttributes = void 0;
let SliderValueDataAttributes = exports.SliderValueDataAttributes = /*#__PURE__*/function (SliderValueDataAttributes) {
  /**
   * Present while the user is dragging.
   */
  SliderValueDataAttributes["dragging"] = "data-dragging";
  /**
   * Indicates the orientation of the slider.
   * @type {'horizontal' | 'vertical'}
   */
  SliderValueDataAttributes["orientation"] = "data-orientation";
  /**
   * Present when the slider is disabled.
   */
  SliderValueDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the slider is in valid state (when wrapped in Field.Root).
   */
  SliderValueDataAttributes["valid"] = "data-valid";
  /**
   * Present when the slider is in invalid state (when wrapped in Field.Root).
   */
  SliderValueDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the slider has been touched (when wrapped in Field.Root).
   */
  SliderValueDataAttributes["touched"] = "data-touched";
  /**
   * Present when the slider's value has changed (when wrapped in Field.Root).
   */
  SliderValueDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the slider is focused (when wrapped in Field.Root).
   */
  SliderValueDataAttributes["focused"] = "data-focused";
  return SliderValueDataAttributes;
}({});
