// @base-ui/react/esm/radio/indicator/RadioIndicatorDataAttributes.js
import { TransitionStatusDataAttributes } from "../../utils/stateAttributesMapping.js";
export let RadioIndicatorDataAttributes = function (RadioIndicatorDataAttributes) {
  /**
   * Present when the radio is checked.
   */
  RadioIndicatorDataAttributes["checked"] = "data-checked";
  /**
   * Present when the radio is not checked.
   */
  RadioIndicatorDataAttributes["unchecked"] = "data-unchecked";
  /**
   * Present when the radio is disabled.
   */
  RadioIndicatorDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the radio is readonly.
   */
  RadioIndicatorDataAttributes["readonly"] = "data-readonly";
  /**
   * Present when the radio is required.
   */
  RadioIndicatorDataAttributes["required"] = "data-required";
  /**
   * Present when the radio indicator is animating in.
   */
  RadioIndicatorDataAttributes[RadioIndicatorDataAttributes["startingStyle"] = TransitionStatusDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the radio indicator is animating out.
   */
  RadioIndicatorDataAttributes[RadioIndicatorDataAttributes["endingStyle"] = TransitionStatusDataAttributes.endingStyle] = "endingStyle";
  /**
   * Present when the radio is in valid state (when wrapped in Field.Root).
   */
  RadioIndicatorDataAttributes["valid"] = "data-valid";
  /**
   * Present when the radio is in invalid state (when wrapped in Field.Root).
   */
  RadioIndicatorDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the radio has been touched (when wrapped in Field.Root).
   */
  RadioIndicatorDataAttributes["touched"] = "data-touched";
  /**
   * Present when the radio's value has changed (when wrapped in Field.Root).
   */
  RadioIndicatorDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the radio is checked (when wrapped in Field.Root).
   */
  RadioIndicatorDataAttributes["filled"] = "data-filled";
  /**
   * Present when the radio is focused (when wrapped in Field.Root).
   */
  RadioIndicatorDataAttributes["focused"] = "data-focused";
  return RadioIndicatorDataAttributes;
}({});
