// @base-ui/react/popover/index.parts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Arrow", {
  enumerable: true,
  get: function () {
    return _PopoverArrow.PopoverArrow;
  }
});
Object.defineProperty(exports, "Backdrop", {
  enumerable: true,
  get: function () {
    return _PopoverBackdrop.PopoverBackdrop;
  }
});
Object.defineProperty(exports, "Close", {
  enumerable: true,
  get: function () {
    return _PopoverClose.PopoverClose;
  }
});
Object.defineProperty(exports, "Description", {
  enumerable: true,
  get: function () {
    return _PopoverDescription.PopoverDescription;
  }
});
Object.defineProperty(exports, "Handle", {
  enumerable: true,
  get: function () {
    return _PopoverHandle.PopoverHandle;
  }
});
Object.defineProperty(exports, "Popup", {
  enumerable: true,
  get: function () {
    return _PopoverPopup.PopoverPopup;
  }
});
Object.defineProperty(exports, "Portal", {
  enumerable: true,
  get: function () {
    return _PopoverPortal.PopoverPortal;
  }
});
Object.defineProperty(exports, "Positioner", {
  enumerable: true,
  get: function () {
    return _PopoverPositioner.PopoverPositioner;
  }
});
Object.defineProperty(exports, "Root", {
  enumerable: true,
  get: function () {
    return _PopoverRoot.PopoverRoot;
  }
});
Object.defineProperty(exports, "Title", {
  enumerable: true,
  get: function () {
    return _PopoverTitle.PopoverTitle;
  }
});
Object.defineProperty(exports, "Trigger", {
  enumerable: true,
  get: function () {
    return _PopoverTrigger.PopoverTrigger;
  }
});
Object.defineProperty(exports, "Viewport", {
  enumerable: true,
  get: function () {
    return _PopoverViewport.PopoverViewport;
  }
});
Object.defineProperty(exports, "createHandle", {
  enumerable: true,
  get: function () {
    return _PopoverHandle.createPopoverHandle;
  }
});
var _PopoverRoot = require("./root/PopoverRoot");
var _PopoverTrigger = require("./trigger/PopoverTrigger");
var _PopoverPortal = require("./portal/PopoverPortal");
var _PopoverPositioner = require("./positioner/PopoverPositioner");
var _PopoverPopup = require("./popup/PopoverPopup");
var _PopoverArrow = require("./arrow/PopoverArrow");
var _PopoverBackdrop = require("./backdrop/PopoverBackdrop");
var _PopoverTitle = require("./title/PopoverTitle");
var _PopoverDescription = require("./description/PopoverDescription");
var _PopoverClose = require("./close/PopoverClose");
var _PopoverViewport = require("./viewport/PopoverViewport");
var _PopoverHandle = require("./store/PopoverHandle");
