// @base-ui/react/esm/menu/checkbox-item-indicator/MenuCheckboxItemIndicatorDataAttributes.js
import { TransitionStatusDataAttributes } from "../../utils/stateAttributesMapping.js";
export let MenuCheckboxItemIndicatorDataAttributes = function (MenuCheckboxItemIndicatorDataAttributes) {
  /**
   * Present when the menu checkbox item is checked.
   */
  MenuCheckboxItemIndicatorDataAttributes["checked"] = "data-checked";
  /**
   * Present when the menu checkbox item is not checked.
   */
  MenuCheckboxItemIndicatorDataAttributes["unchecked"] = "data-unchecked";
  /**
   * Present when the menu checkbox item is disabled.
   */
  MenuCheckboxItemIndicatorDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the indicator is animating in.
   */
  MenuCheckboxItemIndicatorDataAttributes[MenuCheckboxItemIndicatorDataAttributes["startingStyle"] = TransitionStatusDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the indicator is animating out.
   */
  MenuCheckboxItemIndicatorDataAttributes[MenuCheckboxItemIndicatorDataAttributes["endingStyle"] = TransitionStatusDataAttributes.endingStyle] = "endingStyle";
  return MenuCheckboxItemIndicatorDataAttributes;
}({});
