// @base-ui/react/esm/navigation-menu/arrow/NavigationMenuArrowDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let NavigationMenuArrowDataAttributes = function (NavigationMenuArrowDataAttributes) {
  /**
   * Present when the popup is open.
   */
  NavigationMenuArrowDataAttributes[NavigationMenuArrowDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  NavigationMenuArrowDataAttributes[NavigationMenuArrowDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  NavigationMenuArrowDataAttributes[NavigationMenuArrowDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  NavigationMenuArrowDataAttributes[NavigationMenuArrowDataAttributes["align"] = CommonPopupDataAttributes.align] = "align";
  /**
   * Present when the popup arrow is uncentered.
   */
  NavigationMenuArrowDataAttributes["uncentered"] = "data-uncentered";
  return NavigationMenuArrowDataAttributes;
}({});
