// @base-ui/react/combobox/popup/ComboboxPopupDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComboboxPopupDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let ComboboxPopupDataAttributes = exports.ComboboxPopupDataAttributes = function (ComboboxPopupDataAttributes) {
  /**
   * Present when the popup is open.
   */
  ComboboxPopupDataAttributes[ComboboxPopupDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  ComboboxPopupDataAttributes[ComboboxPopupDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the popup is animating in.
   */
  ComboboxPopupDataAttributes[ComboboxPopupDataAttributes["startingStyle"] = _popupStateMapping.CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the popup is animating out.
   */
  ComboboxPopupDataAttributes[ComboboxPopupDataAttributes["endingStyle"] = _popupStateMapping.CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  ComboboxPopupDataAttributes[ComboboxPopupDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  ComboboxPopupDataAttributes[ComboboxPopupDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  /**
   * Present if animations should be instant.
   * @type {'click' | 'dismiss'}
   */
  ComboboxPopupDataAttributes["instant"] = "data-instant";
  /**
   * Present when the items list is empty.
   */
  ComboboxPopupDataAttributes["empty"] = "data-empty";
  return ComboboxPopupDataAttributes;
}({});
