// @base-ui/react/popover/popup/PopoverPopupDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PopoverPopupDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let PopoverPopupDataAttributes = exports.PopoverPopupDataAttributes = function (PopoverPopupDataAttributes) {
  /**
   * Present when the popup is open.
   */
  PopoverPopupDataAttributes[PopoverPopupDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  PopoverPopupDataAttributes[PopoverPopupDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the popup is animating in.
   */
  PopoverPopupDataAttributes[PopoverPopupDataAttributes["startingStyle"] = _popupStateMapping.CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the popup is animating out.
   */
  PopoverPopupDataAttributes[PopoverPopupDataAttributes["endingStyle"] = _popupStateMapping.CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  PopoverPopupDataAttributes["side"] = "data-side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  PopoverPopupDataAttributes["align"] = "data-align";
  /**
   * Present if animations should be instant.
   * @type {'click' | 'dismiss'}
   */
  PopoverPopupDataAttributes["instant"] = "data-instant";
  return PopoverPopupDataAttributes;
}({});
