// @base-ui/react/esm/input/InputDataAttributes.js
export let InputDataAttributes = /*#__PURE__*/function (InputDataAttributes) {
  /**
   * Present when the input is disabled.
   */
  InputDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the input is in valid state (when wrapped in Field.Root).
   */
  InputDataAttributes["valid"] = "data-valid";
  /**
   * Present when the input is in invalid state (when wrapped in Field.Root).
   */
  InputDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the input has been touched (when wrapped in Field.Root).
   */
  InputDataAttributes["touched"] = "data-touched";
  /**
   * Present when the input's value has changed (when wrapped in Field.Root).
   */
  InputDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the input is filled (when wrapped in Field.Root).
   */
  InputDataAttributes["filled"] = "data-filled";
  /**
   * Present when the input is focused (when wrapped in Field.Root).
   */
  InputDataAttributes["focused"] = "data-focused";
  return InputDataAttributes;
}({});
