// @base-ui/react/esm/preview-card/arrow/PreviewCardArrowDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let PreviewCardArrowDataAttributes = function (PreviewCardArrowDataAttributes) {
  /**
   * Present when the preview card is open.
   */
  PreviewCardArrowDataAttributes[PreviewCardArrowDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the preview card is closed.
   */
  PreviewCardArrowDataAttributes[PreviewCardArrowDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  PreviewCardArrowDataAttributes[PreviewCardArrowDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  PreviewCardArrowDataAttributes[PreviewCardArrowDataAttributes["align"] = CommonPopupDataAttributes.align] = "align";
  /**
   * Present when the preview card arrow is uncentered.
   */
  PreviewCardArrowDataAttributes["uncentered"] = "data-uncentered";
  return PreviewCardArrowDataAttributes;
}({});
