// @base-ui/react/number-field/index.parts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Decrement", {
  enumerable: true,
  get: function () {
    return _NumberFieldDecrement.NumberFieldDecrement;
  }
});
Object.defineProperty(exports, "Group", {
  enumerable: true,
  get: function () {
    return _NumberFieldGroup.NumberFieldGroup;
  }
});
Object.defineProperty(exports, "Increment", {
  enumerable: true,
  get: function () {
    return _NumberFieldIncrement.NumberFieldIncrement;
  }
});
Object.defineProperty(exports, "Input", {
  enumerable: true,
  get: function () {
    return _NumberFieldInput.NumberFieldInput;
  }
});
Object.defineProperty(exports, "Root", {
  enumerable: true,
  get: function () {
    return _NumberFieldRoot.NumberFieldRoot;
  }
});
Object.defineProperty(exports, "ScrubArea", {
  enumerable: true,
  get: function () {
    return _NumberFieldScrubArea.NumberFieldScrubArea;
  }
});
Object.defineProperty(exports, "ScrubAreaCursor", {
  enumerable: true,
  get: function () {
    return _NumberFieldScrubAreaCursor.NumberFieldScrubAreaCursor;
  }
});
var _NumberFieldRoot = require("./root/NumberFieldRoot");
var _NumberFieldGroup = require("./group/NumberFieldGroup");
var _NumberFieldIncrement = require("./increment/NumberFieldIncrement");
var _NumberFieldDecrement = require("./decrement/NumberFieldDecrement");
var _NumberFieldInput = require("./input/NumberFieldInput");
var _NumberFieldScrubArea = require("./scrub-area/NumberFieldScrubArea");
var _NumberFieldScrubAreaCursor = require("./scrub-area-cursor/NumberFieldScrubAreaCursor");
