// @base-ui/react/esm/combobox/trigger/ComboboxTriggerDataAttributes.js
import { CommonTriggerDataAttributes } from "../../utils/popupStateMapping.js";
export let ComboboxTriggerDataAttributes = function (ComboboxTriggerDataAttributes) {
  /**
   * Present when the corresponding popup is open.
   */
  ComboboxTriggerDataAttributes[ComboboxTriggerDataAttributes["popupOpen"] = CommonTriggerDataAttributes.popupOpen] = "popupOpen";
  /**
   * Present when the trigger is pressed.
   */
  ComboboxTriggerDataAttributes[ComboboxTriggerDataAttributes["pressed"] = CommonTriggerDataAttributes.pressed] = "pressed";
  /**
   * Present when the component is disabled.
   */
  ComboboxTriggerDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the component is readonly.
   */
  ComboboxTriggerDataAttributes["readonly"] = "data-readonly";
  /**
   * Indicates which side the corresponding popup is positioned relative to its anchor.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start' | null}
   */
  ComboboxTriggerDataAttributes["popupSide"] = "data-popup-side";
  /**
   * Present when the component is required.
   */
  ComboboxTriggerDataAttributes["required"] = "data-required";
  /**
   * Present when the component is in valid state (when wrapped in Field.Root).
   */
  ComboboxTriggerDataAttributes["valid"] = "data-valid";
  /**
   * Present when the component is in invalid state (when wrapped in Field.Root).
   */
  ComboboxTriggerDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the component has been touched (when wrapped in Field.Root).
   */
  ComboboxTriggerDataAttributes["touched"] = "data-touched";
  /**
   * Present when the component's value has changed (when wrapped in Field.Root).
   */
  ComboboxTriggerDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the component has a value (when wrapped in Field.Root).
   */
  ComboboxTriggerDataAttributes["filled"] = "data-filled";
  /**
   * Present when the trigger is focused (when wrapped in Field.Root).
   */
  ComboboxTriggerDataAttributes["focused"] = "data-focused";
  /**
   * Present when the corresponding items list is empty.
   */
  ComboboxTriggerDataAttributes["listEmpty"] = "data-list-empty";
  /**
   * Present when the combobox doesn't have a value.
   */
  ComboboxTriggerDataAttributes["placeholder"] = "data-placeholder";
  return ComboboxTriggerDataAttributes;
}({});
