// @base-ui/react/select/trigger/SelectTriggerDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SelectTriggerDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let SelectTriggerDataAttributes = exports.SelectTriggerDataAttributes = function (SelectTriggerDataAttributes) {
  /**
   * Present when the corresponding select is open.
   */
  SelectTriggerDataAttributes[SelectTriggerDataAttributes["popupOpen"] = _popupStateMapping.CommonTriggerDataAttributes.popupOpen] = "popupOpen";
  /**
   * Present when the trigger is pressed.
   */
  SelectTriggerDataAttributes[SelectTriggerDataAttributes["pressed"] = _popupStateMapping.CommonTriggerDataAttributes.pressed] = "pressed";
  /**
   * Present when the select is disabled.
   */
  SelectTriggerDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the select is readonly.
   */
  SelectTriggerDataAttributes["readonly"] = "data-readonly";
  /**
   * Present when the select is required.
   */
  SelectTriggerDataAttributes["required"] = "data-required";
  /**
   * Present when the select is in valid state (when wrapped in Field.Root).
   */
  SelectTriggerDataAttributes["valid"] = "data-valid";
  /**
   * Present when the select is in invalid state (when wrapped in Field.Root).
   */
  SelectTriggerDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the select has been touched (when wrapped in Field.Root).
   */
  SelectTriggerDataAttributes["touched"] = "data-touched";
  /**
   * Present when the select's value has changed (when wrapped in Field.Root).
   */
  SelectTriggerDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the select has a value (when wrapped in Field.Root).
   */
  SelectTriggerDataAttributes["filled"] = "data-filled";
  /**
   * Present when the select trigger is focused (when wrapped in Field.Root).
   */
  SelectTriggerDataAttributes["focused"] = "data-focused";
  /**
   * Present when the select doesn't have a value.
   */
  SelectTriggerDataAttributes["placeholder"] = "data-placeholder";
  return SelectTriggerDataAttributes;
}({});
