// @base-ui/react/navigation-menu/arrow/NavigationMenuArrowDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NavigationMenuArrowDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let NavigationMenuArrowDataAttributes = exports.NavigationMenuArrowDataAttributes = function (NavigationMenuArrowDataAttributes) {
  /**
   * Present when the popup is open.
   */
  NavigationMenuArrowDataAttributes[NavigationMenuArrowDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  NavigationMenuArrowDataAttributes[NavigationMenuArrowDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  NavigationMenuArrowDataAttributes[NavigationMenuArrowDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  NavigationMenuArrowDataAttributes[NavigationMenuArrowDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  /**
   * Present when the popup arrow is uncentered.
   */
  NavigationMenuArrowDataAttributes["uncentered"] = "data-uncentered";
  return NavigationMenuArrowDataAttributes;
}({});
