// @base-ui/react/esm/combobox/input-group/ComboboxInputGroupDataAttributes.js
export let ComboboxInputGroupDataAttributes = /*#__PURE__*/function (ComboboxInputGroupDataAttributes) {
  /**
   * Present when the corresponding popup is open.
   */
  ComboboxInputGroupDataAttributes["popupOpen"] = "data-popup-open";
  /**
   * Present when the input group is pressed.
   */
  ComboboxInputGroupDataAttributes["pressed"] = "data-pressed";
  /**
   * Present when the component is disabled.
   */
  ComboboxInputGroupDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the component is readonly.
   */
  ComboboxInputGroupDataAttributes["readonly"] = "data-readonly";
  /**
   * Indicates which side the corresponding popup is positioned relative to its anchor.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start' | null}
   */
  ComboboxInputGroupDataAttributes["popupSide"] = "data-popup-side";
  /**
   * Present when the component is in valid state (when wrapped in Field.Root).
   */
  ComboboxInputGroupDataAttributes["valid"] = "data-valid";
  /**
   * Present when the component is in invalid state (when wrapped in Field.Root).
   */
  ComboboxInputGroupDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the component has been touched (when wrapped in Field.Root).
   */
  ComboboxInputGroupDataAttributes["touched"] = "data-touched";
  /**
   * Present when the component's value has changed (when wrapped in Field.Root).
   */
  ComboboxInputGroupDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the component has a value (when wrapped in Field.Root).
   */
  ComboboxInputGroupDataAttributes["filled"] = "data-filled";
  /**
   * Present when the component is focused (when wrapped in Field.Root).
   */
  ComboboxInputGroupDataAttributes["focused"] = "data-focused";
  /**
   * Present when the corresponding items list is empty.
   */
  ComboboxInputGroupDataAttributes["listEmpty"] = "data-list-empty";
  /**
   * Present when the combobox doesn't have a value.
   */
  ComboboxInputGroupDataAttributes["placeholder"] = "data-placeholder";
  return ComboboxInputGroupDataAttributes;
}({});
