// @base-ui/react/slider/indicator/SliderIndicatorDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SliderIndicatorDataAttributes = void 0;
let SliderIndicatorDataAttributes = exports.SliderIndicatorDataAttributes = /*#__PURE__*/function (SliderIndicatorDataAttributes) {
  /**
   * Present while the user is dragging.
   */
  SliderIndicatorDataAttributes["dragging"] = "data-dragging";
  /**
   * Indicates the orientation of the slider.
   * @type {'horizontal' | 'vertical'}
   */
  SliderIndicatorDataAttributes["orientation"] = "data-orientation";
  /**
   * Present when the slider is disabled.
   */
  SliderIndicatorDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the slider is in valid state (when wrapped in Field.Root).
   */
  SliderIndicatorDataAttributes["valid"] = "data-valid";
  /**
   * Present when the slider is in invalid state (when wrapped in Field.Root).
   */
  SliderIndicatorDataAttributes["invalid"] = "data-invalid";
  /**
   * Present when the slider has been touched (when wrapped in Field.Root).
   */
  SliderIndicatorDataAttributes["touched"] = "data-touched";
  /**
   * Present when the slider's value has changed (when wrapped in Field.Root).
   */
  SliderIndicatorDataAttributes["dirty"] = "data-dirty";
  /**
   * Present when the slider is focused (when wrapped in Field.Root).
   */
  SliderIndicatorDataAttributes["focused"] = "data-focused";
  return SliderIndicatorDataAttributes;
}({});
