// @base-ui/react/popover/arrow/PopoverArrowDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PopoverArrowDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let PopoverArrowDataAttributes = exports.PopoverArrowDataAttributes = function (PopoverArrowDataAttributes) {
  /**
   * Present when the popup is open.
   */
  PopoverArrowDataAttributes[PopoverArrowDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  PopoverArrowDataAttributes[PopoverArrowDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  PopoverArrowDataAttributes[PopoverArrowDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  PopoverArrowDataAttributes[PopoverArrowDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  /**
   * Present when the popover arrow is uncentered.
   */
  PopoverArrowDataAttributes["uncentered"] = "data-uncentered";
  return PopoverArrowDataAttributes;
}({});
