// @base-ui/react/tooltip/arrow/TooltipArrowDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TooltipArrowDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let TooltipArrowDataAttributes = exports.TooltipArrowDataAttributes = function (TooltipArrowDataAttributes) {
  /**
   * Present when the tooltip is open.
   */
  TooltipArrowDataAttributes[TooltipArrowDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the tooltip is closed.
   */
  TooltipArrowDataAttributes[TooltipArrowDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  TooltipArrowDataAttributes[TooltipArrowDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  TooltipArrowDataAttributes[TooltipArrowDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  /**
   * Present when the tooltip arrow is uncentered.
   */
  TooltipArrowDataAttributes["uncentered"] = "data-uncentered";
  /**
   * Present if animations should be instant.
   * @type {'delay' | 'dismiss' | 'focus'}
   */
  TooltipArrowDataAttributes["instant"] = "data-instant";
  return TooltipArrowDataAttributes;
}({});
