// @base-ui/react/popover/viewport/PopoverViewportDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PopoverViewportDataAttributes = void 0;
let PopoverViewportDataAttributes = exports.PopoverViewportDataAttributes = /*#__PURE__*/function (PopoverViewportDataAttributes) {
  /**
   * Applied to the direct child of the viewport when no transitions are present or the new content when it's entering.
   */
  PopoverViewportDataAttributes["current"] = "data-current";
  /**
   * Applied to the direct child of the viewport that contains the exiting content when transitions are present.
   */
  PopoverViewportDataAttributes["previous"] = "data-previous";
  /**
   * Indicates the direction from which the popup was activated.
   * This can be used to create directional animations based on how the popup was triggered.
   * Contains space-separated values for both horizontal and vertical axes.
   * @type {`${'left' | 'right'} {'top' | 'bottom'}`}
   */
  PopoverViewportDataAttributes["activationDirection"] = "data-activation-direction";
  /**
   * Indicates that the viewport is currently transitioning between old and new content.
   */
  PopoverViewportDataAttributes["transitioning"] = "data-transitioning";
  /**
   * Present if animations should be instant.
   * @type {'dismiss' | 'click'}
   */
  PopoverViewportDataAttributes["instant"] = "data-instant";
  return PopoverViewportDataAttributes;
}({});
