// @base-ui/react/tooltip/popup/TooltipPopupDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TooltipPopupDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let TooltipPopupDataAttributes = exports.TooltipPopupDataAttributes = function (TooltipPopupDataAttributes) {
  /**
   * Present when the tooltip is open.
   */
  TooltipPopupDataAttributes[TooltipPopupDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the tooltip is closed.
   */
  TooltipPopupDataAttributes[TooltipPopupDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the tooltip is animating in.
   */
  TooltipPopupDataAttributes[TooltipPopupDataAttributes["startingStyle"] = _popupStateMapping.CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the tooltip is animating out.
   */
  TooltipPopupDataAttributes[TooltipPopupDataAttributes["endingStyle"] = _popupStateMapping.CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  TooltipPopupDataAttributes[TooltipPopupDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  TooltipPopupDataAttributes[TooltipPopupDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  /**
   * Present if animations should be instant.
   * @type {'delay' | 'dismiss' | 'focus'}
   */
  TooltipPopupDataAttributes["instant"] = "data-instant";
  return TooltipPopupDataAttributes;
}({});
