// @base-ui/react/drawer/popup/DrawerPopupDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DrawerPopupDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let DrawerPopupDataAttributes = exports.DrawerPopupDataAttributes = function (DrawerPopupDataAttributes) {
  /**
   * Present when the drawer is open.
   */
  DrawerPopupDataAttributes[DrawerPopupDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the drawer is closed.
   */
  DrawerPopupDataAttributes[DrawerPopupDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the drawer is animating in.
   */
  DrawerPopupDataAttributes[DrawerPopupDataAttributes["startingStyle"] = _popupStateMapping.CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the drawer is animating out.
   */
  DrawerPopupDataAttributes[DrawerPopupDataAttributes["endingStyle"] = _popupStateMapping.CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Present when the drawer is at the expanded (full-height) snap point.
   */
  DrawerPopupDataAttributes["expanded"] = "data-expanded";
  /**
   * Present when a nested drawer is open.
   */
  DrawerPopupDataAttributes["nestedDrawerOpen"] = "data-nested-drawer-open";
  /**
   * Present when a nested drawer is being swiped.
   */
  DrawerPopupDataAttributes["nestedDrawerSwiping"] = "data-nested-drawer-swiping";
  /**
   * Present when the drawer is dismissed by swiping.
   */
  DrawerPopupDataAttributes["swipeDismiss"] = "data-swipe-dismiss";
  /**
   * Indicates the swipe direction.
   * @type {'up' | 'down' | 'left' | 'right'}
   */
  DrawerPopupDataAttributes["swipeDirection"] = "data-swipe-direction";
  /**
   * Present when the drawer is being swiped.
   */
  DrawerPopupDataAttributes["swiping"] = "data-swiping";
  return DrawerPopupDataAttributes;
}({});
