// @base-ui/react/context-menu/index.parts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Arrow", {
  enumerable: true,
  get: function () {
    return _MenuArrow.MenuArrow;
  }
});
Object.defineProperty(exports, "Backdrop", {
  enumerable: true,
  get: function () {
    return _MenuBackdrop.MenuBackdrop;
  }
});
Object.defineProperty(exports, "CheckboxItem", {
  enumerable: true,
  get: function () {
    return _MenuCheckboxItem.MenuCheckboxItem;
  }
});
Object.defineProperty(exports, "CheckboxItemIndicator", {
  enumerable: true,
  get: function () {
    return _MenuCheckboxItemIndicator.MenuCheckboxItemIndicator;
  }
});
Object.defineProperty(exports, "Group", {
  enumerable: true,
  get: function () {
    return _MenuGroup.MenuGroup;
  }
});
Object.defineProperty(exports, "GroupLabel", {
  enumerable: true,
  get: function () {
    return _MenuGroupLabel.MenuGroupLabel;
  }
});
Object.defineProperty(exports, "Item", {
  enumerable: true,
  get: function () {
    return _MenuItem.MenuItem;
  }
});
Object.defineProperty(exports, "LinkItem", {
  enumerable: true,
  get: function () {
    return _MenuLinkItem.MenuLinkItem;
  }
});
Object.defineProperty(exports, "Popup", {
  enumerable: true,
  get: function () {
    return _MenuPopup.MenuPopup;
  }
});
Object.defineProperty(exports, "Portal", {
  enumerable: true,
  get: function () {
    return _MenuPortal.MenuPortal;
  }
});
Object.defineProperty(exports, "Positioner", {
  enumerable: true,
  get: function () {
    return _MenuPositioner.MenuPositioner;
  }
});
Object.defineProperty(exports, "RadioGroup", {
  enumerable: true,
  get: function () {
    return _MenuRadioGroup.MenuRadioGroup;
  }
});
Object.defineProperty(exports, "RadioItem", {
  enumerable: true,
  get: function () {
    return _MenuRadioItem.MenuRadioItem;
  }
});
Object.defineProperty(exports, "RadioItemIndicator", {
  enumerable: true,
  get: function () {
    return _MenuRadioItemIndicator.MenuRadioItemIndicator;
  }
});
Object.defineProperty(exports, "Root", {
  enumerable: true,
  get: function () {
    return _ContextMenuRoot.ContextMenuRoot;
  }
});
Object.defineProperty(exports, "Separator", {
  enumerable: true,
  get: function () {
    return _Separator.Separator;
  }
});
Object.defineProperty(exports, "SubmenuRoot", {
  enumerable: true,
  get: function () {
    return _MenuSubmenuRoot.MenuSubmenuRoot;
  }
});
Object.defineProperty(exports, "SubmenuTrigger", {
  enumerable: true,
  get: function () {
    return _MenuSubmenuTrigger.MenuSubmenuTrigger;
  }
});
Object.defineProperty(exports, "Trigger", {
  enumerable: true,
  get: function () {
    return _ContextMenuTrigger.ContextMenuTrigger;
  }
});
var _ContextMenuRoot = require("./root/ContextMenuRoot");
var _ContextMenuTrigger = require("./trigger/ContextMenuTrigger");
var _MenuBackdrop = require("../menu/backdrop/MenuBackdrop");
var _MenuPortal = require("../menu/portal/MenuPortal");
var _MenuPositioner = require("../menu/positioner/MenuPositioner");
var _MenuPopup = require("../menu/popup/MenuPopup");
var _MenuArrow = require("../menu/arrow/MenuArrow");
var _MenuGroup = require("../menu/group/MenuGroup");
var _MenuGroupLabel = require("../menu/group-label/MenuGroupLabel");
var _MenuItem = require("../menu/item/MenuItem");
var _MenuCheckboxItem = require("../menu/checkbox-item/MenuCheckboxItem");
var _MenuCheckboxItemIndicator = require("../menu/checkbox-item-indicator/MenuCheckboxItemIndicator");
var _MenuLinkItem = require("../menu/link-item/MenuLinkItem");
var _MenuRadioGroup = require("../menu/radio-group/MenuRadioGroup");
var _MenuRadioItem = require("../menu/radio-item/MenuRadioItem");
var _MenuRadioItemIndicator = require("../menu/radio-item-indicator/MenuRadioItemIndicator");
var _MenuSubmenuRoot = require("../menu/submenu-root/MenuSubmenuRoot");
var _MenuSubmenuTrigger = require("../menu/submenu-trigger/MenuSubmenuTrigger");
var _Separator = require("../separator/Separator");
