// @base-ui/react/esm/popover/positioner/PopoverPositionerCssVars.js
export let PopoverPositionerCssVars = /*#__PURE__*/function (PopoverPositionerCssVars) {
  /**
   * The available width between the trigger and the edge of the viewport.
   * @type {number}
   */
  PopoverPositionerCssVars["availableWidth"] = "--available-width";
  /**
   * The available height between the trigger and the edge of the viewport.
   * @type {number}
   */
  PopoverPositionerCssVars["availableHeight"] = "--available-height";
  /**
   * The anchor's width.
   * @type {number}
   */
  PopoverPositionerCssVars["anchorWidth"] = "--anchor-width";
  /**
   * The anchor's height.
   * @type {number}
   */
  PopoverPositionerCssVars["anchorHeight"] = "--anchor-height";
  /**
   * The coordinates that this element is anchored to. Used for animations and transitions.
   * @type {string}
   */
  PopoverPositionerCssVars["transformOrigin"] = "--transform-origin";
  /**
   * The width of the popover's positioner.
   * It is important to set `width` to this value when using CSS to animate size changes.
   */
  PopoverPositionerCssVars["positionerWidth"] = "--positioner-width";
  /**
   * The height of the popover's positioner.
   * It is important to set `height` to this value when using CSS to animate size changes.
   */
  PopoverPositionerCssVars["positionerHeight"] = "--positioner-height";
  return PopoverPositionerCssVars;
}({});
