// @base-ui/react/esm/popover/positioner/PopoverPositionerDataAttributes.js
import { CommonPopupDataAttributes } from "../../utils/popupStateMapping.js";
export let PopoverPositionerDataAttributes = function (PopoverPositionerDataAttributes) {
  /**
   * Present when the popup is open.
   */
  PopoverPositionerDataAttributes[PopoverPositionerDataAttributes["open"] = CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  PopoverPositionerDataAttributes[PopoverPositionerDataAttributes["closed"] = CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  PopoverPositionerDataAttributes[PopoverPositionerDataAttributes["anchorHidden"] = CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  PopoverPositionerDataAttributes[PopoverPositionerDataAttributes["side"] = CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  PopoverPositionerDataAttributes[PopoverPositionerDataAttributes["align"] = CommonPopupDataAttributes.align] = "align";
  return PopoverPositionerDataAttributes;
}({});
