// @base-ui/react/preview-card/positioner/PreviewCardPositionerDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PreviewCardPositionerDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let PreviewCardPositionerDataAttributes = exports.PreviewCardPositionerDataAttributes = function (PreviewCardPositionerDataAttributes) {
  /**
   * Present when the preview card is open.
   */
  PreviewCardPositionerDataAttributes[PreviewCardPositionerDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the preview card is closed.
   */
  PreviewCardPositionerDataAttributes[PreviewCardPositionerDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  PreviewCardPositionerDataAttributes[PreviewCardPositionerDataAttributes["anchorHidden"] = _popupStateMapping.CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  PreviewCardPositionerDataAttributes[PreviewCardPositionerDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  PreviewCardPositionerDataAttributes[PreviewCardPositionerDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  return PreviewCardPositionerDataAttributes;
}({});
