// @base-ui/react/preview-card/positioner/PreviewCardPositionerCssVars.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PreviewCardPositionerCssVars = void 0;
let PreviewCardPositionerCssVars = exports.PreviewCardPositionerCssVars = /*#__PURE__*/function (PreviewCardPositionerCssVars) {
  /**
   * The available width between the trigger and the edge of the viewport.
   * @type {number}
   */
  PreviewCardPositionerCssVars["availableWidth"] = "--available-width";
  /**
   * The available height between the trigger and the edge of the viewport.
   * @type {number}
   */
  PreviewCardPositionerCssVars["availableHeight"] = "--available-height";
  /**
   * The anchor's width.
   * @type {number}
   */
  PreviewCardPositionerCssVars["anchorWidth"] = "--anchor-width";
  /**
   * The anchor's height.
   * @type {number}
   */
  PreviewCardPositionerCssVars["anchorHeight"] = "--anchor-height";
  /**
   * The coordinates that this element is anchored to. Used for animations and transitions.
   * @type {string}
   */
  PreviewCardPositionerCssVars["transformOrigin"] = "--transform-origin";
  return PreviewCardPositionerCssVars;
}({});
