// @base-ui/react/navigation-menu/positioner/NavigationPositionerDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NavigationMenuPositionerDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let NavigationMenuPositionerDataAttributes = exports.NavigationMenuPositionerDataAttributes = function (NavigationMenuPositionerDataAttributes) {
  /**
   * Present when the popup is open.
   */
  NavigationMenuPositionerDataAttributes[NavigationMenuPositionerDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  NavigationMenuPositionerDataAttributes[NavigationMenuPositionerDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  NavigationMenuPositionerDataAttributes[NavigationMenuPositionerDataAttributes["anchorHidden"] = _popupStateMapping.CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  NavigationMenuPositionerDataAttributes[NavigationMenuPositionerDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to the specified side.
   * @type {'start' | 'center' | 'end'}
   */
  NavigationMenuPositionerDataAttributes[NavigationMenuPositionerDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  /**
   * Present if animations should be instant.
   */
  NavigationMenuPositionerDataAttributes["instant"] = "data-instant";
  return NavigationMenuPositionerDataAttributes;
}({});
