// @base-ui/react/drawer/viewport/DrawerViewportDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DrawerViewportDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let DrawerViewportDataAttributes = exports.DrawerViewportDataAttributes = function (DrawerViewportDataAttributes) {
  /**
   * Present when the drawer is open.
   */
  DrawerViewportDataAttributes[DrawerViewportDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the drawer is closed.
   */
  DrawerViewportDataAttributes[DrawerViewportDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the drawer is animating in.
   */
  DrawerViewportDataAttributes[DrawerViewportDataAttributes["startingStyle"] = _popupStateMapping.CommonPopupDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the drawer is animating out.
   */
  DrawerViewportDataAttributes[DrawerViewportDataAttributes["endingStyle"] = _popupStateMapping.CommonPopupDataAttributes.endingStyle] = "endingStyle";
  /**
   * Present when the drawer is nested within another drawer.
   */
  DrawerViewportDataAttributes["nested"] = "data-nested";
  return DrawerViewportDataAttributes;
}({});
