// @base-ui/react/accordion/panel/AccordionPanelDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AccordionPanelDataAttributes = void 0;
var _stateAttributesMapping = require("../../utils/stateAttributesMapping");
let AccordionPanelDataAttributes = exports.AccordionPanelDataAttributes = function (AccordionPanelDataAttributes) {
  /**
   * Indicates the index of the accordion item.
   * @type {number}
   */
  AccordionPanelDataAttributes["index"] = "data-index";
  /**
   * Present when the accordion panel is open.
   */
  AccordionPanelDataAttributes["open"] = "data-open";
  /**
   * Indicates the orientation of the accordion.
   */
  AccordionPanelDataAttributes["orientation"] = "data-orientation";
  /**
   * Present when the accordion item is disabled.
   */
  AccordionPanelDataAttributes["disabled"] = "data-disabled";
  /**
   * Present when the panel is animating in.
   */
  AccordionPanelDataAttributes[AccordionPanelDataAttributes["startingStyle"] = _stateAttributesMapping.TransitionStatusDataAttributes.startingStyle] = "startingStyle";
  /**
   * Present when the panel is animating out.
   */
  AccordionPanelDataAttributes[AccordionPanelDataAttributes["endingStyle"] = _stateAttributesMapping.TransitionStatusDataAttributes.endingStyle] = "endingStyle";
  return AccordionPanelDataAttributes;
}({});
