// @base-ui/react/combobox/positioner/ComboboxPositionerDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComboboxPositionerDataAttributes = void 0;
var _popupStateMapping = require("../../utils/popupStateMapping");
let ComboboxPositionerDataAttributes = exports.ComboboxPositionerDataAttributes = function (ComboboxPositionerDataAttributes) {
  /**
   * Present when the popup is open.
   */
  ComboboxPositionerDataAttributes[ComboboxPositionerDataAttributes["open"] = _popupStateMapping.CommonPopupDataAttributes.open] = "open";
  /**
   * Present when the popup is closed.
   */
  ComboboxPositionerDataAttributes[ComboboxPositionerDataAttributes["closed"] = _popupStateMapping.CommonPopupDataAttributes.closed] = "closed";
  /**
   * Present when the anchor is hidden.
   */
  ComboboxPositionerDataAttributes[ComboboxPositionerDataAttributes["anchorHidden"] = _popupStateMapping.CommonPopupDataAttributes.anchorHidden] = "anchorHidden";
  /**
   * Indicates which side the popup is positioned relative to the trigger.
   * @type {'top' | 'bottom' | 'left' | 'right' | 'inline-end' | 'inline-start'}
   */
  ComboboxPositionerDataAttributes[ComboboxPositionerDataAttributes["side"] = _popupStateMapping.CommonPopupDataAttributes.side] = "side";
  /**
   * Indicates how the popup is aligned relative to specified side.
   * @type {'start' | 'center' | 'end'}
   */
  ComboboxPositionerDataAttributes[ComboboxPositionerDataAttributes["align"] = _popupStateMapping.CommonPopupDataAttributes.align] = "align";
  /**
   * Present when the items list is empty.
   */
  ComboboxPositionerDataAttributes["empty"] = "data-empty";
  return ComboboxPositionerDataAttributes;
}({});
