// @base-ui/react/combobox/positioner/ComboboxPositionerCssVars.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComboboxPositionerCssVars = void 0;
let ComboboxPositionerCssVars = exports.ComboboxPositionerCssVars = /*#__PURE__*/function (ComboboxPositionerCssVars) {
  /**
   * The available width between the trigger and the edge of the viewport.
   * @type {number}
   */
  ComboboxPositionerCssVars["availableWidth"] = "--available-width";
  /**
   * The available height between the trigger and the edge of the viewport.
   * @type {number}
   */
  ComboboxPositionerCssVars["availableHeight"] = "--available-height";
  /**
   * The anchor's width.
   * @type {number}
   */
  ComboboxPositionerCssVars["anchorWidth"] = "--anchor-width";
  /**
   * The anchor's height.
   * @type {number}
   */
  ComboboxPositionerCssVars["anchorHeight"] = "--anchor-height";
  /**
   * The coordinates that this element is anchored to. Used for animations and transitions.
   * @type {string}
   */
  ComboboxPositionerCssVars["transformOrigin"] = "--transform-origin";
  return ComboboxPositionerCssVars;
}({});
